<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Analyse et modÃ©lisation dâ€™agroÃ©cosystÃ¨mes sur R - 10&nbsp; Association, partitionnement et ordination</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./10-imputation.html" rel="next">
<link href="./08-explorer.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./09-ordination.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Association, partitionnement et ordination</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Analyse et modÃ©lisation dâ€™agroÃ©cosystÃ¨mes sur R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/chavalli/ecologie-mathematique-R/tree/h2024" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">PrÃ©face</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">La science des donnÃ©es avec R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-tableaux.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Organisation des donnÃ©es et opÃ©rations sur des tableaux</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-visualisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualisation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Science ouverte et reproductibilitÃ©</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction Ã  Python</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07a-biostats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Biostatistiques</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07b-bayes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction Ã  lâ€™analyse bayÃ©sienne en Ã©cologie</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-explorer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Explorer R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-ordination.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Association, partitionnement et ordination</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-imputation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">DÃ©tection de valeurs aberrantes et imputation de donnÃ©es manquantes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-series-temporelles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Les sÃ©ries temporelles</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-autoapprentissage.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Introduction Ã  lâ€™autoapprentissage</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-donnees-spatiales.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Les donnÃ©es gÃ©ospatiales</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-modelisation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">ModÃ©lisation de mÃ©canismes Ã©cologiques</span></span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#espaces-danalyse" id="toc-espaces-danalyse" class="nav-link active" data-scroll-target="#espaces-danalyse"><span class="header-section-number">10.1</span> Espaces dâ€™analyse</a>
  <ul class="collapse">
<li><a href="#abondance-et-occurrence" id="toc-abondance-et-occurrence" class="nav-link" data-scroll-target="#abondance-et-occurrence"><span class="header-section-number">10.1.1</span> Abondance et occurrence</a></li>
  <li><a href="#environnement" id="toc-environnement" class="nav-link" data-scroll-target="#environnement"><span class="header-section-number">10.1.2</span> Environnement</a></li>
  </ul>
</li>
  <li>
<a href="#analyse-dassociation" id="toc-analyse-dassociation" class="nav-link" data-scroll-target="#analyse-dassociation"><span class="header-section-number">10.2</span> Analyse dâ€™association</a>
  <ul class="collapse">
<li><a href="#association-entre-objets-mode-q" id="toc-association-entre-objets-mode-q" class="nav-link" data-scroll-target="#association-entre-objets-mode-q"><span class="header-section-number">10.2.1</span> Association entre objets (mode Q)</a></li>
  <li><a href="#associations-entre-variables-mode-r" id="toc-associations-entre-variables-mode-r" class="nav-link" data-scroll-target="#associations-entre-variables-mode-r"><span class="header-section-number">10.2.2</span> Associations entre variables (mode R)</a></li>
  <li><a href="#conclusion-sur-les-associations" id="toc-conclusion-sur-les-associations" class="nav-link" data-scroll-target="#conclusion-sur-les-associations"><span class="header-section-number">10.2.3</span> Conclusion sur les associations</a></li>
  </ul>
</li>
  <li>
<a href="#partitionnement" id="toc-partitionnement" class="nav-link" data-scroll-target="#partitionnement"><span class="header-section-number">10.3</span> Partitionnement</a>
  <ul class="collapse">
<li><a href="#%C3%A9valuation-dun-partitionnement" id="toc-Ã©valuation-dun-partitionnement" class="nav-link" data-scroll-target="#%C3%A9valuation-dun-partitionnement"><span class="header-section-number">10.3.1</span> Ã‰valuation dâ€™un partitionnement</a></li>
  <li><a href="#partitionnement-non-hi%C3%A9rarchique" id="toc-partitionnement-non-hiÃ©rarchique" class="nav-link" data-scroll-target="#partitionnement-non-hi%C3%A9rarchique"><span class="header-section-number">10.3.2</span> Partitionnement non hiÃ©rarchique</a></li>
  <li><a href="#partitionnement-hi%C3%A9rarchique" id="toc-partitionnement-hiÃ©rarchique" class="nav-link" data-scroll-target="#partitionnement-hi%C3%A9rarchique"><span class="header-section-number">10.3.3</span> Partitionnement hiÃ©rarchique</a></li>
  <li><a href="#partitionnement-hi%C3%A9rarchique-bas%C3%A9e-sur-la-densit%C3%A9-des-points" id="toc-partitionnement-hiÃ©rarchique-basÃ©e-sur-la-densitÃ©-des-points" class="nav-link" data-scroll-target="#partitionnement-hi%C3%A9rarchique-bas%C3%A9e-sur-la-densit%C3%A9-des-points"><span class="header-section-number">10.3.4</span> Partitionnement hiÃ©rarchique basÃ©e sur la densitÃ© des points</a></li>
  <li><a href="#conclusion-sur-le-partitionnement" id="toc-conclusion-sur-le-partitionnement" class="nav-link" data-scroll-target="#conclusion-sur-le-partitionnement"><span class="header-section-number">10.3.5</span> Conclusion sur le partitionnement</a></li>
  </ul>
</li>
  <li>
<a href="#ordination" id="toc-ordination" class="nav-link" data-scroll-target="#ordination"><span class="header-section-number">10.4</span> Ordination</a>
  <ul class="collapse">
<li><a href="#ordination-non-contraignante" id="toc-ordination-non-contraignante" class="nav-link" data-scroll-target="#ordination-non-contraignante"><span class="header-section-number">10.4.1</span> Ordination non contraignante</a></li>
  <li><a href="#ordination-contraignante" id="toc-ordination-contraignante" class="nav-link" data-scroll-target="#ordination-contraignante"><span class="header-section-number">10.4.2</span> Ordination contraignante</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-ord" class="quarto-section-identifier"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Association, partitionnement et ordination</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><hr>
<p><strong>Objectifs spÃ©cifiques</strong>:</p>
<p>Ã€ la fin de ce chapitre, vous</p>
<ul>
<li>serez en mesure dâ€™effectuer des calculs permettant de mesurer des diffÃ©rence entre des observations, des groupes dâ€™observation ou des variables observÃ©es</li>
<li>serez en mesure dâ€™effectuer des analyses de partitionnement hiÃ©rarchiques et non-hiÃ©rarchiques</li>
<li>serez en mesure dâ€™effectuer des calculs dâ€™ordination Ã  lâ€™aide des techniques de rÃ©duction dâ€™axe communes: analyse en composante principale, lâ€™analyse de correspondance, lâ€™analyse en coordonnÃ©es principales, analyse discriminante linÃ©aire, lâ€™analyse de redondance et lâ€™analyse canonique des correspondances.</li>
</ul>
<hr>
<p>Les donnÃ©es Ã©cologiques incluent gÃ©nÃ©ralement plusieurs variables qui doivent Ãªtre analysÃ©es conjointement. Les techniques pour lâ€™analyse multivariÃ©e de donnÃ©es Ã©cologiques ont grandi en nombre et en complexitÃ©, laissant Ã©merger lâ€™Ã©cologie numÃ©rique, un nouveau domaine dâ€™Ã©tude scientifique initiÃ© par Pierre Legendre et Louis Legendre dont lâ€™ouvrage <em>Numerical Ecology</em>, aujourdâ€™hui Ã  sa troisiÃ¨me Ã©dition, reste un incontournable pour qui sâ€™intÃ©resse aux mathÃ©matiques sous-jacentes au domaine. Pour la rÃ©daction de ces notes, câ€™est toutefois le livre <em>Numerical ecology with R</em>, Ã©crit par Borcard et al.&nbsp;(2011) pour offrir un guide Ã  qui voudrait une approche plus appliquÃ©e.</p>
<p>Lâ€™Ã©cologie numÃ©rique sera effleurÃ©e dans ce chapitre, qui introduit Ã  trois concepts.</p>
<ol type="1">
<li>Les <strong>associations</strong> permettent de quantifier la ressemblance ou la diffÃ©rence entre deux observation (Ã©chantillons) ou variables (descripteurs). Lorsque lâ€™on a plus de deux variables ou plus de deux site, nous obtenons des matrices dâ€™association.</li>
<li>Le <strong>partitionnement</strong> permet de regrouper des observations ou des variables selon des mÃ©triques dâ€™association.</li>
<li>Lâ€™<strong>ordination</strong> vise par lâ€™intermÃ©diaire de techniques de rÃ©duction dâ€™axe Ã  mettre de lâ€™ordre dans des donnÃ©es dont le nombre Ã©levÃ© de variables peut amener Ã  des difficultÃ©s dâ€™apprÃ©ciation et dâ€™interprÃ©tation.</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://tidyverse.tidyverse.org">"tidyverse"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
âœ” dplyr     1.1.4     âœ” readr     2.1.5
âœ” forcats   1.0.0     âœ” stringr   1.5.1
âœ” ggplot2   3.4.4     âœ” tibble    3.2.1
âœ” lubridate 1.9.3     âœ” tidyr     1.3.0
âœ” purrr     1.0.2     
â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
âœ– dplyr::filter() masks stats::filter()
âœ– dplyr::lag()    masks stats::lag()
â„¹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
</div>
<section id="espaces-danalyse" class="level2" data-number="10.1"><h2 data-number="10.1" class="anchored" data-anchor-id="espaces-danalyse">
<span class="header-section-number">10.1</span> Espaces dâ€™analyse</h2>
<section id="abondance-et-occurrence" class="level3" data-number="10.1.1"><h3 data-number="10.1.1" class="anchored" data-anchor-id="abondance-et-occurrence">
<span class="header-section-number">10.1.1</span> Abondance et occurrence</h3>
<p>Lâ€™abondance est le dÃ©compte dâ€™espÃ¨ces observÃ©es, tandis que lâ€™occurrence est la prÃ©sence ou lâ€™absence dâ€™une espÃ¨ce. Le tableau suivant contient des donnÃ©es dâ€™abondance.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">abundance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span><span class="st">'Bruant familier'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Citelle Ã  poitrine rousse'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Colibri Ã  gorge rubis'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Geai bleu'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Bruant chanteur'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Chardonneret'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">9</span>, <span class="fl">6</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Bruant Ã  gorge blanche'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'MÃ©sange Ã  tÃªte noire'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>                    <span class="st">'Jaseur borÃ©al'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">66</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Ce tableau peut Ãªtre rapidement transformÃ© en donnÃ©es dâ€™occurrence, qui ne comprennent que lâ€™information boolÃ©enne de prÃ©sence (notÃ© 1) et dâ€™absence (notÃ© 0).</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">occurrence</span> <span class="op">&lt;-</span> <span class="va">abundance</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">transmute_all</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">.</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lâ€™<strong>espace des espÃ¨ces</strong> (ou des variables ou descripteurs) est celui oÃ¹ les espÃ¨ces forment les axes et oÃ¹ les sites sont positionnÃ©s dans cet espace. Il sâ€™agit dâ€™une perspective en <em>mode R</em>, qui permet principalement dâ€™identifier quels espÃ¨ces se retrouvent plus couramment ensemble.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">abundance</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">`Bruant chanteur`</span>, <span class="va">Chardonneret</span>, <span class="va">`MÃ©sange Ã  tÃªte noire`</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 Ã— 3
  `Bruant chanteur` Chardonneret `MÃ©sange Ã  tÃªte noire`
              &lt;dbl&gt;        &lt;dbl&gt;                  &lt;dbl&gt;
1                 1            0                     20
2                 0            9                      1
3                 5            6                      1
4                 2            0                      0</code></pre>
</div>
</div>
<p>Dans lâ€™<strong>espace des sites</strong> (ou les Ã©chantillons ou objets), on transpose la matrice dâ€™abondance. On passe ici en <em>mode Q</em>, oÃ¹ chaque point est une espÃ¨ce, et oÃ¹ lâ€™on peut observer quels Ã©chantillons sont similaires.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">abundance</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          [,1] [,2] [,3] [,4]
Bruant familier              1    0    0    3
Citelle Ã  poitrine rousse    1    0    0    0
Colibri Ã  gorge rubis        0    1    0    0
Geai bleu                    3    2    0    0
Bruant chanteur              1    0    5    2
Chardonneret                 0    9    6    0
Bruant Ã  gorge blanche       1    0    0    0
MÃ©sange Ã  tÃªte noire        20    1    1    0
Jaseur borÃ©al               66    0    0    0</code></pre>
</div>
</div>
</section><section id="environnement" class="level3" data-number="10.1.2"><h3 data-number="10.1.2" class="anchored" data-anchor-id="environnement">
<span class="header-section-number">10.1.2</span> Environnement</h3>
<p>Lâ€™<strong>espace de lâ€™environnement</strong> comprend souvent un autre tableau contenant lâ€™information sur lâ€™environnement oÃ¹ se trouve les espÃ¨ces: les coordonnÃ©es et lâ€™Ã©lÃ©vation, la pente, le pH du sol, la pluviomÃ©trie, etc.</p>
</section></section><section id="analyse-dassociation" class="level2" data-number="10.2"><h2 data-number="10.2" class="anchored" data-anchor-id="analyse-dassociation">
<span class="header-section-number">10.2</span> Analyse dâ€™association</h2>
<p>Nous utiliserons le terme <em>association</em> comme une <strong>mesure pour quantifier la ressemblance ou la diffÃ©rence entre deux objets (Ã©chantillons) ou variables (descripteurs)</strong>.</p>
<p>Alors que la corrÃ©lation et la covariance sont des mesures dâ€™association entre des variables (analyse en <em>mode R</em>), la <strong>similaritÃ©</strong> et la <strong>distance</strong> sont deux types de une mesure dâ€™association entre des objets (analyse en <em>mode Q</em>). Une distance de 0 est mesurÃ©e chez deux objets identiques. La distance augmente au fur et Ã  mesure que les objets sont dissociÃ©s. Une similaritÃ© ayant une valeur de 0 indique aucune association, tandis quâ€™une valeur de 1 indique une association parfaite. Ã€ lâ€™opposÃ©, la dissimilaritÃ© est Ã©gale Ã  1-similaritÃ©.</p>
<p>La distance peut Ãªtre liÃ©e Ã  la similaritÃ© par la relation:</p>
<p><span class="math display">\[distance=\sqrt{1-similaritÃ©}\]</span></p>
<p>ou</p>
<p><span class="math display">\[distance=\sqrt{dissimilaritÃ©}\]</span></p>
<p>La racine carrÃ©e permet, pour certains indices de similaritÃ©, dâ€™obtenir des propriÃ©tÃ©s euclidiennes. Pour plus de dÃ©tails, voyez le tableau 7.2 de <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre (2012)</a>.</p>
<p>Les matrices dâ€™association sont gÃ©nÃ©ralement prÃ©sentÃ©es comme des matrices carrÃ©es, dont les dimensions sont Ã©gales au nombre dâ€™objets (<em>mode Q</em>) ou de variables (<em>mode R</em>) dans le tableau. Chaque Ã©lÃ©ment (â€œcelluleâ€) de la matrice est un indice dâ€™association entre un objet (ou une variable) et un autre. Ainsi, la diagonale de la matrice est un vecteur nul (distance ou dissimilaritÃ©) ou unitaire (similaritÃ©), car elle correspond Ã  lâ€™association entre un objet et lui-mÃªme.</p>
<p>Puisque lâ€™association entre A et B est la mÃªme quâ€™entre B et A, et puisque la diagonale retourne une valeur convenue, il est possible dâ€™exprimer une matrice dâ€™association en mode â€œcompactâ€, sous forme de vecteur. Le vecteur dâ€™association entre des objets A, B et C contiendra toute lâ€™information nÃ©cessaire en un vecteur de trois chiffres, <code>[AB, AC, BC]</code>, plutÃ´t quâ€™une matrice de dimension <span class="math inline">\(3 \times 3\)</span>. Lâ€™impact sur la mÃ©moire vive peut Ãªtre considÃ©rable pour les calculs comprenant de nombreuses dimensions.</p>
<p>En R, les calculs de similaritÃ© et de distances peuvent Ãªtre effectuÃ©s avec le module vegan. La fonction <code>vegdist</code> permet de calculer les indices dâ€™association en forme carrÃ©e.</p>
<p>Nous verrons plus tard les mÃ©thodes de mesure de similaritÃ© et de distance plus loin. Pour lâ€™instant, utilisons la mÃ©thode de <em>Jaccard</em> pour une dÃ©monstration sur des donnÃ©es dâ€™occurrence.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/vegandevs/vegan">"vegan"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: permute</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: lattice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is vegan 2.6-4</code></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">occurrence</span>, method <span class="op">=</span> <span class="st">"jaccard"</span>,</span>
<span>        diag <span class="op">=</span> <span class="cn">TRUE</span>, upper <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          1         2         3         4
1 0.0000000 0.7777778 0.7500000 0.7142857
2 0.7777778 0.0000000 0.6000000 1.0000000
3 0.7500000 0.6000000 0.0000000 0.7500000
4 0.7142857 1.0000000 0.7500000 0.0000000</code></pre>
</div>
</div>
<p>Remarquez que <code>vegdist</code> retourne une matrice dont la diagonale est de 0 (on lâ€™affiche en spÃ©cifiant <code>diag = TRUE</code>). La diagonale est lâ€™association dâ€™un objet avec lui-mÃªme. Or la similaritÃ© dâ€™un objet avec lui-mÃªme devrait Ãªtre de 1! En fait, par convention <code>vegdist</code> retourne des dissimilaritÃ©s, non pas des similaritÃ©s. La matrice de distance serait donc calculÃ©e en extrayant la racine carrÃ©e des Ã©lÃ©ments de la matrice de dissimilaritÃ©:</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dissimilarity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">occurrence</span>, method <span class="op">=</span> <span class="st">"jaccard"</span>,</span>
<span>                         diag <span class="op">=</span> <span class="cn">TRUE</span>, upper <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">distance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">dissimilarity</span><span class="op">)</span></span>
<span><span class="va">distance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          1         2         3         4
1 0.0000000 0.8819171 0.8660254 0.8451543
2 0.8819171 0.0000000 0.7745967 1.0000000
3 0.8660254 0.7745967 0.0000000 0.8660254
4 0.8451543 1.0000000 0.8660254 0.0000000</code></pre>
</div>
</div>
<p>Dans le chapitre sur lâ€™analyse compositionnelle, nous avons abordÃ© les significations diffÃ©rentes que peuvent prendre le zÃ©ro. Lâ€™information fournie par un zÃ©ro peut Ãªtre diffÃ©rente selon les circonstances. Dans le cas dâ€™une variable continue, un zÃ©ro signifie gÃ©nÃ©ralement une mesure sous le seuil de dÃ©tection. Deux tissus dont la concentration en cuivre est nulle ont une affinitÃ© sous la perspective de la concentration en cuivre. Dans le cas de mesures dâ€™abondance (dÃ©compte) ou dâ€™occurrence (prÃ©sence-absence), on pourra dÃ©crire comme similaires deux niches Ã©cologiques oÃ¹ lâ€™on retrouve une espÃ¨ce en particulier. Mais deux sites oÃ¹ lâ€™on de retrouve pas dâ€™ours polaires ne correspondent pas nÃ©cessairement Ã  des niches similaires! En effet, il peut exister de nombreuses raisons Ã©cologiques et mÃ©thodologiques pour lesquelles lâ€™espÃ¨ce ou les espÃ¨ces nâ€™ont pas Ã©tÃ© observÃ©es. Câ€™est le problÃ¨me des <strong>double-zÃ©ros</strong> (espÃ¨ces non observÃ©es Ã  deux sites), problÃ¨me qui est amplifiÃ© avec les grilles comprenant des espÃ¨ces rares.</p>
<p>La ressemblance entre des objets comprenant des donnÃ©es continues devrait Ãªtre calculÃ©e grÃ¢ce Ã  des indicateurs <em>symÃ©triques</em>. Inversement, les affinitÃ©s entre les objets dÃ©crits par des donnÃ©es dâ€™abondance ou dâ€™occurrence susceptibles de gÃ©nÃ©rer des problÃ¨mes de double-zÃ©ros devraient Ãªtre Ã©valuÃ©es grÃ¢ce Ã  des indicateurs <em>asymÃ©triques</em>. Un dÃ©fi supplÃ©mentaire arrive lorsque les donnÃ©es sont de type mixte.</p>
<p>Nous utiliserons la convention de <code>vegan</code> et nous calculerons la dissimilaritÃ©, non pas la similaritÃ©. Les mesures de dissimilaritÃ© sont calculÃ©es sur des donnÃ©es dâ€™abondance ou des donnÃ©es dâ€™occurrence. Notons quâ€™il existe beaucoup de confusion dans la littÃ©rature sur la maniÃ¨re de nommer les dissimilaritÃ©s (ce qui nâ€™est pas le cas des distances, dont les noms sont reconnus). Dans les sections suivantes, nous noterons la dissimilaritÃ© avec un <span class="math inline">\(d\)</span> minuscule et la distance avec un <span class="math inline">\(D\)</span> majuscule.</p>
<section id="association-entre-objets-mode-q" class="level3" data-number="10.2.1"><h3 data-number="10.2.1" class="anchored" data-anchor-id="association-entre-objets-mode-q">
<span class="header-section-number">10.2.1</span> Association entre objets (mode Q)</h3>
<section id="objets-abondance" class="level4" data-number="10.2.1.1"><h4 data-number="10.2.1.1" class="anchored" data-anchor-id="objets-abondance">
<span class="header-section-number">10.2.1.1</span> Objets: Abondance</h4>
<p>La <strong>dissimilaritÃ© de Bray-Curtis</strong> est asymÃ©trique. Elle est aussi appelÃ©e lâ€™indice de Steinhaus, de Czekanowski ou de SÃ¸rensen. Il est important de sâ€™assurer de bien sâ€™entendre la mÃ©thode Ã  laquelle on fait rÃ©fÃ©rence. Lâ€™Ã©quation enlÃ¨ve toute ambiguÃ¯tÃ©. La dissimilaritÃ© de Bray-Curtis entre les points A et B est calculÃ©e comme suit.</p>
<p><span class="math display">\[d_{AB} =  \frac {\sum \left| A_{i} - B_{i} \right| }{\sum \left(A_{i}+B_{i}\right)}\]</span></p>
<p>Utilisons <code>vegdist</code> pour gÃ©nÃ©rer les matrices dâ€™association. Le format â€œlisteâ€ de R est pratique pour enregistrer la collection dâ€™objets, dont les matrice dâ€™association que nous allons crÃ©er dans cette section.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'BrayCurtis'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">abundance</span>, method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'BrayCurtis'</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          1         2         3
2 0.9433962                    
3 0.9619048 0.4400000          
4 0.9591837 1.0000000 0.7647059</code></pre>
</div>
</div>
<p>La dissimilaritÃ© de Bray-Curtis est souvent utilisÃ©e dans la littÃ©rature. Toutefois, la version originale de Bray-Curtis nâ€™est pas tout Ã  fait mÃ©trique (semimÃ©trique). ConsÃ©quemment, la <strong>dissimilaritÃ© de Ruzicka</strong> (une variante de la dissimilaritÃ© de Jaccard pour les donnÃ©es dâ€™abondance) est mÃ©trique, et devrait probablement Ãªtre prÃ©fÃ©rÃ© Ã  Bray-Curtis (<a href="http://ocw.um.es/ciencias/geobotanica/otros-recursos-1/documentos/vegantutorial.pdf">Oksanen, 2006</a>).</p>
<p><span class="math display">\[d_{AB, Ruzicka} =  \frac { 2 \times d_{AB, Bray-Curtis} }{1 + d_{AB, Bray-Curtis}}\]</span></p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'Ruzicka'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">associations_abund</span><span class="op">[[</span><span class="st">'BrayCurtis'</span><span class="op">]</span><span class="op">]</span> <span class="op">*</span> <span class="fl">2</span> <span class="op">/</span> <span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">associations_abund</span><span class="op">[[</span><span class="st">'BrayCurtis'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La <strong>dissimilaritÃ© de Kulczynski</strong> (aussi Ã©crit Kulsinski) est asymÃ©trique et semimÃ©trique, tout comme celle de Bray-Curtis. Elle est calculÃ©e comme suit.</p>
<p><span class="math display">\[d_{AB} = 1-\frac{1}{2} \times \left[ \frac{\sum min(A_i, B_i)}{\sum A_i} + \frac{\sum min(A_i, B_i)}{\sum B_i} \right]\]</span></p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'Kulczynski'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">abundance</span>, method <span class="op">=</span> <span class="st">"kulczynski"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Une approche commune pour mesurer lâ€™association entre sites dÃ©crits par des donnÃ©es dâ€™abondance est la <strong>distance de Hellinger</strong>. Notez quâ€™il sâ€™agit ici dâ€™une distance, non pas dâ€™une dissimilaritÃ©. Pour lâ€™obtenir, on doit dâ€™abord diviser chaque donnÃ©e dâ€™abondance par lâ€™abondance totale pour chaque site pour obtenir les espÃ¨ces en tant que proportions, puis on extrait la racine carrÃ©e de chaque Ã©lÃ©ment. Enfin, on calcule la distance euclidienne entre les proportions de chaque site. Pour rappel, une distance euclidienne est la gÃ©nÃ©ralisation en plusieurs dimensions du thÃ©orÃ¨me de Pythagore, <span class="math inline">\(c = \sqrt{a^2 + b^2}\)</span>.</p>
<p><span class="math display">\[D_{AB} = \sqrt {\sum \left( \frac{A_i}{\sum A_i} - \frac{B_i}{\sum B_i} \right)^2}\]</span></p>
<table class="table">
<colgroup>
<col style="width: 26%">
<col style="width: 66%">
</colgroup>
<tbody><tr class="odd">
<td style="text-align: left;">ğŸ˜±&nbsp;<strong>Attention</strong>
</td>
<td style="text-align: right;">La distance dâ€™Hellinger hÃ©rite des biais liÃ©es aux donnÃ©es compositionnelles. Elle peut Ãªtre substituÃ©e par une matrice de distances dâ€™Aitchison.</td>
</tr></tbody>
</table>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'Hellinger'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">abundance</span>, method<span class="op">=</span><span class="st">"hellinger"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Toute comme la distance dâ€™Hellinger, la <strong>distance de chord</strong> est calculÃ©e par une distance euclidienne sur des donnÃ©es dâ€™abondance transformÃ©es de sorte que chaque ligne ait une longueur (norme) de 1.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'Chord'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">abundance</span>, method<span class="op">=</span><span class="st">"normalize"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La <strong>mÃ©trique du chi-carrÃ©</strong>, ou <span class="math inline">\(\chi\)</span>-carrÃ©, ou chi-square, donne davantage de poids aux espÃ¨ces rares quâ€™aux espÃ¨ces communes. Son utilisation est recommandÃ©e lorsque les espÃ¨ces rares sont de bons indicateurs de conditions Ã©cologiques particuliÃ¨res (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>, p.&nbsp;308).</p>
<p><span class="math display">\[  d_{AB} = \sqrt{\sum _j \frac{1}{\sum y_j} \left( \frac{A_j}{\sum A} - \frac{B_j}{\sum B} \right)^2 }  \]</span></p>
<p>La mÃ©trique peut Ãªtre transformÃ©e en distance en la multipliant par la racine carrÃ©e de la somme totale des espÃ¨ces dans la matrice dâ€™abondance (<span class="math inline">\(X\)</span>).</p>
<p><span class="math display">\[ D_{AB} = \sqrt{\sum X} \times d_{AB} \]</span></p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund</span><span class="op">[[</span><span class="st">'ChiSquare'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">abundance</span>, method<span class="op">=</span><span class="st">"chi.square"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Une maniÃ¨re visuellement plus intÃ©ressante de prÃ©senter une matrice dâ€™association est un graphique de type <em>heatmap</em>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_abund_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">associations_abund</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">associations_abund</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span>key<span class="op">=</span><span class="va">row</span><span class="op">)</span></span>
<span>  <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">column</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span></span>
<span>  <span class="va">associations_abund_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">associations_abund</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">associations_abund_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">associations_abund_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">associations_abund_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">row</span>, y<span class="op">=</span><span class="va">column</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">dist</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">value</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#00ccff"</span>, mid <span class="op">=</span> <span class="st">"#aad400"</span>, high <span class="op">=</span> <span class="st">"#ff0066"</span>, midpoint <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Site"</span>, y<span class="op">=</span><span class="st">"Site"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-dist-comparaison-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Peu importe le type dâ€™association utilisÃ©e, les <em>heatmaps</em> montrent les mÃªmes tendances. Les associations de dissimilaritÃ© (Bray-Curtis, Kulczynski et Ruzicka) sâ€™Ã©talent de 0 Ã  1, tandis que les distances (Chi-Square, Chord et Hellinger) partent de zÃ©ro, mais nâ€™ont pas de limite supÃ©rieure. On note les plus grandes diffÃ©rences entre les sites 2 et 4, tandis que les sites 2 et 3 sont les plus semblables pour toutes les mesures dâ€™association Ã  lâ€™exception de la dissimilaritÃ© de Kulczynski.</p>
</section><section id="objets-occurrence-prÃ©sence-absence" class="level4" data-number="10.2.1.2"><h4 data-number="10.2.1.2" class="anchored" data-anchor-id="objets-occurrence-prÃ©sence-absence">
<span class="header-section-number">10.2.1.2</span> Objets: Occurrence (prÃ©sence-absence)</h4>
<p>Des indices dâ€™association diffÃ©rents devraient Ãªtre utilisÃ©s lorsque des donnÃ©es sont compilÃ©es sous forme boolÃ©enne. En gÃ©nÃ©ral, les tableaux de donnÃ©es dâ€™occurrence seront compilÃ©s avec des 1 (prÃ©sence) et des 0 (absence).</p>
<p>La <strong>similaritÃ© de Jaccard</strong> entre le site A et le site B est la proportion de double 1 (prÃ©sences de 1 dans A et B) parmi les espÃ¨ces. La dissimilaritÃ© est la proportion complÃ©mentaire (comprenant [1, 0], [0, 1] et [0, 0]). La distance de Jaccard est la racine carrÃ©e de la dissimilaritÃ©.</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_occ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">associations_occ</span><span class="op">[[</span><span class="st">'Jaccard'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">occurrence</span>, method <span class="op">=</span> <span class="st">"jaccard"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Les <strong>distances dâ€™Hellinger, de chord et de chi-carrÃ©</strong> sont aussi appropriÃ©es pour les calculs de distances sur des tableaux dâ€™occurrence.</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_occ</span><span class="op">[[</span><span class="st">'Hellinger'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">occurrence</span>, method<span class="op">=</span><span class="st">"hellinger"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">associations_occ</span><span class="op">[[</span><span class="st">'Chord'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">occurrence</span>, method<span class="op">=</span><span class="st">"normalize"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">associations_occ</span><span class="op">[[</span><span class="st">'ChiSquare'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">occurrence</span>, method<span class="op">=</span><span class="st">"chi.square"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Graphiquement,</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_occ_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">associations_occ</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">associations_occ</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span>key<span class="op">=</span><span class="va">row</span><span class="op">)</span></span>
<span>  <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">column</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span></span>
<span>  <span class="va">associations_occ_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">associations_occ</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">associations_occ_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">associations_occ_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">associations_occ_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">row</span>, y<span class="op">=</span><span class="va">column</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">dist</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">value</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#00ccff"</span>, mid <span class="op">=</span> <span class="st">"#aad400"</span>, high <span class="op">=</span> <span class="st">"#ff0066"</span>, midpoint <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Site"</span>, y<span class="op">=</span><span class="st">"Site"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-occ-comparaison-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Il est attendu que les matrices dâ€™association sur lâ€™occurrence sont semblables Ã  celles sur lâ€™abondance. Dans ce cas-ci, la distance dâ€™Hellinger donne des rÃ©sultats semblables Ã  la dissimilaritÃ© de Jaccard.</p>
</section><section id="objets-donnÃ©es-quantitatives" class="level4" data-number="10.2.1.3"><h4 data-number="10.2.1.3" class="anchored" data-anchor-id="objets-donnÃ©es-quantitatives">
<span class="header-section-number">10.2.1.3</span> Objets: DonnÃ©es quantitatives</h4>
<p>Les donnÃ©es quantitative en Ã©cologie peuvent dÃ©crire lâ€™Ã©tat de lâ€™environnement: le climat, lâ€™hydrologie, lâ€™hydrogÃ©ochimie, la pÃ©dologie, etc. En rÃ¨gle gÃ©nÃ©rale, les coordonnÃ©es des sites ne sot pas des variables environnementales, Ã  que lâ€™on soupÃ§onne la coordonnÃ©e elle-mÃªme dâ€™Ãªtre responsable dâ€™effets sur notre systÃ¨me: mais il sâ€™agira la plupart du temps dâ€™effets confondants (par exemple, on peut mesurer un effet de la latitude sur le rendement des agrumes, mais il sâ€™agira probablement avant tout dâ€™effets dus aux conditions climatiques, qui elles changent en fonction de la latitude). Dâ€™autre types de donnÃ©es quantitative pouvant Ãªtre apprÃ©hendÃ©es par des distances sont les traits phÃ©nologiques, les ionomes, les gÃ©nomes, etc.</p>
<p>La <strong>distance euclidienne</strong> est la racine carrÃ©e de la somme des carrÃ©s des distances sur tous les axes. Il sâ€™agit dâ€™une application multidimensionnelle du thÃ©orÃ¨me de Pythagore. La <strong>distance dâ€™Aitchison</strong>, couverte dans le <a href="08-explorer.html">chapitre&nbsp;<span>9</span></a>, est une distance euclidienne calculÃ©e sur des donnÃ©es compositionnelles prÃ©alablement transformÃ©es. La distance euclidienne est sensible aux unitÃ©s utilisÃ©s: utiliser des millimÃ¨tres plutÃ´t que des mÃ¨tres enflera la distance euclidienne. Il est recommandÃ© de porter une attention particuliÃ¨re aux unitÃ©s, et de standardiser les donnÃ©es au besoin (par exemple, en centrant la moyenne Ã  zÃ©ro et en fixant lâ€™Ã©cart-type Ã  1).</p>
<p>On pourrait, par exemple, mesurer la distance entre des observations des dimensions de diffÃ©rentes espÃ¨ces dâ€™iris. Ce tableau est inclus dans R par dÃ©faut.</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span></span>
<span><span class="va">iris</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
1          5.3         3.7          1.5         0.2    setosa
2          6.4         3.1          5.5         1.8 virginica
3          4.9         3.6          1.4         0.1    setosa
4          7.9         3.8          6.4         2.0 virginica
5          4.4         2.9          1.4         0.2    setosa</code></pre>
</div>
</div>
<p>Les mesures du tableau sont en centimÃ¨tres. Pour Ã©viter de donner davantage de poids aux longueurs des sÃ©pales et en mÃªme temps de nÃ©gliger la largeur des pÃ©tales, nous allons standardiser le tableau.</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris_sc</span> <span class="op">&lt;-</span> <span class="va">iris</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Species <span class="op">=</span> <span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span> </span>
<span><span class="va">iris_sc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 150 Ã— 5
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
 1       -0.898      1.02          -1.34       -1.31 setosa 
 2       -1.14      -0.132         -1.34       -1.31 setosa 
 3       -1.38       0.327         -1.39       -1.31 setosa 
 4       -1.50       0.0979        -1.28       -1.31 setosa 
 5       -1.02       1.25          -1.34       -1.31 setosa 
 6       -0.535      1.93          -1.17       -1.05 setosa 
 7       -1.50       0.786         -1.34       -1.18 setosa 
 8       -1.02       0.786         -1.28       -1.31 setosa 
 9       -1.74      -0.361         -1.34       -1.31 setosa 
10       -1.14       0.0979        -1.28       -1.44 setosa 
# â„¹ 140 more rows</code></pre>
</div>
</div>
<p>Pour les comparaisons des dimensions, prenons la moyenne des dimensions (mises Ã  lâ€™Ã©chelle) par espÃ¨ce.</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris_means</span> <span class="op">&lt;-</span> <span class="va">iris_sc</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Species</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise_all.html">summarise_all</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span></span>
<span><span class="va">iris_means</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 Ã— 4
  Sepal.Length Sepal.Width Petal.Length Petal.Width
         &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
1       -1.01        0.850       -1.30       -1.25 
2        0.112      -0.659        0.284       0.166
3        0.899      -0.191        1.02        1.08 </code></pre>
</div>
</div>
<p>Nous pouvons utiliser la distance euclidienne, commune en gÃ©omÃ©trie, pour comparer les espÃ¨ces. La distance euclidienne est calculÃ©e comme suit.</p>
<p><span class="math display">\[ \mathcal{E} = \sqrt{\Sigma_i \left( A_i - B_i \right) ^2 } \]</span></p>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_cont</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">associations_cont</span><span class="op">[[</span><span class="st">'Euclidean'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">iris_sc</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"euclidean"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La <strong>distance de Mahalanobis</strong> est semblable Ã  la distance euclidienne, mais qui tient compte de la covariance de la matrice des objets. Cette covariance peut Ãªtre utilisÃ©e pour dÃ©crire la structure dâ€™un nuage de points. La distance de Mahalanobis se calcule comme suit.</p>
<p><span class="math display">\[\mathcal{M} = \sqrt{(A - B)^T S^{-1} (A-B)}\]</span></p>
<p>Notez quâ€™il sâ€™agit dâ€™une gÃ©nÃ©ralisation de la distance euclidienne, qui Ã©quivaut Ã  une distance de Mahalanobis dont la matrice de covariance est une matrice identitÃ©.</p>
<p>La distance de Mahalanobis permet de reprÃ©senter des distances dans un espace fortement corrÃ©lÃ©. Elle est couramment utilisÃ©e pour dÃ©tecter les valeurs aberrantes selon des critÃ¨res de distance Ã  partir du centre dâ€™un jeu de donnÃ©es multivariÃ©es.</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_cont</span><span class="op">[[</span><span class="st">'Mahalanobis'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">iris_sc</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>, <span class="st">'mahalanobis'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La <strong>distance de Manhattan</strong> porte aussi le nom de distance de cityblock ou de taxi. Câ€™est la distance que vous devrez parcourir pour vous rendre du point A au point B Ã  Manhattan, câ€™est-Ã -dire selon une sÃ©quence de tronÃ§ons perpendiculaires.</p>
<p><span class="math display">\[ D_{AB} = \sum _i \left| A_i - B_i \right| \]</span></p>
<p>La distance de Manhattan est appropriÃ©e lorsque les gradients (changements dâ€™un Ã©tat Ã  lâ€™autre ou dâ€™une rÃ©gion Ã  lâ€™autre) ne permettent pas des changements simultanÃ©s. Mieux vaut standardiser les variables pour Ã©viter quâ€™une dimension soit prÃ©pondÃ©rante.</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_cont</span><span class="op">[[</span><span class="st">'Manhattan'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">iris_sc</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>, <span class="st">'manhattan'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Avant de prÃ©senter les rÃ©sultats des espÃ¨ces dâ€™iris, voici une reprÃ©sentation des distances euclidiennes (rouge), de Mahalanobis (bleu) et de Manhattan (vert), chacune de 1 et 2 unitÃ©s Ã  partir du centre et, pour ce qui est de la distance de Mahalanobis, selon la covariance.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://r-forge.r-project.org/projects/car/">"car"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: carData</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'car'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    recode</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    some</code></pre>
</div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://www.stats.ox.ac.uk/pub/MASS4/">"MASS"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MASS'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    select</code></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">select</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span> <span class="co"># Ã©viter les conflits de fonctions entre MASS et dplyr</span></span>
<span><span class="va">filter</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span></span>
<span></span>
<span><span class="va">sigma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.6</span>, <span class="fl">0.6</span>, <span class="fl">1</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># matrice de covariance</span></span>
<span><span class="va">mu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span> <span class="co"># centre</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/mvrnorm.html">mvrnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, <span class="va">mu</span>, <span class="va">sigma</span><span class="op">)</span> <span class="co"># gÃ©nÃ©rer des donnÃ©es</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">data</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, asp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## cercles</span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span><span class="op">*</span><span class="va">pi</span>,length<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">c1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">mu</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">*</span><span class="fl">1</span>, <span class="va">mu</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">*</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">c2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">mu</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">sin</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">*</span><span class="fl">2</span>, <span class="va">mu</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Trig.html">cos</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span><span class="op">*</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">c1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="va">c2</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## ellipses</span></span>
<span><span class="va">e1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/car/man/Ellipses.html">ellipse</a></span><span class="op">(</span><span class="va">mu</span>, <span class="va">sigma</span>, radius<span class="op">=</span><span class="fl">1</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">e2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/car/man/Ellipses.html">ellipse</a></span><span class="op">(</span><span class="va">mu</span>, <span class="va">sigma</span>, radius<span class="op">=</span><span class="fl">2</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## carrÃ©s</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">2</span>, <span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-dist-iris-comparaison-1.png" class="img-fluid figure-img" width="480"></p>
</figure>
</div>
</div>
</div>
<p>Et, graphiquement, les rÃ©sultats des distances des iris.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">associations_cont_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">associations_cont</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">associations_cont</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span>key<span class="op">=</span><span class="va">row</span><span class="op">)</span></span>
<span>  <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">column</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">associations_cont_df</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">associations_cont</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">associations_cont_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">associations_cont_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">associations_cont_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">row</span>, y<span class="op">=</span><span class="va">column</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">dist</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, colour <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">#geom_text(aes(label = round(value, 2))) +</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#00ccff"</span>, mid <span class="op">=</span> <span class="st">"#aad400"</span>, high <span class="op">=</span> <span class="st">"#ff0066"</span>, midpoint <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Site"</span>, y<span class="op">=</span><span class="st">"Site"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-dist-comparaison-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Le tableau <code>iris</code> est ordonnÃ© par espÃ¨ce. Les distances euclidienne et de Manhattan permettent aisÃ©ment de distinguer les espÃ¨ces selon les dimensions des pÃ©tales et des sÃ©pales. Toutefois, lâ€™utilisation de la covariance avec la distance de Mahalanobis crÃ©e des distinction moins tranchÃ©es.</p>
</section><section id="objets-donnÃ©es-mixtes" class="level4" data-number="10.2.1.4"><h4 data-number="10.2.1.4" class="anchored" data-anchor-id="objets-donnÃ©es-mixtes">
<span class="header-section-number">10.2.1.4</span> Objets: DonnÃ©es mixtes</h4>
<p>Les donnÃ©es catÃ©gorielles ordinales peuvent Ãªtre transformÃ©es en donnÃ©es continues par gradations linÃ©aires ou quadratiques. Les donnÃ©es catÃ©gorielles nominales, quant Ã  elles, peuvent Ãªtre encodÃ©es (<em>encodage catÃ©goriel</em>) en donnÃ©es similaires Ã  des occurrences. Attention toutefois: contrairement Ã  la rÃ©gression linÃ©aire qui demande dâ€™exclure une catÃ©gorie, lâ€™encodage catÃ©goriel doit inclure toutes les catÃ©gories. Le comportement par dÃ©faut de la fonction <code>model.matrix</code> est dâ€™exclure la catÃ©gorie de rÃ©fÃ©rence: on doit spÃ©cifier que lâ€™intercept est de zÃ©ro, câ€™est-Ã -dire <code>model.matrix(~ + categorie)</code>.</p>
<p>La <strong>similaritÃ© de Gower</strong> a Ã©tÃ© dÃ©veloppÃ©e pour mesurer des associations entre des objets dont les donnÃ©es sont mixtes: boolÃ©ennes, catÃ©gorielles et continues. La similaritÃ© de Gower est calculÃ©e en additionnant les distances calculÃ©es par colonne, individuellement. Si la colonne est boolÃ©enne, on utilise les distances de Jaccard (qui exclue les double-zÃ©ro) de maniÃ¨re univariÃ©e: une variable Ã  la fois. Pour les variables continues, on utilise la distance de Manhattan divisÃ©e par la plage de valeurs de la variable (pour fin de standardisation). Puisquâ€™elle hÃ©rite de la particularitÃ© de la distance de Manhattan et de la similaritÃ© de Jaccard univariÃ©e, la <strong>similaritÃ© de Gower</strong> reste une combinaison linÃ©aire de distances univariÃ©es.</p>
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,</span>
<span>            age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">21</span>, <span class="fl">21</span>, <span class="fl">19</span>, <span class="fl">30</span>, <span class="fl">21</span>, <span class="fl">21</span>, <span class="fl">19</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>            gender <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'M'</span>,<span class="st">'M'</span>,<span class="st">'N'</span>,<span class="st">'M'</span>,<span class="st">'F'</span>,<span class="st">'F'</span>,<span class="st">'F'</span>,<span class="st">'F'</span><span class="op">)</span>,</span>
<span>            civil_status <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'MARRIED'</span>,<span class="st">'SINGLE'</span>,<span class="st">'SINGLE'</span>,<span class="st">'SINGLE'</span>,<span class="st">'MARRIED'</span>,<span class="st">'SINGLE'</span>,<span class="st">'WIDOW'</span>,<span class="st">'DIVORCED'</span><span class="op">)</span>,</span>
<span>            salary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3000.0</span>,<span class="fl">1200.0</span> ,<span class="fl">32000.0</span>,<span class="fl">1800.0</span> ,<span class="fl">2900.0</span> ,<span class="fl">1100.0</span> ,<span class="fl">10000.0</span>,<span class="fl">1500.0</span><span class="op">)</span>,</span>
<span>            children <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>            available_credit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2200</span>,<span class="fl">100</span>,<span class="fl">22000</span>,<span class="fl">1100</span>,<span class="fl">2000</span>,<span class="fl">100</span>,<span class="fl">6000</span>,<span class="fl">2200</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">X</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 Ã— 7
     ID   age gender civil_status salary children available_credit
  &lt;int&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt; &lt;lgl&gt;               &lt;dbl&gt;
1     1    21 M      MARRIED        3000 TRUE                 2200
2     2    21 M      SINGLE         1200 FALSE                 100
3     3    19 N      SINGLE        32000 TRUE                22000
4     4    30 M      SINGLE         1800 TRUE                 1100
5     5    21 F      MARRIED        2900 TRUE                 2000
6     6    21 F      SINGLE         1100 TRUE                  100
7     7    19 F      WIDOW         10000 FALSE                6000
8     8    30 F      DIVORCED       1500 TRUE                 2200</code></pre>
</div>
</div>
<p>Il faut prÃ©alablement procÃ©der Ã  lâ€™encodage catÃ©goriel pour les variables catÃ©gorielles nominales.</p>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">X_dum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">.</span>, <span class="va">X</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">X_dum</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  age genderF genderM genderN civil_statusMARRIED civil_statusSINGLE
1  21       0       1       0                   1                  0
2  21       0       1       0                   0                  1
3  19       0       0       1                   0                  1
4  30       0       1       0                   0                  1
5  21       1       0       0                   1                  0
6  21       1       0       0                   0                  1
7  19       1       0       0                   0                  0
8  30       1       0       0                   0                  0
  civil_statusWIDOW salary childrenTRUE available_credit
1                 0   3000            1             2200
2                 0   1200            0              100
3                 0  32000            1            22000
4                 0   1800            1             1100
5                 0   2900            1             2000
6                 0   1100            1              100
7                 1  10000            0             6000
8                 0   1500            1             2200
attr(,"assign")
 [1] 1 2 2 2 3 3 3 4 5 6
attr(,"contrasts")
attr(,"contrasts")$gender
[1] "contr.treatment"

attr(,"contrasts")$civil_status
[1] "contr.treatment"

attr(,"contrasts")$children
[1] "contr.treatment"</code></pre>
</div>
</div>
<p>Calculons la dissimilaritÃ© de Gower (cette fois le graphique est fait avec <code>pheatmap</code>).</p>
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"pheatmap"</span><span class="op">)</span></span>
<span><span class="va">d_gow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">X_dum</span>, <span class="st">'gower'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">d_gow</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">d_gow</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">$</span><span class="va">ID</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">d_gow</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-mixte-gower-heatmap-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Les dendrogrammes apparaissant sur les axes du graphique sont issus dâ€™un processus de partitionnement basÃ© sur la distance, que nous verrons plus loin dans ce chapitre. Les profils des clients 4 et 7, ainsi que ceux des clients 3 et 7 diffÃ¨rent le plus. Les profils 3 et 4 sont nÃ©anmoins plutÃ´t diffÃ©rents.</p>
</section></section><section id="associations-entre-variables-mode-r" class="level3" data-number="10.2.2"><h3 data-number="10.2.2" class="anchored" data-anchor-id="associations-entre-variables-mode-r">
<span class="header-section-number">10.2.2</span> Associations entre variables (mode R)</h3>
<p>Il existe de nombreuses approches pour mesurer les associations entre variables. La plus connue est la corrÃ©lation. Mais les donnÃ©es dâ€™abondance et dâ€™occurrence demandent des approches diffÃ©rentes.</p>
<section id="variables-abondance" class="level4" data-number="10.2.2.1"><h4 data-number="10.2.2.1" class="anchored" data-anchor-id="variables-abondance">
<span class="header-section-number">10.2.2.1</span> Variables: Abondance</h4>
<p>La distance du chi-carrÃ© est suggÃ©rÃ©e par <a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">abundance_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span></span>
<span><span class="va">D_chisq_R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">abundance_r</span>, method<span class="op">=</span><span class="st">"chi.square"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">D_chisq_R</span>, display_numbers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">D_chisq_R</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-r-abond-chisq-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Des coabondances sont notables pour la mÃ©sange Ã  tÃªte noire, le jaseur borÃ©al, la citelle Ã  poitrine rousse et le bruant Ã  gorge blanche (tache bleu au centre).</p>
</section><section id="variables-occurrence" class="level4" data-number="10.2.2.2"><h4 data-number="10.2.2.2" class="anchored" data-anchor-id="variables-occurrence">
<span class="header-section-number">10.2.2.2</span> Variables: occurrence</h4>
<p>La dissimilaritÃ© de Jaccard peut Ãªtre utilisÃ©e.</p>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">occurrence_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">occurrence</span><span class="op">)</span></span>
<span><span class="va">D_jacc_R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">occurrence_r</span>, method <span class="op">=</span> <span class="st">"jaccard"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">D_jacc_R</span>, display_numbers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">D_jacc_R</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-r-aocc-jaccard-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Des cooccurrences sont notables pour le jaseur borÃ©al, la citelle Ã  poitrine rousse et le bruant Ã  gorge blanche (tache bleu au centre).</p>
</section><section id="variables-quantitÃ©s" class="level4" data-number="10.2.2.3"><h4 data-number="10.2.2.3" class="anchored" data-anchor-id="variables-quantitÃ©s">
<span class="header-section-number">10.2.2.3</span> Variables: QuantitÃ©s</h4>
<p>La matrice des corrÃ©lations de Pearson peut Ãªtre utilisÃ©e pour les donnÃ©es continues. Quant aux variables ordinales, elles devraient idÃ©alement Ãªtre liÃ©es linÃ©airement ou quadratiquement. Si ce nâ€™est pas le cas, câ€™est-Ã -dire que les catÃ©gories sont ordonnÃ©es par rang seulement, vous pourrez avoir recours aux coefficients de corrÃ©lation de Spearman ou de Kendall.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris_cor</span> <span class="op">&lt;-</span> <span class="va">iris</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">iris</span><span class="op">[</span>, <span class="op">-</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>         display_numbers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">iris_cor</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-r-corr-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section></section><section id="conclusion-sur-les-associations" class="level3" data-number="10.2.3"><h3 data-number="10.2.3" class="anchored" data-anchor-id="conclusion-sur-les-associations">
<span class="header-section-number">10.2.3</span> Conclusion sur les associations</h3>
<p>Il nâ€™existe pas de rÃ¨gle claire pour dÃ©terminer quelle technique dâ€™association utiliser. Cela dÃ©pend en premier lieu de vos donnÃ©es. Vous sÃ©lectionnerez votre mÃ©thode dâ€™association selon le type de donnÃ©es que vous abordez, la question Ã  laquelle vous dÃ©sirez rÃ©pondre ainsi lâ€™expÃ©rience dans la littÃ©rature comme celle de vos collÃ¨gues scientifiques. Sâ€™il nâ€™existe pas de rÃ¨gle clair, câ€™est quâ€™il existe des dizaines de mÃ©thodes diffÃ©rentes, et la plupart dâ€™entre elles vous donneront une perspective juste et valide. Il faut nÃ©anmoins faire attention pour Ã©viter de sÃ©lectionner les mÃ©thodes qui ne sont pas appropriÃ©es.</p>
</section></section><section id="partitionnement" class="level2" data-number="10.3"><h2 data-number="10.3" class="anchored" data-anchor-id="partitionnement">
<span class="header-section-number">10.3</span> Partitionnement</h2>
<p>Les donnÃ©es suivantes ont Ã©tÃ© gÃ©nÃ©rÃ©es par <a href="https://github.com/scikit-learn-contrib/hdbscan/blob/master/notebooks/clusterable_data.npy">Leland McInnes</a> (Tutte institute of mathematics, Ottawa). ÃŠtes-vous en mesure dâ€™identifier des groupes? Combien en trouvez-vous?</p>
<div class="cell">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_mcinnes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/clusterable_data.csv"</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span>, skip <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 2309 Columns: 2
â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
Delimiter: ","
dbl (2): x, y

â„¹ Use `spec()` to retrieve the full column specification for this data.
â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df_mcinnes</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-load-mcinnes-data-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>En 2D, lâ€™oeil humain peut facilement dÃ©tecter les groupes. En 3D, câ€™est toujours possible, mais au-delÃ  de 3D, le partitionnement cognitive devient rapidement maladroite. Les algorithmes sont alors dâ€™une aide prÃ©cieuse. Mais ils transportent en pratique tout un bagage de limitations. Quel est le critÃ¨re dâ€™association entre les groupes? Combien de groupe devrions-nous crÃ©er? Comment distinguer une donnÃ©e trop bruitÃ©e pour Ãªtre classifiÃ©e?</p>
<p>Le partitionnement de donnÃ©es (<em>clustering</em> en anglais), et inversement leur regroupement, permet de crÃ©er des ensembles selon des critÃ¨res dâ€™association. On suppose donc que Le partitionnement permet de crÃ©er des groupes selon lâ€™information que lâ€™on fait Ã©merger des donnÃ©es. Il est consÃ©quemment entendu que les donnÃ©es ne sont pas catÃ©gorisÃ©es Ã  priori: <strong>il ne sâ€™agit pas de prÃ©dire la catÃ©gorie dâ€™un objet, mais bien de crÃ©er des catÃ©gories Ã  partir des objets</strong> par exemple selon leurs dimensions, leurs couleurs, leurs signature chimique, leurs comportements, leurs gÃ¨nes, etc.</p>
<p>Plusieurs mÃ©thodes sont aujourdâ€™hui offertes aux analystes pour partitionner leurs donnÃ©es. Dans le cadre de ce manuel, nous couvrirons ici deux grandes tendances dans les algorithmes.</p>
<ol type="1">
<li><p><em>MÃ©thodes hiÃ©rarchique et non hiÃ©rarchiques</em>. Dans un partitionnement hiÃ©rarchique, lâ€™ensemble des objets forme un groupe, comprenant des sous-regroupements, des sous-sous-regroupements, etc., dont les objets forment lâ€™ultime partitionnement. On pourra alors identifier comment se dÃ©cline un partitionnement. Ã€ lâ€™inverse, un partitionnement non-hiÃ©rarchique des algorithmes permettent de crÃ©er les groupes non hiÃ©rarchisÃ©s les plus diffÃ©rents que possible.</p></li>
<li><p><em>Membership exclusif ou flou</em>. Certaines techniques attribuent Ã  chaque objet une classe unique: lâ€™appartenance sera indiquÃ©e par un 1 et la non appartenance par un 0. Dâ€™autres techniques vont attribuer un membership flou oÃ¹ le degrÃ© dâ€™appartenance est une variable continue de 0 Ã  1. Parmi les mÃ©thodes floues, on retrouve les mÃ©thodes probabilistes.</p></li>
</ol>
<section id="Ã©valuation-dun-partitionnement" class="level3" data-number="10.3.1"><h3 data-number="10.3.1" class="anchored" data-anchor-id="Ã©valuation-dun-partitionnement">
<span class="header-section-number">10.3.1</span> Ã‰valuation dâ€™un partitionnement</h3>
<p>Le choix dâ€™une technique de partitionnement parmi de nombreuses disponibles, ainsi que le choix des paramÃ¨tres gouvernant chacune dâ€™entre elles, est avant tout basÃ© sur ce que lâ€™on dÃ©sire dÃ©finir comme Ã©tant un groupe, ainsi que la maniÃ¨re dâ€™interprÃ©ter les groupes. En outre, <strong>le nombre de groupe Ã  dÃ©partager est <em>toujours</em> une dÃ©cision de lâ€™analyste</strong>. NÃ©anmoins, on peut se fier <a href="http://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation">des indicateurs de performance de partitionnement</a>. Parmis ceux-ci, retenons le score <a href="https://rdrr.io/cran/cluster/man/silhouette.html">silhouette</a> ainsi que lâ€™<a href="https://www.tandfonline.com/doi/abs/10.1080/03610927408827101">indice de Calinski-Harabaz</a>.</p>
<section id="score-silhouette" class="level4" data-number="10.3.1.1"><h4 data-number="10.3.1.1" class="anchored" data-anchor-id="score-silhouette">
<span class="header-section-number">10.3.1.1</span> Score silhouette</h4>
<p>En anglais, le <em>h</em> dans silhouette se trouve aprÃ¨s le <em>l</em>: on parle donc de <em>silhouette coefficient</em> pour dÃ©signer le score de chacun des objets dans le partitionnement. Pour chaque objet, on calcule la distance moyenne qui le sÃ©pare des autres points de son groupe (<span class="math inline">\(a\)</span>) ainsi que la distance moyenne qui le sÃ©pare des points du groupe le plus rapprochÃ©.</p>
<p><span class="math display">\[s = \frac{b-a}{max \left(a, b \right)}\]</span></p>
<p>Un coefficient de -1 indique le pire classement, tandis quâ€™un coefficient de 1 indique le meilleur classement. La moyenne des coefficients silhouette est le score silhouette.</p>
</section><section id="indice-de-calinski-harabaz" class="level4" data-number="10.3.1.2"><h4 data-number="10.3.1.2" class="anchored" data-anchor-id="indice-de-calinski-harabaz">
<span class="header-section-number">10.3.1.2</span> Indice de Calinski-Harabaz</h4>
<p>Lâ€™indice de Calinski-Harabaz est proportionnel au ratio des dispersions intra-groupe et la moyenne des dispersions inter-groupes. Plus lâ€™indice est Ã©levÃ©, mieux les groupes sont dÃ©finis. La mathÃ©matique est dÃ©crite <a href="http://scikit-learn.org/stable/modules/clustering.html#calinski-harabaz-index">dans la documentation de scikit-learn</a>, un module dâ€™analyse et autoapprentissage sur Python.</p>
<p><strong>Note</strong>. Les coefficients silhouette et lâ€™indice de Calinski-Harabaz sont plus appropriÃ©s pour les formes de groupes convexes (cercles, sphÃ¨res, hypersphÃ¨res) que pour les formes irrÃ©guliÃ¨res (notamment celles obtenues par la DBSCAN, discutÃ©e ci-dessous).</p>
</section></section><section id="partitionnement-non-hiÃ©rarchique" class="level3" data-number="10.3.2"><h3 data-number="10.3.2" class="anchored" data-anchor-id="partitionnement-non-hiÃ©rarchique">
<span class="header-section-number">10.3.2</span> Partitionnement non hiÃ©rarchique</h3>
<p>Il peut arriver que vous nâ€™ayez pas besoin de comprendre la structure dâ€™agglomÃ©ration des objets (ou variables). Plusieurs techniques de partitionnement non hiÃ©rarchique sont disponibles sur R. On sâ€™intÃ©ressera en particulier aux <em>k-means</em> et au <em>dbscan</em>.</p>
<section id="kmeans" class="level4" data-number="10.3.2.1"><h4 data-number="10.3.2.1" class="anchored" data-anchor-id="kmeans">
<span class="header-section-number">10.3.2.1</span> Kmeans</h4>
<p>Lâ€™objectif des kmeans est de minimiser la distance euclidienne entre un nombre prÃ©dÃ©fini de <em>k</em> groupes exclusifs.</p>
<ol type="1">
<li>Lâ€™algorithme commence par placer une nombre <em>k</em> de centroides au hasard dans lâ€™espace dâ€™un nombre <em>p</em> de variables (vous devez fixer <em>k</em>, et <em>p</em> est le nombre de colonnes de vos donnÃ©es).</li>
<li>Ensuite, chaque objet est Ã©tiquetÃ© comme appartenant au groupe du centroÃ¯de le plus prÃ¨s.</li>
<li>La position du centroÃ¯de est dÃ©placÃ©e Ã  la moyenne de chaque groupe.</li>
<li>Recommencer Ã  partir de lâ€™Ã©tape 2 jusquâ€™Ã  ce que lâ€™assignation des objets aux groupes ne change plus.</li>
</ol>
<img src="https://media.giphy.com/media/12vVAGkaqHUqCQ/giphy.gif" class="img-fluid" data-fig-align="center"><center>
Source: <a href="https://dashee87.github.io/data%20science/general/Clustering-with-Scikit-with-GIFs/">David Sheehan</a>
</center>
<p>La technique des kmeans suppose que les groupes ont des distributions multinormales - reprÃ©sentÃ©es par des cercles en 2D, des sphÃ¨res en 3D, des hypersphÃ¨res en plus de 3D. Cette limitation est problÃ©matique lorsque les groupes se prÃ©sentent sous des formes irrÃ©guliÃ¨res, comme celles du nuage de points de Leland McInnes, prÃ©sentÃ© plus haut. De plus, la technique classique des kmeans est basÃ©e sur des distances euclidiennes: lâ€™utilisation des kmeans nâ€™est appropriÃ©e pour les donnÃ©es comprenant beaucoup de zÃ©ros, comme les donnÃ©es dâ€™abondance, qui devraient prÃ©alablement Ãªtre transformÃ©es en variables centrÃ©es et rÃ©duites (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>). La technique des <strong>mixtures gaussiennes</strong> (<a href="https://www.stat.washington.edu/mclust/"><em>gaussian mixtures</em></a>) est une gÃ©nÃ©ralisation des kmeans permettant dâ€™intÃ©grer la covariance des groupes. Les groupes ne sont plus des hyper-sphÃ¨res, mais des hyper-ellipsoÃ¯des.</p>
<section id="application" class="level5" data-number="10.3.2.1.1"><h5 data-number="10.3.2.1.1" class="anchored" data-anchor-id="application">
<span class="header-section-number">10.3.2.1.1</span> Application</h5>
<p>Nous pouvons utilisÃ© la fonction <code>kmeans</code> de R. Toutefois, puisque lâ€™on dÃ©sire ici effectuer des tests de partitionnement pour plusieurs nombres de groupes, nous utiliserons <code>cascadeKM</code>, du module vegan. Notez que de nombreux paramÃ¨tres par dÃ©faut sont utilisÃ©s dans les exÃ©cutions ci-dessous. Ces notes de cours ne forment pas un travail de recherche scientifique. Lors de travaux de recherche, lâ€™utilsation dâ€™un argument ou dâ€™un autre dans une fonction doit Ãªtre justifiÃ©: quâ€™un paramÃ¨tre soit utilisÃ© par dÃ©faut dans une fonction nâ€™est a priori pas une justification convaincante.</p>
<p>Pour les kmeans, on doit fixer le nombre de groupes. Le graphique des donnÃ©es de Leland McInnes montrent 6 groupes. Toutefois, il est rare que lâ€™on puisse visualiser des dÃ©marcations aussi tranchÃ©es que celles de lâ€™exemple, qui plus est dans des cas oÃ¹ lâ€™on doit traiter de plus de deux dimensions. Je vais donc lancer le partitionnement en boucle pour plusieurs nombres de groupes, de 3 Ã  10 et pour chaque groupe, Ã©valuer le score silhouette et de Calinski-Habaraz. Jâ€™utilise un argument random_state pour mâ€™assurer que les groupes seront les mÃªmes Ã  chaque fois que la cellule sera lancÃ©e.</p>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/vegandevs/vegan">"vegan"</a></span><span class="op">)</span></span>
<span><span class="va">mcinnes_kmeans</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cascadeKM.html">cascadeKM</a></span><span class="op">(</span><span class="va">df_mcinnes</span>, inf.gr <span class="op">=</span> <span class="fl">3</span>, sup.gr <span class="op">=</span> <span class="fl">10</span>, criterion <span class="op">=</span> <span class="st">"calinski"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">mcinnes_kmeans</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 4
 $ partition: int [1:2309, 1:8] 1 1 1 1 1 1 1 1 1 1 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:2309] "1" "2" "3" "4" ...
  .. ..$ : chr [1:8] "3 groups" "4 groups" "5 groups" "6 groups" ...
 $ results  : num [1:2, 1:8] 85.1 2164.5 61.4 2294.6 51.4 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:2] "SSE" "calinski"
  .. ..$ : chr [1:8] "3 groups" "4 groups" "5 groups" "6 groups" ...
 $ criterion: chr "calinski"
 $ size     : int [1:10, 1:8] 1243 561 505 NA NA NA NA NA NA NA ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : chr [1:10] "Group 1" "Group 2" "Group 3" "Group 4" ...
  .. ..$ : chr [1:8] "3 groups" "4 groups" "5 groups" "6 groups" ...
 - attr(*, "class")= chr "cascadeKM"</code></pre>
</div>
</div>
<p>Lâ€™objet <code>mcinnes_kmeans</code>, de type <code>cascadeKM</code>, peut Ãªtre visualisÃ© directement avec la fonction <code>plot</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">mcinnes_kmeans</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09-ordination_files/figure-html/apo-cluster-kmeans-mcinnes-plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>On obtient un maximum de Calinski Ã  4 groupes, qui correspond Ã  la deuxiÃ¨me simulation effectuÃ©e de 3 Ã  10.</p>
<p>Examinons les scores silhouette (module: cluster).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://svn.r-project.org/R-packages/trunk/cluster/">"cluster"</a></span><span class="op">)</span></span>
<span><span class="va">asw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">mcinnes_kmeans</span><span class="op">$</span><span class="va">partition</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mcinnes_kmeans_silhouette</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html">silhouette</a></span><span class="op">(</span><span class="va">mcinnes_kmeans</span><span class="op">$</span><span class="va">partition</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">df_mcinnes</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">asw</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mcinnes_kmeans_silhouette</span><span class="op">)</span><span class="op">$</span><span class="va">avg.width</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">10</span>, <span class="va">asw</span>, type <span class="op">=</span> <span class="st">'b'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-kmeans-mcinnes-silhouette-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Le score silhouette maximum est Ã  3 groupes. La forme des groupes nâ€™Ã©tant pas convexe, il fallait sâ€™attendre Ã  ce que indicateurs maximaux pour les deux indicateurs soient diffÃ©rents. Câ€™est dâ€™ailleurs souvent le cas. Cet exemple supporte que le choix du nombre de groupe Ã  dÃ©partager repose sur lâ€™analyste, non pas uniquement sur les indicateurs de performance. Choisissons 6 groupes, puisque que câ€™est visuellement ce que lâ€™on devrait chercher pour ce cas dâ€™Ã©tude.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">kmeans_group</span> <span class="op">&lt;-</span> <span class="va">mcinnes_kmeans</span><span class="op">$</span><span class="va">partition</span><span class="op">[</span>, <span class="fl">4</span><span class="op">]</span></span>
<span><span class="va">mcinnes_kmeans</span><span class="op">$</span><span class="va">partition</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  3 groups 4 groups 5 groups 6 groups 7 groups 8 groups 9 groups 10 groups
1        1        3        1        5        2        3        1         6
2        1        3        5        4        2        6        1         6
3        1        4        1        5        4        3        6        10</code></pre>
</div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_mcinnes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>kmeans_group <span class="op">=</span> <span class="va">kmeans_group</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># ajouter une colonne de regoupement</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">kmeans_group</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-kmeans-mcinnes-6groupes-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Lâ€™algorithme kmeans est loin dâ€™Ãªtre statisfaisant. Cela est attendu, puisque les kmeans recherchent des distribution gaussiennes sur des groupes vraisemblablement non-gaussiens.</p>
<p>Nous pouvons crÃ©er un graphique silhouette pour nos 6 groupes. Notez quâ€™Ã  cause dâ€™un bogue, il nâ€™est pas possible de prÃ©senter les donnÃ©es clairement lorsquâ€™elles sont nombreuses.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html">silhouette</a></span><span class="op">(</span><span class="va">mcinnes_kmeans</span><span class="op">$</span><span class="va">partition</span><span class="op">[</span>, <span class="fl">6</span><span class="op">]</span>,</span>
<span>                  dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">df_mcinnes</span><span class="op">[</span>, <span class="op">]</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html">sortSilhouette</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sil</span>, col <span class="op">=</span> <span class="st">'black'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-kmeans-mcinnes-silhouette-plot-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="dbscan" class="level4" data-number="10.3.2.2"><h4 data-number="10.3.2.2" class="anchored" data-anchor-id="dbscan">
<span class="header-section-number">10.3.2.2</span> DBSCAN</h4>
<p>La technique DBSCAN (* <strong>D</strong>ensity-<strong>B</strong>ased <strong>S</strong>patial <strong>C</strong>lustering of <strong>A</strong>pplications with <strong>N</strong>oise<em>) sousentend que les groupes sont composÃ©s de zones oÃ¹ lâ€™on retrouve plus de points (zones denses) sÃ©parÃ©es par des zones de faible densitÃ©. Pour lancer lâ€™algorithme, nous devons spÃ©cifier une mesure dâ€™association critique (distance ou dissimilaritÃ©) </em>d* ainsi quâ€™un nombre de point critique <em>k</em> dans le voisinage de cette distance.</p>
<ol type="1">
<li>Lâ€™algorithme commence par Ã©tiqueter chaque point selon lâ€™une de ces catÃ©gories:</li>
</ol>
<ul>
<li>
<em>Noyau</em>: le point a au moins <em>k</em> points dans son voisinage, câ€™est-Ã -dire Ã  une distance infÃ©rieure ou Ã©gale Ã  <em>d</em>.</li>
<li>
<em>Bordure</em>: le point a moins de <em>k</em> points dans son voisinage, mais lâ€™un de des points voisins est un <em>noyau</em>.</li>
<li>
<em>Bruit</em>: le cas Ã©chÃ©ant. Ces points sont considÃ©rÃ©s comme des outliers.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_dbscan_1.svg" class="img-fluid figure-img" width="600"></p>
</figure>
</div>
<ol start="2" type="1">
<li>Les noyaux distancÃ©s de <em>d</em> ou moins sont connectÃ©s entre eux en englobant les bordures.</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_dbscan_2.svg" class="img-fluid figure-img" width="600"></p>
</figure>
</div>
<p>Le nombre de groupes est prescrit par lâ€™algorithme DBSCAN, qui permet du coup de dÃ©tecter des donnÃ©es trop bruitÃ©es pour Ãªtre classÃ©es.</p>
<p><a href="https://doi.org/10.1145/2666310.2666417">Damiani et al.&nbsp;(2014)</a> a dÃ©veloppÃ© une approche utilisant la technique DBSCAN pour partitionner des zones dâ€™escale pour les flux de populations migratoires.</p>
<section id="application-1" class="level5" data-number="10.3.2.2.1"><h5 data-number="10.3.2.2.1" class="anchored" data-anchor-id="application-1">
<span class="header-section-number">10.3.2.2.1</span> Application</h5>
<p>La technique <strong>DBSCAN</strong> nâ€™est pas basÃ©e sur le nombre de groupe, mais sur la densitÃ© des points. Lâ€™argument <code>x</code> ne constitue pas les donnÃ©es, mais une matrice dâ€™association. Lâ€™argument minPts spÃ©cifie le nombre minimal de points qui lâ€™on doit retrouver Ã  une distance critique d* pour la formation des *noyaux et la propagation des groupes, spÃ©cifiÃ©e dans lâ€™argument eps. La distance d peut Ãªtre estimÃ©e en prenant une fraction de la moyenne, mais on aura volontiers recours Ã  sont bon jugement.</p>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mhahsler/dbscan">"dbscan"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dbscan'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    as.dendrogram</code></pre>
</div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mcinnes_dbscan</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dbscan/man/dbscan.html">dbscan</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">df_mcinnes</span><span class="op">[</span>, <span class="op">]</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span>,</span>
<span>                         eps <span class="op">=</span> <span class="fl">0.03</span>, minPts <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">dbscan_group</span> <span class="op">&lt;-</span> <span class="va">mcinnes_dbscan</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">dbscan_group</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 0 2 6 3 4 5</code></pre>
</div>
</div>
<p>Les paramÃ¨tres spÃ©cifiÃ©s donnent 5 groupes (<code>1, 2, ..., 5</code>) et des points trop bruitÃ©s pour Ãªtre classifiÃ©s (Ã©tiquetÃ©s <code>0</code>). Voyons comment les groupes ont Ã©tÃ© formÃ©s.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_mcinnes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dbscan_group <span class="op">=</span> <span class="va">dbscan_group</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># ajouter une colonne de regoupement</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dbscan_group</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-dbscan-mcinnes-5groupes-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Le partitionnement semble plus conforme Ã  ce que lâ€™on recherche. NÃ©anmoins, DBSCAN crÃ© quelques petits groupes indÃ©sirables (groupe 6, en rose) ainsi quâ€™un grand groupe (violet) qui auraient lieu dâ€™Ãªtre partitionnÃ©. Ces dÃ©faut pourraient Ãªtre rÃ©glÃ©s en jouant sur les paramÃ¨tres <code>eps</code> et <code>minPts</code>.</p>
</section></section></section><section id="partitionnement-hiÃ©rarchique" class="level3" data-number="10.3.3"><h3 data-number="10.3.3" class="anchored" data-anchor-id="partitionnement-hiÃ©rarchique">
<span class="header-section-number">10.3.3</span> Partitionnement hiÃ©rarchique</h3>
<p>Les techniques de partitionnement hiÃ©rarchique sont basÃ©es sur les matrices dâ€™association. La technique pour mesurer lâ€™association (entre objets ou variables) dÃ©terminera en grande partie le partitionnement des donnÃ©es. Les partitionnements hiÃ©rarchiques ont lâ€™avantage de pouvoir Ãªtre reprÃ©sentÃ©s sous forme de dendrogramme (ou arbre) de partition. Un tel dendrogramme prÃ©sente des sous-groupes qui se joignent en groupes jusquâ€™Ã  former un seul ensemble.</p>
<p>Le partitionnement hiÃ©rarchique est abondamment utilisÃ© en phylogÃ©nie, pour Ã©tudier les relations de parentÃ© entre organismes vivants, populations dâ€™organismes et espÃ¨ces. La phÃ©nÃ©tique, branche empirique de la phylogÃ©nÃ¨se interspÃ©cifique, fait usage du partitionnement hiÃ©rarchique Ã  partir dâ€™associations gÃ©nÃ©tiques entre unitÃ©s taxonomiques. On retrouve de nombreuses ressources acadÃ©miques en phylogÃ©nÃ©tique ainsi que des outils pour <a href="https://www.springer.com/us/book/9781461495413">R</a> et <a href="https://academic.oup.com/bioinformatics/article/26/12/1569/287181/DendroPy-a-Python-library-for-phylogenetic">Python</a>. Toutefois, la phylogÃ©nÃ©tique en particulier ne fait pas partie de la prÃ©sente itÃ©ration de ce manuel.</p>
<section id="techniques-de-partitionnement-hiÃ©rarchique" class="level4" data-number="10.3.3.1"><h4 data-number="10.3.3.1" class="anchored" data-anchor-id="techniques-de-partitionnement-hiÃ©rarchique">
<span class="header-section-number">10.3.3.1</span> Techniques de partitionnement hiÃ©rarchique</h4>
<p>Le partitionnement hiÃ©rarchique est typiquement effectuÃ© avec une des quatre mÃ©thodes suivantes, dont chacune possÃ¨de ses particularitÃ©s, mais sont toutes agglomÃ©ratives: Ã  chaque Ã©tape dâ€™agglomÃ©ration, on fusionne les deux groupes ayant le plus dâ€™affinitÃ© sur la base des deux sous-groupes les plus rapprochÃ©s.</p>
<p><strong>Single link</strong> (<code>single</code>). Les groupes sont agglomÃ©rÃ©s sur la base des deux points parmi les groupes, qui sont les plus proches.</p>
<p><strong>Complete link</strong> (<code>complete</code>). Ã€ la diffÃ©rence de la mÃ©thode <em>single</em>, on considÃ¨re comme critÃ¨re dâ€™agglomÃ©ration les Ã©lÃ©ments les plus Ã©loignÃ©s de chaque groupe.</p>
<p><strong>AgglomÃ©ration centrale</strong>. Il sâ€™agit dâ€™une famille de mÃ©thodes basÃ©es sur les diffÃ©rences entre les tendances centrales des objets ou des groupes.</p>
<ul>
<li>
<strong>Average</strong> (<code>average</code>). AppelÃ©e UPGMA (Unweighted Pair-Group Method unsing Average), les groupes sont agglomÃ©rÃ©s selon un centre calculÃ©s par la moyenne et le nombre dâ€™objet pondÃ¨re lâ€™agglomÃ©ration (le poids des groupes est retirÃ©). Cette technique est historiquement utilisÃ©e en bioinformatique pour partitionner des groupes phylogÃ©nÃ©tiques (<a href="https://www.cabdirect.org/cabdirect/abstract/19730310919">Sneath et Sokal, 1973</a>).</li>
<li>
<strong>Weighted</strong> (<code>weighted</code>). La version de average, mais non pondÃ©rÃ©e (WPGMA).</li>
<li>
<strong>Centroid</strong> (<code>centroid</code>). Tout comme average, mais le centroÃ¯de (centre gÃ©omÃ©trique) est utilisÃ© au lieu de la moyenne. Accronyme: UPGMC.</li>
<li>
<strong>Median</strong> (<code>median</code>). AppelÃ©e WPGMC. Devinez! ;)</li>
</ul>
<p><strong>Ward</strong> (<code>ward</code>). Lâ€™optimisation vise Ã  minimiser les sommes des carrÃ©s par regroupement.</p>
</section><section id="quel-outil-de-partitionnement-hiÃ©rarchique-utiliser" class="level4" data-number="10.3.3.2"><h4 data-number="10.3.3.2" class="anchored" data-anchor-id="quel-outil-de-partitionnement-hiÃ©rarchique-utiliser">
<span class="header-section-number">10.3.3.2</span> Quel outil de partitionnement hiÃ©rarchique utiliser?</h4>
<p>Alors que le choix de la matrice dâ€™association dÃ©pend des donnÃ©es et de leur contexte, la technique de partitionnement hiÃ©rarchique peut, quant Ã  elle, Ãªtre basÃ©e sur un critÃ¨re numÃ©rique. Il en existe plusieurs, mais le critÃ¨re recommandÃ© pour le choix dâ€™une technique de partitionnement hiÃ©rarchique est la <strong>corrÃ©lation cophÃ©nÃ©tique</strong>. La distance cophÃ©nÃ©tique est la distance Ã  laquelle deux objets ou deux sous-groupes deviennent membres dâ€™un mÃªme groupe. La corrÃ©lation cophÃ©nÃ©tique est la corrÃ©lation de Pearson entre le vecteur dâ€™association des objets et le vecteur de distances cophÃ©nÃ©tiques.</p>
</section><section id="application-2" class="level4" data-number="10.3.3.3"><h4 data-number="10.3.3.3" class="anchored" data-anchor-id="application-2">
<span class="header-section-number">10.3.3.3</span> Application</h4>
<p>Les techniques de partitionnement hiÃ©rarchique prÃ©sentÃ©es ci-dessus sont disponibles dans le module <code>stats</code> de R, qui est chargÃ© automatiquement lors de lâ€™ouverture de R. Nous allons classifier les dimensions des iris grÃ¢ce Ã  la distance de Manhattan.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mcinnes_hclust_distmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">df_mcinnes</span>, method <span class="op">=</span> <span class="st">"manhattan"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clustering_methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'single'</span>, <span class="st">'complete'</span>, <span class="st">'average'</span>, <span class="st">'centroid'</span>, <span class="st">'ward'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clust_l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">coph_corr_l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">clustering_methods</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">clust_l</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">mcinnes_hclust_distmat</span>, method <span class="op">=</span> <span class="va">clustering_methods</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">coph_corr_l</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">mcinnes_hclust_distmat</span>, <span class="fu"><a href="https://rdrr.io/r/stats/cophenetic.html">cophenetic</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The "ward" method has been renamed to "ward.D"; note new "ward.D2"</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span><span class="va">clustering_methods</span>, coph_corr <span class="op">=</span> <span class="va">coph_corr_l</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_reorder.html">fct_reorder</a></span><span class="op">(</span><span class="va">clustering_methods</span>, <span class="op">-</span><span class="va">coph_corr</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">coph_corr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"MÃ©thode de partitionnement"</span>, y <span class="op">=</span> <span class="st">"CorrÃ©lation cophÃ©nÃ©tique"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hierarch-mcinnes-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>La mÃ©thode <code>average</code> retourne la corrÃ©lation la plus Ã©levÃ©e. Pour plus de flexibilitÃ©, enchÃ¢ssons le nom de la mÃ©thode dans une variable. Ainsi, en chageant le nom de cette variable, le reste du code sera consÃ©quent.</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">clustering_methods</span></span>
<span><span class="va">best_method</span> <span class="op">&lt;-</span> <span class="st">"average"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Le partitionnement hiÃ©rarchique peut Ãªtre visualisÃ© par un dendrogramme.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">best_method</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hierarch-mcinnes-plot-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
</section><section id="combien-de-groupes-utiliser" class="level4" data-number="10.3.3.4"><h4 data-number="10.3.3.4" class="anchored" data-anchor-id="combien-de-groupes-utiliser">
<span class="header-section-number">10.3.3.4</span> Combien de groupes utiliser?</h4>
<p>La longueur des lignes verticales est la distance sÃ©parant les groupes enfants. Bien que la sÃ©lection du nombre de groupe soit avant tout basÃ©e sur les besoins du problÃ¨me, nous pouvons nous appuyer sur certains outils. La hauteur totale peut servir de critÃ¨re pour dÃ©finir un nombre de groupes adÃ©quat. On pourra sÃ©lectionner le nombre de groupe oÃ¹ la hauteur se stabilise en fonction du nombre de groupe. On pourra aussi utiliser le <em>graphique silhouette</em>, comprenant une collection de <em>largeurs de silhouette</em>, reprÃ©sentant le degrÃ© dâ€™appartenance Ã  son groupe. La fonction <code>sklearn.metrics.silhouette_score</code>, du module scikit-learn, sâ€™en occupe.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">asw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">num_groups</span> <span class="op">&lt;-</span> <span class="fl">3</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">num_groups</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html">silhouette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">best_method</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="va">num_groups</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="va">mcinnes_hclust_distmat</span><span class="op">)</span></span>
<span>  <span class="va">asw</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span><span class="op">$</span><span class="va">avg.width</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">num_groups</span>, <span class="va">asw</span>, type <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hierarch-mcinnes-ngroupes-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Le nombre optimal de groupes serait de 5. Coupons le dendrorgamme Ã  la hauteur correspondant Ã  5 groupes avec la fonction <code>cutree</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">k_opt</span> <span class="op">&lt;-</span> <span class="va">num_groups</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">asw</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">hclust_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">best_method</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="va">k_opt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">best_method</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/rect.hclust.html">rect.hclust</a></span><span class="op">(</span><span class="va">clust_l</span><span class="op">[[</span><span class="va">best_method</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="va">k_opt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hierarch-mcinnes-5groupes-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
<p>La classification hiÃ©rarchique, uniquement basÃ©e sur la distance, peut Ãªtre inappropriÃ©e pour dÃ©finir des formes complexes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_mcinnes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>hclust_group <span class="op">=</span> <span class="va">hclust_group</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># ajouter une colonne de regoupement</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">hclust_group</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hierarch-mcinnes-5groupes-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="partitionnement-hiÃ©rarchique-basÃ©e-sur-la-densitÃ©-des-points" class="level3" data-number="10.3.4"><h3 data-number="10.3.4" class="anchored" data-anchor-id="partitionnement-hiÃ©rarchique-basÃ©e-sur-la-densitÃ©-des-points">
<span class="header-section-number">10.3.4</span> Partitionnement hiÃ©rarchique basÃ©e sur la densitÃ© des points</h3>
<p>La tecchinque HDBSCAN, dont lâ€™algorithme est relativement rÃ©cent (<a href="https://link.springer.com/chapter/10.1007%2F978-3-642-37456-2_14">Campello et al., 2013</a>), permet une partitionnement hiÃ©rarchique sur le mÃªme principe des zones de densitÃ© de la technique DBSCAN. Le HDBSCAN a Ã©tÃ© utilisÃ©e pour partitionner les lieux dâ€™escale dâ€™oiseaux migrateurs en Chine (<a href="https://www.jstage.jst.go.jp/article/dsj/12/0/12_WDS-027/_article">Xu et al., 2013</a>).</p>
<p>Avec DBSCAN, un rayon est fixÃ© dans une mÃ©trique appropriÃ©e. Pour chaque point, on compte le nombre de point voisins, câ€™est Ã  dire le nombre de point se situant Ã  une distance (ou une dissimilaritÃ©) Ã©gale ou infÃ©rieure au rayon fixÃ©. Avec HDBSCAN, on spÃ©cifie le nombre de points devant Ãªtre recouverts et on calcule le rayon nÃ©cessaire pour les recouvrir. Ainsi, chaque point est associÃ© Ã  un rayon critique que lâ€™on nommera <span class="math inline">\(d_{noyau}\)</span>. La mÃ©trique initiale est ensuite altÃ©rÃ©e: on remplace les associations entre deux objets A et B par la valeur maximale entre cette association, le rayon critique de A et le rayon critique de B. Cette nouvelle distance est appelÃ©e la <em>distance dâ€™atteinte mutuelle</em>: elle accentue les distances pour les points se trouvant dans des zones peu denses. On applique par la suite un algorithme semblable Ã  la partition hiÃ©rarchique <em>single link</em>: En sâ€™Ã©largissant, les rayons se superposent, chaque superposition de rayon forment graduellement des groupes qui sâ€™agglomÃ¨rent ainsi de maniÃ¨re hiÃ©rarchique. Au lieu dâ€™effectuer une tranche Ã  une hauteur donnÃ©e dans un dendrogramme de partitionnement, la technique HDBSCAN se base sur un dendrogramme condensÃ© qui discarte les sous-groupes comprenant moins de <em>n</em> objets (<span class="math inline">\(n_{gr min}\)</span>). Dans nouveau dendrogramme, on recherche des groupes qui occupent bien lâ€™espace dâ€™analyse. Pour ce faire, on utilise lâ€™inverse de la distance pour crÃ©er un indicateur de <em>persistance</em> (semblable Ã  la similaritÃ©), <span class="math inline">\(\lambda\)</span>. Pour chaque groupe hiÃ©rarchique dans le dendrogramme condensÃ©, on peut calculer la persistance oÃ¹ le groupe prend naissance. De plus, pour chaque objet dâ€™un groupe, on peut aussi calculer une distance Ã  laquelle il quitte le groupe. La <em>stabilitÃ©</em> dâ€™un groupe est la somme des diffÃ©rences de persistance entre la persistance Ã  la naissance et les persistances des objets. On descend dans le dendrogramme. Si la somme des stabilitÃ© des groupes enfants est plus grande que la stabilitÃ© du groupe parent, on accepte la division. Sinon, le parent forme le groupe. La <a href="http://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html">documentation du module <code>hdbscan</code></a> pour Python offre une description intuitive et plus exhaustive des principes et algorithme de HDBSCAN.</p>
<section id="paramÃ¨tres" class="level4" data-number="10.3.4.1"><h4 data-number="10.3.4.1" class="anchored" data-anchor-id="paramÃ¨tres">
<span class="header-section-number">10.3.4.1</span> ParamÃ¨tres</h4>
<p>Outre la mÃ©trique dâ€™association dont nous avons discutÃ©, HDBSCAN demande dâ€™Ãªtre nourri avec <a href="https://www.rdocumentation.org/packages/dbscan/versions/1.1-3/topics/hdbscan">quelques paramÃ¨tres importants</a>. En particulier, le <strong>nombre minimum dâ€™objets par groupe</strong>, <span class="math inline">\(n_{gr min}\)</span> dÃ©pend de la quantitÃ© de donnÃ©es que vous avez Ã  votre disposition, ainsi que de la quantitÃ© dâ€™objets que vous jugez suffisante pour crÃ©er des groupes. Nous utiliserons lâ€™implÃ©mentation de HDBSCAN du module dbscan. Si vous dÃ©sirez davantage dâ€™options, vous prÃ©fÃ©rerez probablement lâ€™<a href="https://www.rdocumentation.org/packages/largeVis/versions/0.2.1.1/topics/hdbscan">implÃ©mentation du module largeVis</a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mcinnes_hdbscan</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dbscan/man/hdbscan.html">hdbscan</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">df_mcinnes</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span>,</span>
<span>                           minPts <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                           gen_hdbscan_tree <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           gen_simplified_tree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">hdbscan_group</span> <span class="op">&lt;-</span> <span class="va">mcinnes_hdbscan</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">hdbscan_group</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6 0 4 3 5 1 2</code></pre>
</div>
</div>
<p>Nous avons 6 groupes, numÃ©rotÃ©s de 1 Ã  6, ainsi que des Ã©tiquettes identifiant des objets dÃ©signÃ©s comme Ã©tant du bruit de fond, numÃ©rotÃ© 0. Le dendrogramme non condensÃ© peu Ãªtre produit.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">mcinnes_hdbscan</span><span class="op">$</span><span class="va">hdbscan_tree</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hbdscan-mcinnes-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Difficile dâ€™y voir clair avec autant dâ€™objets. Lâ€™objet <code>mcinnes_hdbscan</code> a un nombre minimum dâ€™objets par groupe de 20. Ce qui permet de prÃ©senter le dendrogramme de maniÃ¨re condensÃ©e.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">mcinnes_hdbscan</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hbdscan-mcinnes-plot2-1.png" class="img-fluid figure-img" width="1440"></p>
</figure>
</div>
</div>
</div>
<p>Enfin, un aperÃ§u des stratÃ©gies de partitionnement utilisÃ©s jusquâ€™ici.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clustering_group</span> <span class="op">&lt;-</span> <span class="va">df_mcinnes</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">kmeans_group</span>,</span>
<span>         <span class="va">hclust_group</span>,</span>
<span>         <span class="va">dbscan_group</span>,</span>
<span>         <span class="va">hdbscan_group</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/gather.html">gather</a></span><span class="op">(</span><span class="op">-</span><span class="va">x</span>, <span class="op">-</span><span class="va">y</span>, key <span class="op">=</span> <span class="st">"method"</span>, value <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: attributes are not identical across measure variables; they will be
dropped</code></pre>
</div>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clustering_group</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">clustering_group</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="va">clustering_group</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">method</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-cluster-hbdscan-mcinnes-plot3-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Clairement, le partitionnement avec HDBSCAN donne les meilleurs rÃ©sultats.</p>
</section></section><section id="conclusion-sur-le-partitionnement" class="level3" data-number="10.3.5"><h3 data-number="10.3.5" class="anchored" data-anchor-id="conclusion-sur-le-partitionnement">
<span class="header-section-number">10.3.5</span> Conclusion sur le partitionnement</h3>
<p>Au <a href="04-visualisation.html">chapitre&nbsp;<span>4</span></a>, nous avons vu avec le jeu de donnÃ©es â€œdatasaurusâ€ que la visualisation peut permettre de dÃ©tecter des structures en segmentant les donnÃ©es selon des groupes.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_datasaurus_mix.png" class="img-fluid figure-img" width="400"></p>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_datasaurus_facet.png" class="img-fluid figure-img"></p>
</figure>
</div>
<p>Or, si les donnÃ©es nâ€™Ã©taient pas Ã©tiquetÃ©es, leur structure serait indÃ©tectable avec les algorithmes disponibles actuellement. Le partitionnement permet dâ€™explorer des donnÃ©es, de dÃ©tecter des tendances et de dÃ©gager des groupes permettant la prise de dÃ©cision.</p>
<p>Plusieurs techniques de partitionnement ont Ã©tÃ© prÃ©sentÃ©es. Le choix de la technique sera dÃ©terminante sur la maniÃ¨re dont les groupes seront partitionnÃ©s. La dÃ©finition dâ€™un groupe variant dâ€™un cas Ã  lâ€™autre, il nâ€™existe pas de rÃ¨gle pour prescrire une mÃ©thode ou une autre. La partitionnement hiÃ©rarchique a lâ€™avantage de permettre de visualiser comment les groupes sâ€™agglomÃ¨rent. Parmi les mÃ©thodes de partitionnement hiÃ©rarchique disponibles, les mÃ©thodes basÃ©es sur la densitÃ© permettent une grande flexibilitÃ©, ainsi quâ€™une dÃ©tection dâ€™observations ne faisant partie dâ€™aucun groupe.</p>
</section></section><section id="ordination" class="level2" data-number="10.4"><h2 data-number="10.4" class="anchored" data-anchor-id="ordination">
<span class="header-section-number">10.4</span> Ordination</h2>
<p>En Ã©cologie, biologie, agronomie comme en foresterie, la plupart des tableaux de donnÃ©es comprennent de nombreuses variables: pH, nutriments, climat, espÃ¨ces ou cultivars, etc. Lâ€™ordination vise Ã  mettre de lâ€™ordre dans des donnÃ©es dont le nombre Ã©levÃ© de variables peut amener Ã  des difficultÃ©s dâ€™apprÃ©ciation et dâ€™interprÃ©tation (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>). Plus prÃ©cisÃ©ment, le terme ordination est utilisÃ© en Ã©cologie pour dÃ©signer les techniques de rÃ©duction dâ€™axe. Lâ€™analyse en composante principale est probablement la plus connue de ces techniques. Mais de nombreuses techniques dâ€™ordination ont Ã©tÃ© dÃ©veloppÃ©es au cours des derniÃ¨res annÃ©es, chacune ayant ses domaines dâ€™application.</p>
<p>Les techniques de rÃ©duction dâ€™axe permettent de dÃ©gager lâ€™information la plus importante en projetant une synthÃ¨se des relations entre les observations et entre les variables. Les techniques ne supposant aucune structure <em>a priori</em> sont dites <em>non-contraignantes</em>: elles ne comprennent pas de tests statistiques. Ã€ lâ€™inverse, les ordinations contraignantes lient des variables descriptives avec une ou plusieurs variables prÃ©dictives.</p>
<p>La rÃ©fÃ©rence en la matiÃ¨re est indiscutablement (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>). Cette section en couvrira quelques unes et vous guidera vers la technique la plus appropriÃ©e pour vos donnÃ©es.</p>
<section id="ordination-non-contraignante" class="level3" data-number="10.4.1"><h3 data-number="10.4.1" class="anchored" data-anchor-id="ordination-non-contraignante">
<span class="header-section-number">10.4.1</span> Ordination non contraignante</h3>
<p>Cette section couvrira lâ€™<strong>analyse en composantes principales</strong> (ACP), lâ€™<strong>analyse de correspondance</strong> (AC), <strong>lâ€™analyse factorielle</strong> (AF) ainsi que lâ€™<strong>analyse en coordonnÃ©es principales</strong> (ACoP).</p>
<table class="table">
<colgroup>
<col style="width: 33%">
<col style="width: 33%">
<col style="width: 33%">
</colgroup>
<thead><tr class="header">
<th>MÃ©thode</th>
<th>Distance prÃ©servÃ©e</th>
<th>Variables</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Analyse en composantes principales (ACP)</td>
<td>Distance euclidienne</td>
<td>DonnÃ©es quantitatives, relations linÃ©aires (attention aux double-zÃ©ros)</td>
</tr>
<tr class="even">
<td>Analyse de correspondance (AC)</td>
<td>Distance de <span class="math inline">\(\chi^2\)</span>
</td>
<td>DonnÃ©es non-nÃ©gatives, dimentionnellement homogÃ¨nes ou binaires, abondance ou occurrence</td>
</tr>
<tr class="odd">
<td>Positionnement multidimensionnel (PoMd)</td>
<td>Toute mesure de dissimilaritÃ©</td>
<td>DonnÃ©es quantitatives, qualitatives nominales/ordinales ou mixtes</td>
</tr>
</tbody>
</table>
<p>Source: AdaptÃ© de (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>, chapitre 9)</p>
<section id="analyse-en-composantes-principales" class="level4" data-number="10.4.1.1"><h4 data-number="10.4.1.1" class="anchored" data-anchor-id="analyse-en-composantes-principales">
<span class="header-section-number">10.4.1.1</span> Analyse en composantes principales</h4>
<p>Lâ€™objectif dâ€™une ACP est de reprÃ©senter les donnÃ©es dans un nombre rÃ©duit de dimensions reprÃ©sentant le plus possible la variation dâ€™un tableau de donnÃ©es: elle permet de projeter les donnÃ©es dans un espace oÃ¹ les variables sont combinÃ©es en axes orthogonaux dont le premier axe capte le maximum de variance. Lâ€™ACP peut par exemple Ãªtre utilisÃ©e pour analyser des corrÃ©lations entre variables ou dÃ©gager lâ€™information la plus pertinente dâ€™un tableau de donnÃ©es mÃ©tÃ©o ou de signal en un nombre plus retreint de variables.</p>
<p>Lâ€™ACP effectue une rotation des axes Ã  partir du centre (moyenne) du nuage de points effectuÃ©e de maniÃ¨re Ã  ce que le premier axe dÃ©finisse la direction oÃ¹ lâ€™on retrouve la variance maximale. Ce premier axe est une combinaison linÃ©aire des variables et forme la premiÃ¨re composante principale. Une fois cet axe dÃ©finit, on trouve de deuxiÃ¨me axe, orthogonal au premier, oÃ¹ lâ€™on retrouve la variance maximale - cet axe forme la deuxiÃ¨me composante principale, et ainsi de suite jusquâ€™Ã  ce que le nombre dâ€™axe corresponde au nombre de variables. Les projections des observations sur ces axes principaux sont appelÃ©s les <strong>scores</strong>. Les projections des variables sur les axes principaux sont les <strong>vecteurs propres</strong> (<em>eigenvectors</em>, ou <em>loadings</em>). La variance des composantes principales diminue de la premiÃ¨re Ã  la derniÃ¨re, et peut Ãªtre calculÃ©e comme une proportion de la variance totale: câ€™est le <strong>pourcentage dâ€™inertie</strong>. Par convention, on utilise les <strong>valeurs propres</strong> (<em>eigenvalues</em>) pour mesurer lâ€™importance des axes. Si la premiÃ¨re composante principale a une inertie de 50% et la deuxiÃ¨me a une inertie de 30%, la reprÃ©sentation en 2D des projection reprÃ©sentera 80% de la variance du nuage de points.</p>
<p>Lâ€™hÃ©tÃ©rogÃ©nÃ©itÃ© des Ã©chelles de mesure peut avoir une grande importance sur les rÃ©sultats dâ€™une ACP (les donnÃ©es doivent Ãªtre dimensionnellement homogÃ¨nes). En effet, la hauteur dâ€™un ceriser aura une variance plus grande que le diamÃ¨tre dâ€™une cerise exprimÃ© dans les mÃªmes unitÃ©s, et cette derniÃ¨re aura plus de variance que la teneur en cuivre dâ€™une feuille. Il est consÃ©quemment avisÃ© de mettre les donnÃ©es Ã  lâ€™Ã©chelle en centrant la moyenne Ã  zÃ©ro et lâ€™Ã©cart-type Ã  1 avant de procÃ©der Ã  une ACP.</p>
<p>Lâ€™ACP a Ã©tÃ© conÃ§ue pour projeter en un nombre moindre de dimensions des observations dont les distributions sont multinormales. Bien que lâ€™ACP soit une technique robuste, il est prÃ©fÃ©rable de transformer prÃ©alablement les variables dont la distribution est particuliÃ¨rement asymÃ©triques (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012, p.&nbsp;450</a>). Le cas Ã©chÃ©ant, les valeurs extrÃªmes pourraient faire dÃ©vier les vecteurs propres et biaiser lâ€™analyse. En particulier, les donnÃ©es ACP menÃ©es sur des donnÃ©es compositionnelles sont rÃ©putÃ©es pour gÃ©nÃ©rer des analyses biaisÃ©es (<a href="http://sp.lyellcollection.org/content/specpubgsl/264/1/1.full.pdf">Pawlowsky-Glahn and Egozcue, 2006</a>). Le test de Mardia (<a href="https://journal.r-project.org/archive/2014-2/korkmaz-goksuluk-zararsiz.pdf">Korkmaz, 2014</a>) peut Ãªtre utilisÃ© pour tester la multinormalitÃ©. Une distribution multinormale devrait gÃ©nÃ©rer des scores en forme dâ€™hypersphÃ¨re (en forme de cercle sur un biplot: voir plus loin).</p>
<section id="vecteurs-propres-et-valeurs-propres" class="level5" data-number="10.4.1.1.1"><h5 data-number="10.4.1.1.1" class="anchored" data-anchor-id="vecteurs-propres-et-valeurs-propres">
<span class="header-section-number">10.4.1.1.1</span> Vecteurs propres et valeurs propres</h5>
<p>Une matrice carrÃ©e (comme une matrice de covariance <span class="math inline">\(\Sigma\)</span>) multipliÃ©e par un vecteur propre <span class="math inline">\(e\)</span> est Ã©gale aux valeurs propres <span class="math inline">\(\lambda\)</span> multipliÃ©es par les vecteurs propres <span class="math inline">\(e\)</span>.</p>
<p><span class="math display">\[ \Sigma e = \lambda e \]</span></p>
<p>De maniÃ¨re intuitive, les vecteurs propres indiquent lâ€™orientation de la covariance, et les valeurs propres indique la longueur associÃ©e Ã  cette direction. Lâ€™ACP est basÃ©e sur le calcul des vecteurs propres et des valeurs propres de la matrice de covariance des variables. Pour dâ€™abord obtenir les valeurs propres <span class="math inline">\(\lambda\)</span>, il faut rÃ©soudre lâ€™Ã©quation</p>
<p><span class="math display">\[ det(cov(X) - \lambda I) = 0 \]</span>,</p>
<p>oÃ¹ <span class="math inline">\(det\)</span> est lâ€™opÃ©ration permettant de calculer le dÃ©terminant, <span class="math inline">\(cov\)</span> est lâ€™opÃ©ration pour calculer la covariance, <span class="math inline">\(X\)</span> est la matrice de donnÃ©es, <span class="math inline">\(\lambda\)</span> sont les valeurs propres et <span class="math inline">\(I\)</span> est une matrice dâ€™identitÃ©.</p>
<p>Pour <span class="math inline">\(p\)</span> variables dans votre tableau <span class="math inline">\(X\)</span>, vous obtiendrex <span class="math inline">\(p\)</span> valeurs propres. Ensuite, on trouve les vecteurs propres en rÃ©solvant lâ€™Ã©quation $ e = e $.</p>
<p>Bien quâ€™il soit possible dâ€™<a href="https://www.youtube.com/watch?v=2fCBE7DWgd0&amp;list=PLBv09BD7ez_5_yapAg86Od6JeeypkS4YM">effectuer cette opÃ©ration Ã  la main</a> pour des cas trÃ¨s simples, vous aurez avantage Ã  utiliser un langage de programmation.</p>
<p>Chargeons les donnÃ©es dâ€™iris, puis isolons seulement les deux dimensions des sÃ©pales lâ€™espÃ¨ce <em>setosa</em>.</p>
<div class="cell">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"iris"</span><span class="op">)</span></span>
<span><span class="va">setosa_sepal</span> <span class="op">&lt;-</span> <span class="va">iris</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Species</span> <span class="op">==</span> <span class="st">"setosa"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span><span class="st">"Sepal"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">setosa_sepal</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Sepal.Length Sepal.Width
1           5.1         3.5
2           4.9         3.0
3           4.7         3.2
4           4.6         3.1
5           5.0         3.6
6           5.4         3.9
7           4.6         3.4
8           5.0         3.4
9           4.4         2.9
10          4.9         3.1
11          5.4         3.7
12          4.8         3.4
13          4.8         3.0
14          4.3         3.0
15          5.8         4.0
16          5.7         4.4
17          5.4         3.9
18          5.1         3.5
19          5.7         3.8
20          5.1         3.8
21          5.4         3.4
22          5.1         3.7
23          4.6         3.6
24          5.1         3.3
25          4.8         3.4
26          5.0         3.0
27          5.0         3.4
28          5.2         3.5
29          5.2         3.4
30          4.7         3.2
31          4.8         3.1
32          5.4         3.4
33          5.2         4.1
34          5.5         4.2
35          4.9         3.1
36          5.0         3.2
37          5.5         3.5
38          4.9         3.6
39          4.4         3.0
40          5.1         3.4
41          5.0         3.5
42          4.5         2.3
43          4.4         3.2
44          5.0         3.5
45          5.1         3.8
46          4.8         3.0
47          5.1         3.8
48          4.6         3.2
49          5.3         3.7
50          5.0         3.3</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"MVN"</span><span class="op">)</span></span>
<span><span class="va">setosa_sepal_mvn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MVN/man/mvn.html">mvn</a></span><span class="op">(</span><span class="va">setosa_sepal</span>, mvnTest <span class="op">=</span> <span class="st">"mardia"</span><span class="op">)</span></span>
<span><span class="va">setosa_sepal_mvn</span><span class="op">$</span><span class="va">multivariateNormality</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             Test          Statistic           p value Result
1 Mardia Skewness  0.759503524380438 0.943793240544741    YES
2 Mardia Kurtosis 0.0934600553610254 0.925538081956867    YES
3             MVN               &lt;NA&gt;              &lt;NA&gt;    YES</code></pre>
</div>
</div>
<p>Pour considÃ©rer la distribution comme multinormale, la p-value de la distortion (<code>Mardia Skewness</code>) <strong>et</strong> la statistique de Kurtosis (<code>Mardia Kurtosis</code>) doit Ãªtre Ã©gale ou plus Ã©levÃ©e que 0.05 (<a href="https://www.rdocumentation.org/packages/MVN/versions/5.6/topics/mvn">Kormaz, 2019, fiche dâ€™aide de la fonction <code>mvn</code> de R</a>). Câ€™est bien le cas pour les donnÃ©es du tableau <code>setosa_sepal</code>.</p>
<p>Retirons de la matrice de covariance les valeurs et vecteurs propres avec la fonction <code>eigen</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">setosa_eigen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/eigen.html">eigen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cov</a></span><span class="op">(</span><span class="va">setosa_sepal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">setosa_eigenval</span> <span class="op">&lt;-</span> <span class="va">setosa_eigen</span><span class="op">$</span><span class="va">values</span></span>
<span><span class="va">setosa_eigenvec</span> <span class="op">&lt;-</span> <span class="va">setosa_eigen</span><span class="op">$</span><span class="va">vectors</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Le premier vecteur propre correspond Ã  la premiÃ¨re colonne, et le second Ã  la deuxiÃ¨me. Les coordonnÃ©es x et y sont les premiÃ¨res et deuxiÃ¨mes lignes. Les vecteurs propres ont une longueur unitaire (norme de 1). Ils peuvent Ãªtre mis Ã  lâ€™Ã©chelles Ã  la racine carrÃ©e des valeurs propres.</p>
<div class="cell">
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">setosa_eigenvec_sc</span> <span class="op">&lt;-</span> <span class="va">setosa_eigenvec</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">setosa_eigen</span><span class="op">$</span><span class="va">values</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Pour effectuer une translation des vecteurs propres au centre du nuage de point, nous avons besoin du centroÃ¯de.</p>
<div class="cell">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">centroid</span> <span class="op">&lt;-</span> <span class="va">setosa_sepal</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">.</span>, <span class="fl">2</span>, <span class="va">mean</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">setosa_sepal</span>, asp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># vecteurs propres brutes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"green"</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># vecteur propre 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"green"</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="co"># vecteur propre 1</span></span>
<span></span>
<span><span class="co"># vecteurs propres Ã  l'Ã©chelle</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec_sc</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec_sc</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># vecteur propre 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec_sc</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>      y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">setosa_eigenvec_sc</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># vecteur propre 1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">centroid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, y<span class="op">=</span><span class="va">centroid</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">2</span>, col  <span class="op">=</span><span class="st">"blue"</span><span class="op">)</span> <span class="co"># centroid</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-iris-sepal-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>On peut observer que, comme je lâ€™ai mentionnÃ© plus haut, <strong>les vecteurs propres indiquent lâ€™orientation de la covariance, et les valeurs propres indique la longueur associÃ©e Ã  cette direction</strong>.</p>
</section><section id="biplot" class="level5" data-number="10.4.1.1.2"><h5 data-number="10.4.1.1.2" class="anchored" data-anchor-id="biplot">
<span class="header-section-number">10.4.1.1.2</span> Biplot</h5>
<p>Imaginez un nuage de points en 3D, axes y compris. Vous tournez votre nuage de points pour trouver la perspective en 2D qui fera en sorte que vos donnÃ©es soient les plus dispersÃ©es possibles. Avec une lampe de poche, vous illuminez votre nuage de points dans lâ€™axe de cette perspective: vous venez dâ€™effectuer une analyse en composantes principales, et lâ€™ombre des points et des axes sur le mur formera votre biplot.</p>
<p>Pour crÃ©er un biplot, on juxtapose les descripteurs (variables) en tant que vecteurs propres, reprÃ©sentÃ©s par des flÃ¨ches, et les objets (observations) en tant que scores, reprÃ©sentÃ©s par des points. Les rÃ©sultats dâ€™une ordination peuvent Ãªtre prÃ©sentÃ©s selon deux types de biplots (<a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Legendre et Legendre, 2012</a>).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_biplot-meteo-sc2.svg" class="img-fluid figure-img"></p>
</figure>
</div>
<center>
Biplot de corrÃ©lation permettant de visualiser les corrÃ©lations entre des variables mÃ©tÃ©orologiques.
</center>
<p>Deux types de projection sont couramment utilisÃ©s.</p>
<p><strong>Biplot de distance</strong>. Ce type de projection permet de visualiser la position des objets entre eux et par rapport aux descripteurs et dâ€™apprÃ©cier la contribution des descripteurs pour crÃ©er les composantes principales. Pour crÃ©er un biplot de distance, on projette directement les vecteurs propres (<span class="math inline">\(U\)</span>) en guise de descripteurs. Pour ce qui est des objets, on utilise les scores de lâ€™ACP (<span class="math inline">\(F\)</span>). De cette maniÃ¨re,</p>
<ol type="1">
<li>les distances euclidiennes entre les scores sont des approximations des distances euclidiennes dans lâ€™espace multidimensionnel,</li>
<li>la projection dâ€™un objet sur un descripteur perpendiculairement Ã  ce dernier est une approximation de la position de lâ€™objet sur le descripteur et</li>
<li>la projection dâ€™un descripteur sur un axe principal est proportionnelle Ã  sa contribution pour gÃ©nÃ©rer lâ€™axe.</li>
</ol>
<p><strong>Biplot de corrÃ©lation</strong>. Cette projection permet dâ€™apprÃ©cier les corrÃ©lations entre les descripteurs. Pour ce faire, les objets et les valeurs propres doivent Ãªtre transformÃ©s. Pour gÃ©nÃ©rer les descripteurs, les vecteurs propres (<span class="math inline">\(U\)</span>) doivent Ãªtre multipliÃ©s par la matrice diagonalisÃ©e de la racine carrÃ©e des valeurs propres (<span class="math inline">\(\Lambda\)</span>), câ€™est-Ã -dire <span class="math inline">\(U \Lambda ^{\frac{1}{2}}\)</span>. En ce qui a trait aux objets, on multiplie les scores par (<span class="math inline">\(F\)</span>) par la racine carrÃ©e nÃ©gative des valeurs propres diagonalisÃ©es, câ€™est-Ã -dire <span class="math inline">\(F \Lambda ^{- \frac{1}{2}}\)</span>. De cette maniÃ¨re,</p>
<ol type="1">
<li>tout comme câ€™est le cas pour le biplot de distance, la projection dâ€™un objet sur un descripteur perpendiculairement Ã  ce dernier est une approximation de la position de lâ€™objet sur le descripteur,</li>
<li>la projection dâ€™un descripteur sur un axe principal est proportionnelle Ã  son Ã©cart-type et</li>
<li>les <strong>angles</strong> entre les descripteurs sont proportionnelles Ã  leur corrÃ©lation (et non pas leur proximitÃ©).</li>
</ol>
<p>En dâ€™autres mots, le bilot de distances devrait Ãªtre utilisÃ© pour apprÃ©cier la distance entre les objets et le biplot de corrÃ©lation devrait Ãªtre utilisÃ© pour apprÃ©cier les corrÃ©lations entre les descripteurs. Mais dans tous les cas, <strong>le type de biplot utilisÃ© doit Ãªtre indiquÃ©</strong>.</p>
<p>Le <em>triplot</em> est une forme apparentÃ©e au biplot, auquel on ajoute des variables prÃ©dictives. Le triplot est utile pour reprÃ©senter les rÃ©sultats des ordinations contraignantes comme les analyses de redondance et les analyse de correspondance canoniques.</p>
</section><section id="application-3" class="level5" data-number="10.4.1.1.3"><h5 data-number="10.4.1.1.3" class="anchored" data-anchor-id="application-3">
<span class="header-section-number">10.4.1.1.3</span> Application</h5>
<p>Bien que lâ€™ACP puisse Ãªtre effectuÃ©e grÃ¢ce Ã  des modules de base de R, nous utiliserons le module vegan. Le tableau <a href="https://rdrr.io/rforge/vegan/man/varechem.html"><code>varechem</code></a> comprend des donnÃ©es issues dâ€™analyse de sols identifiÃ©s par leur composition chimique, leur pH, leur profondeur totale et la profondeur de lâ€™humus publiÃ©es dans <a href="http://onlinelibrary.wiley.com/doi/10.2307/3236351/abstract">VÃ¤re et al.&nbsp;(1995)</a> et exportÃ©es du module <a href="https://rdrr.io/rforge/vegan/">vegan</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/vegandevs/vegan">"vegan"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"varechem"</span><span class="op">)</span></span>
<span><span class="va">varechem</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      N    P     K    Ca    Mg    S    Al   Fe    Mn   Zn  Mo Baresoil Humdepth
27 20.6 60.8 233.7 834.0 127.2 40.7  15.4  4.4 132.0 10.7 0.2     18.7      2.9
28 29.8 73.5 260.0 748.6 105.3 42.5  17.9  2.4 106.6  9.3 0.3     17.6      3.0
18 19.8 42.1 139.9 519.4  90.0 32.3  39.0 40.9  58.1  4.5 0.3     43.9      2.2
5  33.1 22.7  43.6 240.3  25.7 14.9  39.0  8.4  26.8  8.4 0.2      8.1      1.0
6  19.1 26.4  61.1 259.1  37.0 21.4 155.1 81.4  20.6  4.0 0.6      5.8      1.9
    pH
27 2.8
28 2.8
18 2.7
5  3.1
6  3.0</code></pre>
</div>
</div>
<p>Comme nous lâ€™avons vu prÃ©cedemment, les donnÃ©es de concentration sont de type <em>compositionnelles</em>. Les donnÃ©es compositionnelles du tableau <code>varechem</code> mÃ©riteraient dâ€™Ãªtre transformÃ©es (<a href="http://doi.wiley.com/10.1111/1467-9876.00275">Aitchison et Greenacre, 2002</a>). Utilisons les log-ratios centrÃ©s (<em>clr</em>).</p>
<div class="cell">
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://www.stat.boogaart.de/compositions/">"compositions"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to compositions, a package for compositional data analysis.
Find an intro with "? compositions"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'compositions'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    anova, cor, cov, dist, var</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:graphics':

    segments</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    %*%, norm, scale, scale.default</code></pre>
</div>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">varecomp</span> <span class="op">&lt;-</span> <span class="va">varechem</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Baresoil</span>, <span class="op">-</span><span class="va">Humdepth</span>, <span class="op">-</span><span class="va">pH</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Fv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">.</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fl">1e6</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vareclr</span> <span class="op">&lt;-</span> <span class="va">varecomp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/compositions/man/acomp.html">acomp</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/compositions/man/clr.html">clr</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">varechem</span> <span class="op">|&gt;</span></span>
<span>              <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="va">Baresoil</span>, <span class="va">Humdepth</span>, <span class="va">pH</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vareclr</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 Ã— 15
      N      P     K    Ca      Mg      S    Al     Fe     Mn    Zn    Mo    Fv
* &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 -1.88 -0.600 0.641  1.48 -0.684  -0.805 1.30  -0.156 -1.14  -2.52 -4.68  9.04
2 -1.69 -0.624 0.830  1.59 -0.0312 -0.729 0.189 -0.626 -0.331 -2.60 -5.49  9.53
3 -1.76 -0.553 0.565  2.11  0.575  -0.706 0.159 -1.20  -1.30  -1.95 -4.99  9.05
4 -1.78 -0.613 0.354  1.72 -0.174  -0.645 0.932 -1.02  -0.935 -2.31 -4.85  9.32
5 -1.34 -0.760 0.673  2.07  0.549  -0.762 0.181 -1.60  -0.696 -2.26 -5.39  9.34
# â„¹ 3 more variables: Baresoil &lt;dbl&gt;, Humdepth &lt;dbl&gt;, pH &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Effectuons lâ€™ACP. Pour cet exemple, nous standardiserons les donnÃ©es Ã©tant donnÃ©es que les colonnes Baresoil, Humedepth et pH ne sont pas Ã  la mÃªme Ã©chelle que les colonnes des clr.</p>
<div class="cell">
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vareclr_sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/compositions/man/scale.html">scale</a></span><span class="op">(</span><span class="va">vareclr</span><span class="op">)</span></span>
<span><span class="va">vare_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">rda</a></span><span class="op">(</span><span class="va">vareclr_sc</span><span class="op">)</span> <span class="co"># ou bien rda(vareclr, scale = TRUE, mais la mise Ã  l'Ã©chelle prÃ©alable est plus explicite)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lâ€™objet <code>vareclr_pca</code> contient lâ€™information nÃ©cessaire pour mener notre ACP.</p>
<div class="cell">
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">vare_pca</span>, scaling <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># scaling = 2 pour obtenir les infos pour les biplots de corrÃ©lation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
rda(X = vareclr_sc) 

Partitioning of variance:
              Inertia Proportion
Total              15          1
Unconstrained      15          1

Eigenvalues, and their contribution to the variance 

Importance of components:
                         PC1    PC2    PC3     PC4     PC5     PC6     PC7
Eigenvalue            7.1523 2.4763 2.1122 0.93015 0.57977 0.48786 0.36646
Proportion Explained  0.4768 0.1651 0.1408 0.06201 0.03865 0.03252 0.02443
Cumulative Proportion 0.4768 0.6419 0.7827 0.84473 0.88338 0.91590 0.94034
                          PC8     PC9    PC10     PC11     PC12     PC13
Eigenvalue            0.29432 0.19686 0.15434 0.107357 0.095635 0.042245
Proportion Explained  0.01962 0.01312 0.01029 0.007157 0.006376 0.002816
Cumulative Proportion 0.95996 0.97308 0.98337 0.990527 0.996902 0.999719
                           PC14
Eigenvalue            0.0042200
Proportion Explained  0.0002813
Cumulative Proportion 1.0000000

Scaling 2 for species and site scores
* Species are scaled proportional to eigenvalues
* Sites are unscaled: weighted dispersion equal on all dimensions
* General scaling constant of scores:  4.309777 


Species scores

             PC1     PC2     PC3      PC4        PC5       PC6
N         0.1437  0.7606 -0.6792  0.19837  0.1128526 -0.050149
P         0.8670 -0.3214 -0.2950 -0.22940  0.1437960 -0.042884
K         0.9122 -0.3857  0.2357  0.03469  0.2737020  0.075717
Ca        0.9649 -0.3362 -0.2147  0.17757 -0.2188717  0.008051
Mg        0.8263 -0.2723  0.1035  0.52135 -0.1495399 -0.342214
S         0.8825 -0.3169  0.3539 -0.21216  0.1176279 -0.191386
Al       -1.0105 -0.2442  0.2146  0.02674 -0.1005560 -0.043569
Fe       -1.0338 -0.2464  0.1492  0.13162  0.1512218  0.081571
Mn        0.9556  0.1041 -0.1256 -0.21300  0.2565831  0.275275
Zn        0.7763 -0.1031 -0.3123 -0.36493 -0.5665691  0.153089
Mo       -0.2152  0.8717  0.4065 -0.33643 -0.2134335 -0.167725
Fv        0.2360  0.5776 -0.8112  0.12736  0.1280097 -0.109737
Baresoil  0.5147  0.4210  0.4472  0.54980 -0.1438570  0.463148
Humdepth  0.7455  0.4379  0.5194  0.16493  0.0004757 -0.273056
pH       -0.5754 -0.5864 -0.5957  0.23408 -0.1517661 -0.056641


Site scores (weighted sums of species scores)

        PC1       PC2      PC3      PC4      PC5      PC6
18  0.16862  0.423777  0.46731  0.91175  1.10380  1.06421
15 -0.09705 -0.097482  0.61143 -0.29049  1.14916  0.40622
24  0.02831 -0.795737  0.74176 -0.19097 -2.43337 -0.81762
27  1.39081 -0.354376 -0.19377 -0.45160  0.46020 -0.31446
23  1.30346  0.357866  0.29887  0.76856  0.20913 -0.64145
19  0.43636  0.495037  1.21722  1.18128 -0.98242 -0.74474
22  1.07306  0.467575 -0.32245  0.03717  0.13956 -0.64972
16  0.02545  0.659714 -0.28861 -0.01424  0.47105  0.45173
28  1.42005  0.007356 -0.29000 -0.78474  0.97592 -0.80263
13 -0.50638 -0.220909  1.52981  0.26289  0.42135  0.94054
14  0.45392  0.649297  0.44573 -0.26620 -0.74522 -0.53228
20  0.18623  0.259640  0.89112  0.21096 -0.51393  2.24361
25  1.26264  0.225744 -0.96668 -0.69334  0.61990  0.43312
7  -1.48685  0.739545 -0.20926  1.09256  0.61856 -0.87999
5  -0.50622  1.108685 -2.61287 -1.00433 -1.35383  1.21964
6  -1.28653  0.898663 -0.38778 -0.47556 -0.02449 -0.29419
3  -1.72773  0.476962 -0.48878  0.71156  1.06398 -1.33473
4  -0.82844 -0.296515  1.20315 -1.49821 -0.18330  1.05231
2  -1.00247 -0.609253  0.25185 -0.85420  0.71031  0.14854
9  -0.43405 -0.338912  0.55348 -1.35776 -0.81986 -1.02468
12 -0.05083  0.122645 -0.04611 -0.56047 -0.26151 -0.98053
10  0.17891 -2.315489 -0.69084 -0.19547  0.80628  0.04291
11 -0.46443 -2.592018 -1.21615  1.56359 -0.62334  0.28748
21  0.46316  0.728185 -0.49843  1.89726 -0.80791  0.72671</code></pre>
</div>
</div>
<p>La deuxiÃ¨me ligne de <code>Importance of components</code>, <code>Proportion Explained</code>, indique la proportion de la variance totale captÃ©e successivement par les axes principaux. Le premier axe principal comporte 47.68% de la variance. Le deuxiÃ¨me axe principal ajoutant une proportion de 16,51%, une reprÃ©sentation en deux axes principaux prÃ©sentent 64.19 % de la variance.</p>
<div class="cell">
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop_expl</span> <span class="op">&lt;-</span> <span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span></span>
<span><span class="va">prop_expl</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         PC1          PC2          PC3          PC4          PC5          PC6 
0.4768180610 0.1650859388 0.1408156459 0.0620101490 0.0386511040 0.0325238535 
         PC7          PC8          PC9         PC10         PC11         PC12 
0.0244303815 0.0196215021 0.0131238464 0.0102890284 0.0071571089 0.0063756951 
        PC13         PC14 
0.0028163495 0.0002813359 </code></pre>
</div>
</div>
<p>La dÃ©cision du nombre dâ€™axes principaux Ã  retenir est arbitraire. Elle peut dÃ©pendre dâ€™un nombre maximal de paramÃ¨tre Ã  retenir pour Ã©viter de surdimensionner un modÃ¨le (<em>curse of dimensionality</em>, section 11) ou dâ€™un seuil de pourcentage de variance minimal Ã  retenir, par exemple 75%. Ou bien, vous retiendrez deux composantes principales si vous dÃ©sirez prÃ©senter un seul biplot.</p>
<p>Lâ€™approche de <em>Kaiser-Guttmann</em> (<a href="http://www.springer.com/us/book/9781441979759">Borcard et al., 2011</a>) consiste Ã  sÃ©lectionner les composantes principales dont la valeur propre est supÃ©rieure Ã  leur moyenne.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varechem-pca-kaiser-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Lâ€™approche du <em>broken stick</em> consiste Ã  couper un bÃ¢ton dâ€™une longueur de 1 en n tranches. La premiÃ¨re tranche est de longueur <span class="math inline">\(\frac{1}{n}\)</span>. La tranche suivante est dâ€™une longueur de la tranche prÃ©cÃ©dente Ã  laquelle on aditionne <span class="math inline">\(\frac{1}{longueur~restante}\)</span>. Puis on place les longueurs en ordre dÃ©croissant. On retient les composantes principales dont les valeurs propres cumulÃ©es sont plus grandes que le broken stick.</p>
<div class="cell">
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">broken_stick</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">bsm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="st">"numeric"</span>, length <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">bsm</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="va">x</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">bsm</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">bsm</span><span class="op">[</span><span class="va">i</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="fl">1</span><span class="op">-</span><span class="va">i</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">bsm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="va">bsm</span><span class="op">/</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">bsm</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Le graphique du <em>broken stick</em>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">prop_expl</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span>,</span>
<span>      y <span class="op">=</span> <span class="fu">broken_stick</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">vare_pca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varechem-pca-bs-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Les approches <em>Kaiser-Guttmann</em> et <em>broken stick</em> suggÃ¨rent que les trois premiÃ¨res composantes sont suffisantes pour dÃ©crire la dispersion des donnÃ©es.</p>
<p>Examinons les <em>loadings</em> (vecteurs propres) plus en particulier. Dans le langage du module vegan, les vecteurs propres sont les espÃ¨ces (<code>species</code>) et les scores sont les <code>sites</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vare_eigenvec</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">vare_pca</span>, scaling <span class="op">=</span> <span class="fl">2</span>, display <span class="op">=</span> <span class="st">"species"</span>, choices <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">vareclr</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vare_eigenvec</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                PC1        PC2        PC3        PC4           PC5          PC6
N         0.1437343  0.7606006 -0.6792046  0.1983670  0.1128526122 -0.050148980
P         0.8669892 -0.3213683 -0.2949864 -0.2294036  0.1437959857 -0.042883754
K         0.9122089 -0.3857245  0.2356904  0.0346904  0.2737019601  0.075717162
Ca        0.9648855 -0.3361651 -0.2147486  0.1775746 -0.2188716732  0.008050762
Mg        0.8263327 -0.2723055  0.1035276  0.5213484 -0.1495399242 -0.342213793
S         0.8824519 -0.3169039  0.3538854 -0.2121562  0.1176278503 -0.191386377
Al       -1.0105173 -0.2441785  0.2145614  0.0267422 -0.1005559874 -0.043569364
Fe       -1.0337676 -0.2463987  0.1491865  0.1316173  0.1512218115  0.081571443
Mn        0.9555632  0.1041030 -0.1256178 -0.2130047  0.2565830557  0.275275174
Zn        0.7763480 -0.1030878 -0.3122919 -0.3649341 -0.5665691228  0.153089144
Mo       -0.2152399  0.8717229  0.4064967 -0.3364279 -0.2134335302 -0.167725160
Fv        0.2360040  0.5775863 -0.8111953  0.1273582  0.1280096553 -0.109737235
Baresoil  0.5147445  0.4209983  0.4472351  0.5497950 -0.1438569673  0.463148072
Humdepth  0.7455213  0.4379436  0.5193895  0.1649306  0.0004756685 -0.273056212
pH       -0.5753858 -0.5863743 -0.5957495  0.2340826 -0.1517660977 -0.056640816
                 PC7         PC8          PC9        PC10         PC11
N        -0.09111164 -0.06122008  0.315645453  0.08090232  0.019251478
P         0.26894062  0.34111276  0.021124287  0.08756299  0.045741546
K        -0.21662612 -0.01641260  0.143099440 -0.08737113 -0.183005607
Ca        0.03630015  0.04775616 -0.073609828 -0.10601799 -0.161460554
Mg        0.04617838 -0.12098602 -0.051599273  0.18373857  0.009862571
S        -0.26825994  0.15822845  0.038378858  0.05100717  0.138785063
Al       -0.22737412  0.10598673  0.040586196 -0.14473132  0.089462074
Fe        0.10553041 -0.09254655 -0.079426433  0.09908706  0.006376211
Mn        0.20224538 -0.19347804 -0.038859808 -0.07637994  0.083300112
Zn       -0.12332232 -0.14862229  0.024026151  0.02643462  0.064973307
Mo        0.13788948  0.17165900  0.032981311  0.01419924 -0.128814989
Fv       -0.20911147  0.11289753 -0.281443886 -0.08391004  0.012456867
Baresoil -0.02103009  0.23028292  0.004554036  0.02604286  0.061147847
Humdepth  0.17061078 -0.11310394  0.027515405 -0.23068827  0.102189307
pH        0.19890884  0.12152266  0.150118818 -0.15240317  0.037691048
                 PC12        PC13         PC14
N         0.045420621 -0.05020956  0.002340519
P         0.145128883  0.03337551 -0.010109130
K         0.002260341  0.10566808  0.001169065
Ca        0.041210064 -0.14341793  0.007419161
Mg       -0.063493608  0.03782662 -0.023575986
S        -0.117144869 -0.06075094  0.025874035
Al        0.058212507 -0.01983102 -0.037901576
Fe        0.049837173  0.01169516  0.036048221
Mn       -0.133353213 -0.02679781 -0.021373612
Zn        0.051057277  0.06538348  0.010896560
Mo       -0.114803631  0.01989539 -0.001335923
Fv       -0.020157331  0.05448619  0.005707928
Baresoil -0.019696758  0.01640490  0.003823725
Humdepth  0.109293684  0.02485030  0.016559206
pH       -0.153813168  0.04523353  0.014193061
attr(,"const")
[1] 4.309777</code></pre>
</div>
</div>
<p>Lâ€™ordre dâ€™importance des vecteurs propres est Ã©tabli en ordre croissant des Ã©lÃ©ment des vecteurs propres associÃ©es. Un vecteur propre est une combinaison linÃ©aire des variables. Par exemple, le premier vecteur propre pointe surtout dans la direction du Fe (-1.497) et de lâ€™Al (-1.463). Le deuxiÃ¨me pointe surtout vers le Mo (2.145). Les vecteurs (<em>loadings</em>) dâ€™un biplot de distance prÃ©sentant les des deux premiÃ¨res composantes principales prendront les coordonnÃ©es des deux premiÃ¨res colonnes. Le vecteur Al aura la coordonnÃ©e [-1.463 ; -0.601], le vecteur de Fe sera placÃ© Ã  [-1.497 ; -0.606] et le vecteur Mo Ã  [-0.312 ; 2.145]. Il existe diffÃ©rentes fonctions dâ€™affichage des biplots. Notez que leur longueur peut Ãªtre magnifiÃ©e pour amÃ©liorer la visualisation.</p>
<p>LanÃ§ons la fonction <code>biplot</code> pour crÃ©er un biplot de distance et un autre de corrÃ©lation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/biplot.html">biplot</a></span><span class="op">(</span><span class="va">vare_pca</span>, scaling <span class="op">=</span> <span class="fl">1</span>, main <span class="op">=</span> <span class="st">"Biplot de distance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/biplot.html">biplot</a></span><span class="op">(</span><span class="va">vare_pca</span>, scaling <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">"Biplot de corrÃ©lation"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varechem-pca-biplots-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Le biplot de distance permet de dÃ©gager les variables qui expliquent davantage la variabilitÃ© dans notre tableau: les clr du Fe et de lâ€™Al forment en grande partie le premier axe principal, alors que le clr du Mo forme en grande partie le second axe. Le biplot de corrÃ©lation montre que les clr du Fe et du Al sont corrÃ©lÃ©s dans le mÃªme sens, mais das le sens contraire du clr du Mn. Lâ€™information sur la teneur en Fe et celle de lâ€™Al est en grande partie redondante. Toutefois, le clr du Mo est presque indÃ©pendant du clr du Fe, ceux-ci Ã©tant Ã  angle presque droit (~90Â°). Ces relations peuvent Ãªtre explorÃ©es directement.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">vareclr</span><span class="op">$</span><span class="va">Al</span>, <span class="va">vareclr</span><span class="op">$</span><span class="va">Fe</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">vareclr</span><span class="op">$</span><span class="va">Mo</span>, <span class="va">vareclr</span><span class="op">$</span><span class="va">Fe</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varechem-pca-al-fe-mo-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Nous avons mentionnÃ© que lâ€™ACP est une rotation. Prenons un second exemple pour bien en saisir les tenants et aboutissants. Le tableau de donnÃ©es que nous chargerons provient dâ€™un infographie dâ€™un dauphin, intitullÃ©e <em>Bottlenose Dolphin</em>, conÃ§u par lâ€™artiste <a href="https://www.blendswap.com/blends/view/83681">Tarnyloo</a>. Les points correspondent Ã  la surface dâ€™un dauphin. Jâ€™ai ajoutÃ© une colonne <code>anatomy</code>, qui indique Ã  quelle partie anatomique le point appartient.</p>
<div class="cell">
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dolphin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"data/07_dolphin.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 12969 Columns: 4
â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
Delimiter: ","
chr (1): anatomy
dbl (3): x, y, z

â„¹ Use `spec()` to retrieve the full column specification for this data.
â„¹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dolphin</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 Ã— 4
        x      y       z anatomy   
    &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;     
1  0.0762 -0.521 -0.0803 Head      
2  0.472   1.00   1.28   Caudal fin
3 -0.172  -0.361  0.0223 Head      
4  0.0461 -0.571 -0.0907 Head      
5  0.0152 -0.710 -0.125  Head      </code></pre>
</div>
</div>
<p>Voici en vue isomÃ©trique ce en quoi consiste ce nuage de points.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"scatterplot3d"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scatterplot3d/man/scatterplot3d.html">scatterplot3d</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dolphin</span><span class="op">$</span><span class="va">x</span>, y <span class="op">=</span> <span class="va">dolphin</span><span class="op">$</span><span class="va">y</span>, z <span class="op">=</span> <span class="va">dolphin</span><span class="op">$</span><span class="va">z</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex.symbols <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-pca-dauphin-3d-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Effectuons lâ€™ACP sur le dauphin.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dolph_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">rda</a></span><span class="op">(</span><span class="va">dolphin</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">z</span><span class="op">)</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/biplot.html">biplot</a></span><span class="op">(</span><span class="va">dolph_pca</span>, scaling <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-pca-dauphin-biplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>On nâ€™y voit pas grand chose, mais si lâ€™on extrait les scores et que lâ€™on raccourcit les vecteurs:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dolph_scores</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">dolph_pca</span>, display <span class="op">=</span> <span class="st">"sites"</span><span class="op">)</span></span>
<span><span class="va">dolph_loads</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">dolph_pca</span>, display <span class="op">=</span> <span class="st">"species"</span><span class="op">)</span></span>
<span><span class="va">dolph_loads</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          PC1         PC2
x -0.02990131  0.01608095
y -7.13731672 -1.43221776
z -4.56612084  2.23859843
attr(,"const")
[1] 9.089026</code></pre>
</div>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">dolph_scores</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.24</span>, asp <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">dolphin</span><span class="op">$</span><span class="va">anatomy</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">segments</span><span class="op">(</span>x0 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, y0 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>         x1 <span class="op">=</span> <span class="va">dolph_loads</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">/</span><span class="fl">50</span>,</span>
<span>         y1 <span class="op">=</span> <span class="va">dolph_loads</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="fl">50</span>,</span>
<span>         col <span class="op">=</span> <span class="st">"chocolate"</span>, lwd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-pca-dauphin-biplot2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>La meilleure reprÃ©sentation du dauphin en 2D, selon la variance, est son profil - en effet, il est plus long et haut que large.</p>
<div class="alert alert-block alert-info">
<strong>Note</strong>. Une ACP effectue seulement une rotation des points. Les distances euclidiennes entre les points sont maintenues.
</div>
<div class="alert alert-block alert-info">
<strong>Note</strong>. Lâ€™ACP a Ã©tÃ© conÃ§ue pour projetter en un nombre moindre de dimensions des observations dont les distributions sont multinormales (ce nâ€™est Ã©videmment pas le cas du dauphin).
</div>
<div class="alert alert-block alert-info">
<strong>Note</strong>. Les axes principaux dâ€™une ACP sont des variables alÃ©atoires. Elles peuvent Ãªtre assujetties Ã  des tests ststistiques, des modÃ¨les, du partitionnement de donnÃ©es, etc.
</div>
<div class="alert alert-block alert-success">
<strong>Excercice</strong>. Effectuez maintenant une ACP avec les donnÃ©es dâ€™iris.
</div>
</section></section><section id="analyse-de-correspondance-ac" class="level4" data-number="10.4.1.2"><h4 data-number="10.4.1.2" class="anchored" data-anchor-id="analyse-de-correspondance-ac">
<span class="header-section-number">10.4.1.2</span> Analyse de correspondance (AC)</h4>
<p>Lâ€™analyse de correspondance (AC) est particuliÃ¨rement appropriÃ©e pour traiter des donnÃ©es dâ€™abondance et dâ€™occurrence. Tout comme lâ€™analyse en composantes principales, les donnÃ©es apportÃ©s vers une AC doivent Ãªtre dimensionnellement homogÃ¨nes, câ€™est-Ã -dire que chaque variable doit Ãªtre de mÃªme mÃ©trique: pour des donnÃ©es dâ€™abondance, cela signifie que les dÃ©comptes rÃ©fÃ¨rent tous au mÃªme concept: individus, colonies, surfaces occupÃ©es, etc. Alors que la distance euclidienne est prÃ©servÃ©e avec lâ€™ACP, lâ€™AC prÃ©serve la distance du <span class="math inline">\(\chi^2\)</span>, qui est insensible aux double-zÃ©ros.</p>
<p>Lâ€™AC produit <span class="math inline">\(min(n,p)-1\)</span> axes principaux orthogonaux qui captent non pas le maximum de variance, mais la proportion de mesures aux carrÃ© par rapport Ã  la somme des carrÃ©s de la matrice. Le biplot obtenu peut Ãªtre prÃ©sentÃ© sous forme de biplot de site (<em>scaling 1</em>), oÃ¹ la distance du <span class="math inline">\(\chi^2\)</span> est prÃ©servÃ©e entre les sites ou biplot dâ€™espÃ¨ces (<em>scaling 2</em>), ou la distance du <span class="math inline">\(\chi^2\)</span> est prÃ©servÃ©e entre les espÃ¨ces. Lâ€™AC hÃ©rite du coup une propriÃ©tÃ© importate de la distance du <span class="math inline">\(\chi^2\)</span>, qui accorde davantage de distance entre un compte de 0 et de 1 quâ€™entre 1 et 2, et davantage entre 1 et 2 quâ€™entre 2 et 3.</p>
<p>Par exemple, sur ces trois sites, on a comptÃ© un individu A de moins que dâ€™individu B.</p>
<div class="cell">
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">abundance_0123</span> <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>Site <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Site 1"</span>, <span class="st">"Site 2"</span>, <span class="st">"Site 3"</span><span class="op">)</span>,</span>
<span>                        A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">9</span><span class="op">)</span>,</span>
<span>                        B <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">abundance_0123</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 Ã— 3
  Site       A     B
  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;
1 Site 1     0     1
2 Site 2     1     2
3 Site 3     9    10</code></pre>
</div>
</div>
<p>Pourtant, la distance du <span class="math inline">\(\chi^2\)</span> est plus Ã©levÃ©e entre le site 1 et le site 2 quâ€™entre le site 2 et le site 3.</p>
<div class="cell">
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">dist</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/decostand.html">decostand</a></span><span class="op">(</span><span class="va">abundance_0123</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Site</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"chi.square"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          1         2
2 0.6724111          
3 0.9555316 0.2831205</code></pre>
</div>
</div>
<p>La distance du <span class="math inline">\(\chi^2\)</span> donne davantage dâ€™importance aux espÃ¨ces rares, ce dont une analyse doit tenir compte. Il pourrait Ãªtre envisageable de retirer dâ€™un tableau des espÃ¨ces rare, ou bien prÃ©transformer des donnÃ©es dâ€™abondance par une transformation de chord ou de Hellinger (tel que discutÃ© au chapitre 6), puis procÃ©der Ã  une ACP sur ces donnÃ©es (<a href="https://doi.org/10.1007/s004420100716">Legendre et Gallagher, 2001</a>).</p>
<section id="application-4" class="level5" data-number="10.4.1.2.1"><h5 data-number="10.4.1.2.1" class="anchored" data-anchor-id="application-4">
<span class="header-section-number">10.4.1.2.1</span> Application</h5>
<p>Le tableau <a href="https://rdrr.io/rforge/vegan/man/varechem.html"><code>varespec</code></a> comprend des donnÃ©es de surface de couverture de 44 espÃ¨ces de plantes en lien avec les donnÃ©es environnementales du tableau <code>varechem</code>. Ces donnÃ©es ont Ã©tÃ© publiÃ©es dans <a href="http://onlinelibrary.wiley.com/doi/10.2307/3236351/abstract">VÃ¤re et al.&nbsp;(1995)</a> et exportÃ©es du module <a href="https://rdrr.io/rforge/vegan/">vegan</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"varespec"</span><span class="op">)</span></span>
<span><span class="va">varespec</span> <span class="op">|&gt;</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Callvulg Empenigr Rhodtome Vaccmyrt Vaccviti Pinusylv Descflex Betupube
27     0.00    15.13     2.42     5.92    15.97     0.00     3.70        0
6      0.30     5.75     0.00     0.00    10.50     0.10     0.00        0
7      0.00     5.30     0.00     0.00     8.20     0.00     0.05        0
5      0.00     0.13     0.00     0.00     2.75     0.03     0.00        0
2      0.05     9.30     0.00     0.00     8.50     0.03     0.00        0
   Vacculig Diphcomp Dicrsp Dicrfusc Dicrpoly Hylosple Pleuschr Polypili
27     1.12     0.00      0     3.63     0.00      6.7    58.07     0.00
6      0.00     0.00      0     0.85     0.00      0.0     0.05     0.03
7      8.10     0.28      0     0.45     0.03      0.0     0.10     0.00
5      0.00     0.00      0     0.25     0.03      0.0     0.03     0.18
2      0.00     0.00      0     0.03     0.00      0.0     0.75     0.00
   Polyjuni Polycomm Pohlnuta Ptilcili Barbhatc Cladarbu Cladrang Cladstel
27     0.00     0.13     0.02     0.08     0.08     1.42     7.63     2.55
6      0.08     0.00     0.00     0.08     0.00    39.00    37.50    11.30
7      0.25     0.00     0.03     0.00     0.00    35.00    42.50     0.28
5      0.65     0.00     0.00     0.00     0.00    18.50    59.00     0.98
2      0.03     0.00     0.00     0.03     0.00     0.48    24.50    75.00
   Cladunci Cladcocc Cladcorn Cladgrac Cladfimb Cladcris Cladchlo Cladbotr
27     0.15     0.00     0.38     0.12     0.10     0.03     0.00     0.02
6      3.45     0.18     0.20     0.25     0.25     0.23     0.03     0.00
7      0.35     0.08     0.20     0.25     0.18     0.13     0.08     0.00
5      0.28     0.23     0.23     0.23     0.10     0.05     0.00     0.00
2      0.20     0.00     0.03     0.03     0.05     0.03     0.03     0.00
   Cladamau Cladsp Cetreric Cetrisla Flavniva Nepharct Stersp Peltapht Icmaeric
27     0.00   0.02     0.00        0     0.00      0.0   0.00     0.07     0.00
6      0.00   0.03     0.35        0     0.08      0.0   0.03     0.00     0.00
7      0.00   0.00     0.05        0     0.23      0.2   0.93     0.00     0.03
5      0.03   0.00     0.18        0     0.28      0.0  10.28     0.00     0.10
2      0.00   0.00     0.00        0     0.00      0.0   0.00     0.00     0.00
   Cladcerv Claddefo Cladphyl
27     0.00     0.15        0
6      0.00     0.28        0
7      0.00     0.10        0
5      0.00     0.25        0
2      0.03     0.03        0</code></pre>
</div>
</div>
<p>Pour effectuer lâ€™AC, nous utiliserons, comme pour lâ€™ACP, le module vegan mais cette fois-ci avec la fonction <code>cca</code>. Lâ€™AC en <em>scaling 1</em> est effectuÃ©e sur le tableau des abondances avec les espÃ¨ces comme colonnes et les sites comme lignes. Les matrices dâ€™abondance transposÃ©es indique les sites oÃ¹ chaque espÃ¨ce ont Ã©tÃ© dÃ©nombrÃ©es: pour une analyse en <em>scaling 2</em>, on effectue une analyse de correspondance sur la matrice dâ€™abondance (ou dâ€™occurrence) transposÃ©e.</p>
<p>Pour chacune des AC, je filtre pour mâ€™assurer que toutes les lignes contiennent au moins une observation. Ce nâ€™est pas nÃ©cessaire dans notre cas, mais je le laisse pour lâ€™exemple.</p>
<div class="cell">
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vare_cca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">cca</a></span><span class="op">(</span><span class="va">varespec</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">vare_cca</span>, scaling <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
cca(X = varespec %&gt;% filter(rowSums(.) &gt; 0)) 

Partitioning of scaled Chi-square:
              Inertia Proportion
Total           2.083          1
Unconstrained   2.083          1

Eigenvalues, and their contribution to the scaled Chi-square 

Importance of components:
                         CA1    CA2    CA3     CA4     CA5     CA6     CA7
Eigenvalue            0.5249 0.3568 0.2344 0.19546 0.17762 0.12156 0.11549
Proportion Explained  0.2520 0.1713 0.1125 0.09383 0.08526 0.05835 0.05544
Cumulative Proportion 0.2520 0.4233 0.5358 0.62962 0.71489 0.77324 0.82868
                          CA8     CA9    CA10    CA11    CA12    CA13     CA14
Eigenvalue            0.08894 0.07318 0.05752 0.04434 0.02546 0.01710 0.014896
Proportion Explained  0.04269 0.03513 0.02761 0.02129 0.01222 0.00821 0.007151
Cumulative Proportion 0.87137 0.90650 0.93411 0.95539 0.96762 0.97583 0.982978
                          CA15     CA16     CA17     CA18     CA19      CA20
Eigenvalue            0.010160 0.007830 0.006032 0.004008 0.002865 0.0019275
Proportion Explained  0.004877 0.003759 0.002896 0.001924 0.001375 0.0009253
Cumulative Proportion 0.987855 0.991614 0.994510 0.996434 0.997809 0.9987341
                           CA21      CA22      CA23
Eigenvalue            0.0018074 0.0005864 0.0002434
Proportion Explained  0.0008676 0.0002815 0.0001168
Cumulative Proportion 0.9996017 0.9998832 1.0000000

Scaling 1 for species and site scores
* Sites are scaled proportional to eigenvalues
* Species are unscaled: weighted dispersion equal on all dimensions


Species scores

                CA1       CA2      CA3       CA4        CA5       CA6
Callvulg  0.0303167 -1.597460  0.11455 -2.894569  0.1376073  2.291129
Empenigr  0.0751030  0.379305  0.39303  0.023675  0.8568729 -0.400964
Rhodtome  1.1052309  1.499299  3.04284  0.120106  3.2324306 -0.283510
Vaccmyrt  1.4614812  1.622935  2.72375  0.231688  0.4604556  0.712538
Vaccviti  0.1468014  0.313436  0.14696  0.243505  0.6868371 -0.147815
Pinusylv -0.4820096  0.588517 -0.36020 -0.127094  0.4064754  0.386604
Descflex  1.5348239  1.218806  1.87562 -0.001340 -1.3136979 -0.070731
Betupube  0.6694503  1.951826  3.84017  1.389423  7.5959115 -0.244478
Vacculig -0.0830789 -1.629259  1.05063  0.802648 -0.3058811 -1.625341
Diphcomp -0.5446464 -1.037570  0.52282  0.940275  0.3682126 -1.082929
Dicrsp    1.8120408  0.360290 -4.92082  3.088562  1.3867372  0.157815
Dicrfusc  1.2704743 -0.562978 -0.39718 -2.929542  0.3848272 -2.408710
Dicrpoly  0.7248118  1.409347  0.80341  1.915549  4.5674148  1.295447
Hylosple  2.0062408  1.743883  2.27549  0.928884 -3.7648428  2.254851
Pleuschr  1.3102086  0.583036 -0.01004  0.137298 -1.1216144  0.200422
Polypili -0.3805097 -1.243904  0.54593  1.477188 -0.7276341 -0.387641
Polyjuni  1.0133795  0.099043 -2.24697  1.510641  0.7729714 -3.062378
Polycomm  0.8468241  1.321773  1.13585  1.140723  2.6836594 -0.605038
Pohlnuta -0.0136453  0.589290 -0.35542  0.135481  0.9369707  0.397246
Ptilcili  0.4223631  1.598584  3.43474  1.400065  6.3209491  0.198935
Barbhatc  0.5018348  2.119334  4.57303  1.693188  8.1101807  0.645995
Cladarbu -0.1531729 -1.483884  0.20024  0.193680  0.0734141  0.358926
Cladrang -0.5502561 -1.084008  0.40552  0.724060 -0.3357992 -0.335924
Cladstel -1.4373146  1.077753 -0.44397 -0.375926 -0.2421525  0.004212
Cladunci  0.8151727 -1.006186 -1.82587 -1.389523  1.6046713  3.675908
Cladcocc -0.2133215 -0.584429 -0.21434 -0.567886 -0.0003788 -0.145303
Cladcorn  0.2631227 -0.177858 -0.44464  0.272422  0.3992282 -0.306738
Cladgrac  0.1956947 -0.311167 -0.23894  0.379013  0.4933026  0.037581
Cladfimb  0.0009213 -0.161418  0.18463 -0.435908  0.4831233 -0.143751
Cladcris  0.3373031 -0.470369 -0.05093 -0.823855  0.7182250  0.636140
Cladchlo -0.6200021  1.207278  0.21889  0.426447  1.9506082  0.120722
Cladbotr  0.5647242  1.047333  2.65330  0.907734  4.4946805  1.201655
Cladamau -0.6598144 -1.512880  0.83251  1.577699 -0.0407227 -1.419139
Cladsp   -0.8209003  0.476164 -0.49752 -0.998241 -0.2393208  0.390785
Cetreric  0.2458192 -0.689228 -1.68427 -0.131681  0.7439412  2.374535
Cetrisla -0.3465221  1.362693  0.85897  0.396752  2.7526968  0.396591
Flavniva -1.4391907 -0.833589 -0.12919  0.007071 -1.4841375  2.956977
Nepharct  1.6813309  0.199484 -4.33509  2.229917  0.9561223 -5.472858
Stersp   -0.5172793 -2.280900  0.99775  2.377013 -0.8892757 -1.441228
Peltapht  0.4035858 -0.043265  0.04538  0.711040  0.1824679 -0.841227
Icmaeric  0.0378754 -2.419595  0.72135  0.361302 -0.3736424 -2.092136
Cladcerv -0.9232858 -0.005233 -1.22058  0.305290 -0.8142627  0.414135
Claddefo  0.5190399 -0.496632 -0.15271 -0.695927  0.9042143  0.909191
Cladphyl -1.2836161  1.155872 -0.79912 -0.741170 -0.1608002  0.490526


Site scores (weighted averages of species scores)

         CA1      CA2       CA3      CA4        CA5      CA6
18 -0.108122 -0.53705  0.229574  0.24412  0.1405624 -0.14253
15  0.697118 -0.14441 -0.031788 -0.21743 -0.2738522 -0.08146
24  0.987603  0.15042 -1.348447  0.80472  0.3095168  0.46773
27  0.851765  0.49901  0.443559  0.12277 -0.4814871  0.07589
23  0.359881 -0.05608  0.145813  0.15087  0.2405263 -0.17770
19  0.003545  0.37017  0.027760  0.06168 -0.1158930 -0.03413
22  0.860732 -0.11504  0.110869 -1.02169  0.0772348 -0.60530
16  0.636936 -0.33250  0.001120 -0.79797  0.0130769 -0.54049
28  1.279352  0.81557  0.670053  0.23137 -0.8929976  0.41783
13 -0.195009 -0.80564  0.117686 -0.58286 -0.0007212  0.53071
14  0.528532 -0.70420 -0.517771 -0.86836  0.5713441  0.91671
20  0.382866 -0.18686 -0.004789  0.10156  0.0458125  0.21087
25  0.990715  0.11967 -1.110040  0.44929  0.1885902 -0.70694
7  -0.264704 -1.06013  0.334900  0.45973 -0.0326631 -0.19945
5  -0.428410 -1.20765  0.374344  0.74970 -0.2596294 -0.30467
6  -0.330534 -0.77498  0.130760  0.22391  0.0632686  0.09060
3  -0.899601  0.12075 -0.075742  0.03842 -0.1489585 -0.12031
4  -0.770294 -0.35351 -0.033779 -0.01795 -0.3007839  0.44303
2  -0.992193  0.50319 -0.157505 -0.07070 -0.1065172 -0.09928
9  -0.937173  0.78688 -0.258119 -0.19377 -0.0343535 -0.01259
12 -0.726413  0.49163 -0.157235 -0.08698 -0.0105774 -0.02801
10 -1.002083  0.71239 -0.236526 -0.18643 -0.0231666 -0.04928
11 -0.322647 -0.03871 -0.001297  0.09029 -0.1481448  0.06934
21  0.259527  0.80746  1.124258  0.36083  1.5437866  0.07051</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">varespec_eigenval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/eigenvals.html">eigenvals</a></span><span class="op">(</span><span class="va">vare_cca</span>, scaling <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">prop_expl</span> <span class="op">&lt;-</span> <span class="va">varespec_eigenval</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">vare_cca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">prop_expl</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>      y <span class="op">=</span> <span class="fu">broken_stick</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varespec-cca-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>CrÃ©ons les biplots.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">vare_cca</span>, scaling <span class="op">=</span> <span class="fl">1</span>, main <span class="op">=</span> <span class="st">"Biplot des espÃ¨ces"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">vare_cca</span>, scaling <span class="op">=</span> <span class="fl">2</span>, main <span class="op">=</span> <span class="st">"Biplot des sites"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varespec-cca-biplot-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>Le <strong>biplot des espÃ¨ces</strong>, Ã  gauche (<code>scaling = 1</code>), montre la distribution des sites selon les espÃ¨ces. Les emplacements des scores (en noir) montrent les contrastes entre sites selon les espÃ¨ces qui les recouvrent. Les sites 14 et 15, par exemple, contrastent les sites 19, 20, 21 et 22 selon le 2iÃ¨me axe principal. Par ailleurs, les axes principaux sont formÃ© de plusieurs espÃ¨ces dont aucune ne domine clairement.</p>
<p>Le <strong>biplot des sites</strong>, Ã  droite (<code>scaling = 2</code>), montre la distribution des recouvrements dâ€™espÃ¨ces selon les sites. Par exemple, les espÃ¨ces Betupube (<a href="https://fr.wikipedia.org/wiki/Betula_pubescens"><em>Betula pubescens</em></a>) et Barbhatc (<a href="https://en.wikipedia.org/wiki/Marchantiophyta"><em>Barbilophozia hatcheri </em></a>) se recouvrent en particulier le site 24. Le site 1 est difficile Ã  identifier, car il est couvert par plusieurs noms dâ€™espÃ¨ces, au bas au centre. Les sites 3 et 13 se confondent avec Dicrsp (une espÃ¨ce de <a href="https://en.wikipedia.org/wiki/Dicranum"><em>Dicranum</em></a>) qui le recouvre amplement.</p>
<p>Pour les deux types de biplot, les sites oÃ¹ les espÃ¨ces situÃ©s prÃ¨s de lâ€™origine, car ils peuvent Ãªtre soit prÃ¨s de la moyenne, soit distribuÃ©s uniformÃ©ment.</p>
<p>Le nombre de composantes Ã  retenir peut Ãªtre Ã©valuÃ© par les approches <em>Kaiser-Guttmann</em> et <em>broken-stick</em>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scaling</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">varespec_eigenval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/eigenvals.html">eigenvals</a></span><span class="op">(</span><span class="va">vare_cca</span>, scaling <span class="op">=</span> <span class="va">scaling</span><span class="op">)</span> <span class="co"># peut Ãªtre effectuÃ© sur les deux types de scaling</span></span>
<span></span>
<span><span class="va">prop_expl</span> <span class="op">&lt;-</span> <span class="va">varespec_eigenval</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">vare_cca</span><span class="op">$</span><span class="va">CA</span><span class="op">$</span><span class="va">eig</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span>,</span>
<span>     main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Eigenvalue - Kaiser-Guttmann, scaling ="</span>, <span class="va">scaling</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>     y <span class="op">=</span> <span class="va">prop_expl</span>,</span>
<span>     type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Rang de la valeur propre"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Valeur propre"</span>,</span>
<span>     main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Proportion - broken stick, scaling ="</span>, <span class="va">scaling</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span>,</span>
<span>      y <span class="op">=</span> <span class="fu">broken_stick</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">varespec_eigenval</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-varespec-cca-analysis-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>Pour les deux scalings, lâ€™approche <em>Kaiser-Guttmann</em> propose 7 axes, tandis que lâ€™approche <em>broken-stick</em> en propose 5.</p>
<p>Les reprÃ©sentations biplot dâ€™analyse de correspondance peuvent prendre la forme dâ€™un boomerang, en particulier celles qui sont basÃ©es sur des donnÃ©es dâ€™occurrence. Le tableau suivant initialement de <a href="http://pbil.univ-lyon1.fr/R/pdf/pps047.pdf">Chessel et al.&nbsp;(1987)</a> et est distribuÃ© dans le module ade4.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://pbil.univ-lyon1.fr/ADE-4/">"ade4"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"doubs"</span><span class="op">)</span></span>
<span><span class="va">fish</span> <span class="op">&lt;-</span> <span class="va">doubs</span><span class="op">$</span><span class="va">fish</span></span>
<span><span class="va">doubs_cca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">cca</a></span><span class="op">(</span><span class="va">fish</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">doubs_cca</span>, scaling <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-doubs-cca-biplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Les numÃ©ros de sites correspondent Ã  la position dans une riviÃ¨re, 1 Ã©tant en amont et 30 en aval. Le premier axe discrimine lâ€™amont et lâ€™aval, tandis que le deuxiÃ¨me montre deux niches en amont. Bien que lâ€™on observe une discontinuitÃ© dans le cours dâ€™eau, il y a une continuitÃ© dans les abondances. Cet effet peut Ãªtre corrigÃ© en retirant la tendance de lâ€™analyse de correspondance par une <em>detrended correspondance analysis</em>. Pour cela, il faudra utiliser la fonction <code>decorana</code>, ce qui ne sera pas couvert ici.</p>
<p>Lâ€™<em>analyse des correspondances multiples</em> (ACM) est utile pour lâ€™ordination des donnÃ©es catÃ©gorielles. Le module <a href="https://larmarange.github.io/analyse-R/analyse-des-correspondances-multiples.html#acm-avec-ade4"><em>ade4</em></a> est en mesure dâ€™effectuer des AMC, mais nâ€™est pas couvert dans ce manuel.</p>
<div class="alert alert-block alert-success">
<strong>Excercice</strong>. Effectuez et analysez une AC avec les donnÃ©es de recouvrement <code>varespec</code>.
</div>
</section></section><section id="positionnement-multidimensionnel-pomd" class="level4" data-number="10.4.1.3"><h4 data-number="10.4.1.3" class="anchored" data-anchor-id="positionnement-multidimensionnel-pomd">
<span class="header-section-number">10.4.1.3</span> Positionnement multidimensionnel (PoMd)</h4>
<p>Le positionnement multidimensionnel (PoMd), ou <a href="http://scikit-learn.org/stable/modules/manifold.html"><em>manifold analysis</em></a>, se base sur les associations entre les objets (mode Q) ou les variables (mode R) pour en rÃ©duire les dimensions. Alors que lâ€™analyse en composantes principales conserve la distance euclidienne et que lâ€™analyse de correspondance conserve la distance du <span class="math inline">\(\chi^2\)</span>, le PoMd conserve lâ€™association que vous sÃ©lectionnerez Ã  votre convenance. Le PoMd vise Ã  reprÃ©senter en un nombre limitÃ© de dimensions (souvent 2) la distance (ou dissimilaritÃ©) quâ€™ont les objets (ou des variables) les uns par rapport aux autres dans lâ€™espace multidimensionnel.</p>
<p>Il existe deux types dâ€™AEM. Le <strong>PoMd-mÃ©trique</strong> (<em>metric multidimentional scaling</em> MMDS, parfois le <em>metric</em> est retirÃ©, MDS, et parfois lâ€™on parle de <em>classic MDS</em>) vise Ã  reprÃ©senter fidÃ¨lement la distance entre les objets ou les variables. Le PoMd-mÃ©trique ne devrait Ãªtre utilisÃ©e que lorsque la mÃ©trique nâ€™est ni euclidienne, ni de <span class="math inline">\(\chi^2\)</span> et que lâ€™on dÃ©sire prÃ©server les distances entre les objets. Lâ€™PoMd-mÃ©trique aussi appelÃ©e <em>analyse en coordonnÃ©es principales</em> (ACoP ou de lâ€™anglais <em>PCoA</em>) .</p>
<p>Le <strong>PoMd-non-mÃ©trique</strong> (<em>nonmetric multidimentional scaling</em>, NMDS) vise quant Ã  lui Ã  reprÃ©senter lâ€™ordre des distances entre les objets ou les variables. Câ€™est une approche par rang: le PoMd-non-mÃ©trique vise reprÃ©senter les objets sont plus proches ou plus Ã©loignÃ©es les uns des autres plutÃ´t que de reprÃ©senter leur similaritÃ© dans lâ€™espace multidimentionnelle.</p>
<p>Lâ€™<strong>IsoMap</strong>, pour <em>isometric feature mapping</em>, est une extension du PoMd qui reconstruit les distances selon les points retrouvÃ©s dans le voisinage. Les isomaps sont en mesure dâ€™aplatir des donnÃ©es ayant des formes complexes.</p>
<p>Nous ne traitons pour lâ€™instant que de lâ€™PoMd-mÃ©trique (fonction <code>vegan::cmdscale</code>) et des PoMd-non-mÃ©trique (fonction <code><a href="https://rdrr.io/pkg/vegan/man/metaMDS.html">vegan::metaMDS</a></code>).</p>
<section id="application-5" class="level5" data-number="10.4.1.3.1"><h5 data-number="10.4.1.3.1" class="anchored" data-anchor-id="application-5">
<span class="header-section-number">10.4.1.3.1</span> Application</h5>
<p>Utilisons les donnÃ©es dâ€™abondance que nous avions au tout dÃ©but de ce chapitre. La matrice dâ€™association de Bray-Curtis sera utilisÃ©e.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">assoc_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegdist</a></span><span class="op">(</span><span class="va">abundance</span>, method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">assoc_mat</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>         display_numbers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">assoc_mat</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/aop-pomd-abun-data-1.png" class="img-fluid figure-img" width="288"></p>
</figure>
</div>
</div>
</div>
<p>Les sites 2 et 3 devraient Ãªtre plus prÃ¨s lâ€™un et lâ€™autre, puis les sites 3 et 4. Les autres associations sont Ã©loignÃ©s dâ€™environ la mÃªme distance. LanÃ§ons le calcul de la PoMd-mÃ©trique.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pcoa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html">cmdscale</a></span><span class="op">(</span><span class="va">assoc_mat</span>, k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">spec_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/wascores.html">wascores</a></span><span class="op">(</span><span class="va">pcoa</span><span class="op">$</span><span class="va">points</span>, <span class="va">abundance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">pcoa</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">'t'</span>, cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>species scores not available</code></pre>
</div>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">spec_scores</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">spec_scores</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, cex <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/aop-pomd-abun-compute-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>On observe en effet que les sites 2 et 3 sont les plus prÃ¨s. Les sites 3 et 4sont plus Ã©loignÃ©s. Les sites 1, 2 et 4 font Ã  peu prÃ¨s un triangle Ã©quilatÃ©ral, ce qui correspond Ã  ce Ã  quoi on devrait sâ€™attendre. Les wa-scores permettent de juxtaposer les espÃ¨ces sur les sites, pour rÃ©fÃ©rence. Le colibri nâ€™est prÃ©sent que sur le site 2. Le site 1 est populÃ© par des jaseurs et des mÃ©sanges, et câ€™est le seul site oÃ¹ lâ€™on a observÃ© une citelle. On a observÃ© des chardonnerets sur les sites 2 et 3. Sur le site 4, on nâ€™a observÃ© que des bruants, que lâ€™on a aussi observÃ© ailleurs, sauf au site 2.</p>
<p>Le PoMd-non-mÃ©trique (<em>non metric dimensional scaling, NMDS</em>) fonctionne de la mÃªme maniÃ¨re que la PoMd-mÃ©trique, Ã  la diffÃ©rence que la distance est basÃ©e sur les rangs. Ã€ cet Ã©gard, le site 4 Ã  une distance de 0.76 du site 3, mais plutÃ´t le deuxiÃ¨me plus loin, aprÃ¨s le site 2 et avant le site 1. Utilisons la fonction <code>metaMDS</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nmds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/metaMDS.html">metaMDS</a></span><span class="op">(</span><span class="va">assoc_mat</span>, k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">abundance</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Run 0 stress 0 
Run 1 stress 0 
... Procrustes: rmse 0.1216366  max resid 0.1634103 
Run 2 stress 0 
... Procrustes: rmse 0.1403208  max resid 0.1888214 
Run 3 stress 0 
... Procrustes: rmse 0.115492  max resid 0.1523862 
Run 4 stress 0 
... Procrustes: rmse 0.1551062  max resid 0.202592 
Run 5 stress 0 
... Procrustes: rmse 0.09908794  max resid 0.1384734 
Run 6 stress 0 
... Procrustes: rmse 0.0987008  max resid 0.1454209 
Run 7 stress 0 
... Procrustes: rmse 0.1164685  max resid 0.1501751 
Run 8 stress 0 
... Procrustes: rmse 0.0831961  max resid 0.09943451 
Run 9 stress 0 
... Procrustes: rmse 0.1429865  max resid 0.1921223 
Run 10 stress 0 
... Procrustes: rmse 0.09337227  max resid 0.1214715 
Run 11 stress 0 
... Procrustes: rmse 0.06970139  max resid 0.1001732 
Run 12 stress 0 
... Procrustes: rmse 0.08842632  max resid 0.1314237 
Run 13 stress 0 
... Procrustes: rmse 0.08200456  max resid 0.1163193 
Run 14 stress 0 
... Procrustes: rmse 0.09483784  max resid 0.1239188 
Run 15 stress 0 
... Procrustes: rmse 0.1513612  max resid 0.1946446 
Run 16 stress 0 
... Procrustes: rmse 0.08664967  max resid 0.1141698 
Run 17 stress 0 
... Procrustes: rmse 0.1179912  max resid 0.1610405 
Run 18 stress 0 
... Procrustes: rmse 0.1099474  max resid 0.1455303 
Run 19 stress 0 
... Procrustes: rmse 0.1063117  max resid 0.1398709 
Run 20 stress 0 
... Procrustes: rmse 0.1009788  max resid 0.1310601 
*** Best solution was not repeated -- monoMDS stopping criteria:
    20: stress &lt; smin</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in metaMDS(assoc_mat, k = nrow(abundance) - 1, eig = TRUE): stress is
(nearly) zero: you may have insufficient data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in postMDS(out$points, dis, plot = max(0, plot - 1), ...): skipping
half-change scaling: too few points below threshold</code></pre>
</div>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spec_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/wascores.html">wascores</a></span><span class="op">(</span><span class="va">nmds</span><span class="op">$</span><span class="va">points</span>, <span class="va">abundance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">nmds</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">'t'</span>, cex <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>species scores not available</code></pre>
</div>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">spec_scores</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">spec_scores</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span>, cex <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/aop-pomd-abun-compute-nmds-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Dans ce cas, entre PoMd-mÃ©trique et non-mÃ©trique, les rÃ©sultats peuvent Ãªtre interprÃ©tÃ©s de maniÃ¨re similaire.</p>
<p>En ce qui a trait au dauphin,</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/aop-pomd-dauphin-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Pour plus de dÃ©tails, je vous invite Ã  vous rÃ©fÃ©rer Ã  <a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>) ou de consulter lâ€™excellent site <a href="https://mb3is.megx.net/gustame/dissimilarity-based-methods/nmds">GUSTA ME</a>.</p>
</section></section><section id="conclusion-sur-lordination-non-contraignante" class="level4" data-number="10.4.1.4"><h4 data-number="10.4.1.4" class="anchored" data-anchor-id="conclusion-sur-lordination-non-contraignante">
<span class="header-section-number">10.4.1.4</span> Conclusion sur lâ€™ordination non contraignante</h4>
<p>Lorsque les donnÃ©es sont euclidiennes, lâ€™analyse en composantes principales (ACP) devrait Ãªtre utilisÃ©e. Lorsque la mÃ©trique est celle du <span class="math inline">\(\chi^2\)</span>, on prÃ©fÃ©rera lâ€™analyse de correspondance (AC). Si la mÃ©trique est autre, le positionnement multidimensionel (PoMd) est prÃ©fÃ©rable. Dans ce dernier cas, si lâ€™on recherche une reprÃ©sentation simplifiÃ©e de la distance entre les objets ou variables, on utilisera un PoMd-mÃ©trique. Ã€ lâ€™inverse, si lâ€™on dÃ©sire une reprÃ©sentation plus fidÃ¨le au rang des distances, on prÃ©fÃ©rera lâ€™PoMd-non-mÃ©trique.</p>
</section></section><section id="ordination-contraignante" class="level3" data-number="10.4.2"><h3 data-number="10.4.2" class="anchored" data-anchor-id="ordination-contraignante">
<span class="header-section-number">10.4.2</span> Ordination contraignante</h3>
<p>Alors que lâ€™ordination non contraignante vous permet de dresser un portrait de vos variables, lâ€™ordination contraignante (ou canonique) permet de tester statistiquement ainsi que de reprÃ©senter la relation entre plusieurs variables explicatives (par exemple, des conditions environnementales) et une ou plusieurs variables rÃ©ponses (par exemple, les espÃ¨ces observÃ©es).</p>
<ul>
<li>Lâ€™analyse discriminante nâ€™a fondamentalement quâ€™une seulement variable rÃ©ponse, et celle-ci doit dÃ©crire lâ€™appartenance Ã  une catÃ©gorie.</li>
<li>Lâ€™analyse de redondance sera prÃ©fÃ©rÃ©e lorsque le nombre de variable est plus restreint (variables ionomiques et indicateurs de performance des cultures). Les dÃ©tails, ainsi que les tenants et aboutissants de ces mÃ©thodes, sont prÃ©sentÃ©s dans <a href="https://www.elsevier.com/books/numerical-ecology/legendre/978-0-444-53868-0">Numerical Ecology (Legendre et Legendre, 2012)</a>.</li>
<li>Lâ€™analyse canonique des corrÃ©lations sera prÃ©fÃ©rÃ©e lorsque les variables sont parsemÃ©es (beaucoup de colonnes avec beaucoup de zÃ©ros, comme les variables dâ€™abondance).</li>
</ul>
<section id="analyse-discriminante" class="level4" data-number="10.4.2.1"><h4 data-number="10.4.2.1" class="anchored" data-anchor-id="analyse-discriminante">
<span class="header-section-number">10.4.2.1</span> Analyse discriminante</h4>
<p>Alors que lâ€™analyse en composante principale vise Ã  prÃ©senter la perspective (les axes) selon laquelle les points sont les plus Ã©clatÃ©es, lâ€™analyse discriminante, le plus souvent utilisÃ© dans sa forme linÃ©aire (ADL) et quadratique (ADQ), vise Ã  prÃ©senter la perspective selon laquelle les groupes sont les plus Ã©clatÃ©s, les groupes formant la variable contraignante. Ces groupes peuvent Ãªtre connus (e.g.&nbsp;cultivar, rÃ©gion gÃ©ographique) ou attribuÃ©s (exemple: par partitionnement). Lâ€™ADL est parfois nommÃ©e <em>analyse canonique de la variance</em>.</p>
<p>Lâ€™AD vise Ã  reprÃ©senter des diffÃ©rences entre des groupes aux moyens de combinaisons linÃ©aires (ADL) ou quadratique (ADQ) de variables mesurÃ©es. Sa reprÃ©sentation sous forme de biplot permet dâ€™apprÃ©cier les diffÃ©rences entre les groupes dâ€™identifier les variables qui sont responsables de la discrimination.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/07_ionome-revisited-figure4d.png" class="img-fluid figure-img" alt="&quot;" width="600"></p>
</figure>
</div>
<center>
Biplot de distance de lâ€™analyse discriminante des ionomes dâ€™espÃ¨ces de plantes Ã  fruits cultivÃ©es sauvages et domestiquÃ©es, Source: <a href="https://doi.org/10.3389/fpls.2013.00039">Parent et al.&nbsp;(2013)</a>
</center>
<p>Lâ€™ADL a Ã©tÃ© dÃ©veloppÃ©e par <a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1469-1809.1936.tb02137.x/abstract">Fisher (1936)</a>, qui Ã  titre dâ€™exemple dâ€™application a utilisÃ© un jeu de donnÃ©es de dimensions dâ€™iris collectÃ©es par Edgar Anderson, du Jardin botanique du Missouri, sur 150 spÃ©cimens dâ€™iris collectÃ©s en GaspÃ©sie (Est du QuÃ©bec), ma rÃ©gion natale (suis-je assez chauvin?). Ce jeu de donnÃ©es est amplement utilisÃ© Ã  titre dâ€™exemple en analyse multivariÃ©e.</p>
<p><a href="http://www.jstor.org/stable/1937836">Williams (1983)</a> a prÃ©sentÃ© les tenants et aboutissants de lâ€™ADL en Ã©cologie. Tout comme les donnÃ©es passant pas une ACP doivent suivre une distribution multinormale pour Ãªtre statistiquement valide, les distributions des groupes dans une ADL doivent Ãªtre multinormales et les variances des points par groupe doivent Ãªtre homogÃ¨nesâ€¦ ce qui est rarement le cas en science. NÃ©anmoins:</p>
<blockquote class="blockquote">
<p>Heureusement, il y a des Ã©vidences dans la littÃ©rature que certaines dâ€™entre [ces rÃ¨gles] peuvent Ãªtre transgressÃ©es modÃ©rÃ©ment sans de grands changement dans les taux de classification. Cette conclusion dÃ©pends, toutefois, de la sÃ©vÃ©ritÃ© des transgressions, et de facteurs structueaux comme la position relative des moyennes des populations et de la nature des dispersions. - <a href="http://www.jstor.org/stable/1937836">Williams (1983)</a></p>
</blockquote>
<p>Lâ€™ADL peut servir autant dâ€™outil dâ€™interprÃ©tation que dâ€™outil de classification, câ€™est Ã  dire de prÃ©dire une catÃ©gorie selon les variables (<a href="12-autoapprentissage.html">chapitre&nbsp;<span>13</span></a>). Dans les deux cas, lorsque le nombre de variables approchent le nombre dâ€™observation, les rÃ©sultats dâ€™une ADL risque dâ€™Ãªtre difficilement interprÃ©tables. Le test appropriÃ© pour Ã©valuer lâ€™homodÃ©nÃ©itÃ© de la covariance est le M-test de Box. Ce test est peu documentÃ© dans la littÃ©rature, est rarement utilisÃ© mais a la <a href="https://en.wikiversity.org/wiki/Box%27s_M">rÃ©putation</a> dâ€™Ãªtre particuliÃ¨rement sÃ©vÃ¨re.</p>
<p>Il est rare que des donnÃ©es Ã©cologiques aient des dispersions (covariances) homogÃ¨nes. Contrairement Ã  lâ€™ADL, lâ€™ADQ ne demande pas Ã  ce que les dispersions (covariances) soient homogÃ¨nes. NÃ©anmoins, lâ€™ADQ ne gÃ©nÃ¨re ni de scores, ni de loadings: il sâ€™agit dâ€™un outil pour prÃ©dire des catÃ©gories (classification), non pas dâ€™un outil dâ€™ordination.</p>
<section id="application-6" class="level5" data-number="10.4.2.1.1"><h5 data-number="10.4.2.1.1" class="anchored" data-anchor-id="application-6">
<span class="header-section-number">10.4.2.1.1</span> Application</h5>
<p>Utilisons les donnÃ©es dâ€™iris.</p>
<div class="cell">
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"iris"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Testons la multinormalitÃ© par groupe. Rappelons-nous que pour considÃ©rer la distribution comme multinormale, la p-value de la distorsion ainsi que la statistique de Kurtosis doivent Ãªtre Ã©gale ou plus Ã©levÃ©e que 0.05. La fonction <code>split</code> sÃ©pare le tableau en listes et la fonction <code>map</code> applique la fonction spÃ©cifiÃ©e Ã  chaque Ã©lÃ©ment de la liste. Cela permet dâ€™effectuer des tests de multinormalitÃ© sur chacune des espÃ¨ces dâ€™iris.</p>
<div class="cell">
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">iris</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/MVN/man/mvn.html">mvn</a></span><span class="op">(</span><span class="va">.x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>,</span>
<span>            mvnTest <span class="op">=</span> <span class="st">"mardia"</span><span class="op">)</span><span class="op">$</span><span class="va">multivariateNormality</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$setosa
             Test        Statistic           p value Result
1 Mardia Skewness 25.6643445196298 0.177185884467652    YES
2 Mardia Kurtosis 1.29499223711605 0.195322907441935    YES
3             MVN             &lt;NA&gt;              &lt;NA&gt;    YES

$versicolor
             Test         Statistic           p value Result
1 Mardia Skewness  25.1850115362466 0.194444483140265    YES
2 Mardia Kurtosis -0.57186635893429 0.567412516528727    YES
3             MVN              &lt;NA&gt;              &lt;NA&gt;    YES

$virginica
             Test         Statistic           p value Result
1 Mardia Skewness  26.2705981752915 0.157059707690356    YES
2 Mardia Kurtosis 0.152614173978342 0.878702546726567    YES
3             MVN              &lt;NA&gt;              &lt;NA&gt;    YES</code></pre>
</div>
</div>
<p>Le test est passÃ© pour toutes les espÃ¨ces. Voyons maintenant lâ€™homogÃ©nÃ©itÃ© de la covariance. Pour ce faire, nous aurons besoin de la fonction <code>boxM</code>, disponible avec le module <code>biotools</code>. Pour que les covariances soient considÃ©rÃ©es comme Ã©gales, la p-vaule doit Ãªtre supÃ©rieure Ã  0.05.</p>
<div class="cell">
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://friendly.github.io/heplots/">"heplots"</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: broom</code></pre>
</div>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://friendly.github.io/heplots/reference/boxM.html">boxM</a></span><span class="op">(</span><span class="va">iris</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>,</span>
<span>     group <span class="op">=</span> <span class="va">iris</span><span class="op">$</span><span class="va">Species</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Box's M-test for Homogeneity of Covariance Matrices

data:  select(iris, -Species)
Chi-Sq (approx.) = 140.94, df = 20, p-value &lt; 2.2e-16</code></pre>
</div>
</div>
<p>On est loin dâ€™un cas oÃ¹ les distributions sont homogÃ¨nes. Nous allons nÃ©anmoins procÃ©der Ã  lâ€™analyse discriminante avec le module ade4. Nous aurons dâ€™abord besoin dâ€™effectuer une ACP avec la fonction <code>dudi.pca</code> de ade4 (en spÃ©cifiant une mise Ã  lâ€™Ã©chelle), que nous projetterons en ADL avec <code>discrimin</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://pbil.univ-lyon1.fr/ADE-4/">"ade4"</a></span><span class="op">)</span></span>
<span><span class="va">iris_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ade4/man/dudi.pca.html">dudi.pca</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">iris</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/MASS/man/lm.ridge.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Species</span><span class="op">)</span>,</span>
<span>                     scannf <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># ne pas gÃ©nÃ©rer de graphique</span></span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">iris_lda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ade4/man/discrimin.html">discrimin</a></span><span class="op">(</span>dudi <span class="op">=</span> <span class="va">iris_pca</span>,</span>
<span>                      fac <span class="op">=</span> <span class="va">iris</span><span class="op">$</span><span class="va">Species</span>,</span>
<span>                      scannf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La visualisation peut Ãªtre effectuÃ©e directement sur lâ€™objet issu de la fonction <code>discrimin</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">iris_lda</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-lda-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Il sâ€™agit toutefois dâ€™une visualisation pour le diagnostic davantage que pour la publication. Si lâ€™objectif est la publication, vous pourriez utiliser la fonction <code>plotDA</code> que jâ€™ai conÃ§ue Ã  cet effet. Jâ€™ai aussi conÃ§u une <a href="https://raw.githubusercontent.com/essicolo/AgFun/master/plotDA_trad.R">fonction similaire qui utilise le module graphique de base de R</a>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/essicolo/AgFun/master/plotDA_gg.R"</span><span class="op">)</span></span>
<span><span class="fu">plotDA</span><span class="op">(</span>scores <span class="op">=</span> <span class="va">iris_lda</span><span class="op">$</span><span class="va">li</span>,</span>
<span>       loadings <span class="op">=</span> <span class="va">iris_lda</span><span class="op">$</span><span class="va">fa</span>,</span>
<span>       fac <span class="op">=</span> <span class="va">iris</span><span class="op">$</span><span class="va">Species</span>,</span>
<span>       level<span class="op">=</span><span class="fl">0.95</span>,</span>
<span>       facname <span class="op">=</span> <span class="st">"Species"</span>,</span>
<span>       propLoadings <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ellipse</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ellipse'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:car':

    ellipse</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:graphics':

    pairs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: grid</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: plyr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>------------------------------------------------------------------------------</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>You have loaded plyr after dplyr - this is likely to cause problems.
If you need functions from both plyr and dplyr, please load plyr first, then dplyr:
library(plyr); library(dplyr)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>------------------------------------------------------------------------------</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'plyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    arrange, count, desc, failwith, id, mutate, rename, summarise,
    summarize</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    compact</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
â„¹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-lda-biplot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Ã€ la diffÃ©rence de lâ€™ACP, lâ€™ADL maximise la sÃ©paration des groupes. Nous avions notÃ© avec lâ€™ACP que les dimensions des pÃ©tales distinguaient les groupes. Puisque nous avions justement des informations sur les groupes, nous aurions pu procÃ©der directement Ã  un ADL pour obtenir des conclusions plus directes. Si la longueur des pÃ©tales permet de distinguer lâ€™espÃ¨ce <em>setosa</em> des deux autres, la largeur des pÃ©tales permet de distinguer <em>virginica</em> et <em>versicolor</em>, bien que les nuages de points se superposent. De maniÃ¨re bivariÃ©e, les rÃ©gions de confiance des moyennes des scores discriminants (petites ellipses) montrent des diffÃ©rence significatives au seuil 0.05.</p>
<div class="alert alert-block alert-success">
<strong>Excercice</strong>. Si lâ€™on effectuait lâ€™ADL sur notre dauphin, avec la colonne <code>anatomy</code> comme variable de regroupement, quâ€™obtiendrions-nous? Si lâ€™on considÃ¨re la nageoire codale (queue) comme faisant partie du corps? Quelles sont les limitations?
</div>
</section></section><section id="analyse-de-redondance-rda" class="level4" data-number="10.4.2.2"><h4 data-number="10.4.2.2" class="anchored" data-anchor-id="analyse-de-redondance-rda">
<span class="header-section-number">10.4.2.2</span> Analyse de redondance (RDA)</h4>
<p>En anglais, on la nomme <em>redundancy analysis</em>, souvent abrÃ©gÃ©e RDA. Elle est utilisÃ©e pour rÃ©sumer les relations linÃ©aires entre des variables rÃ©ponse et des variables explicatives. La â€œredondanceâ€ se situe dans lâ€™utilisation de deux tableaux de donnÃ©es contenant de lâ€™information concordante. Lâ€™analyse de redondance est une maniÃ¨re Ã©lÃ©gante dâ€™effectuer une rÃ©gression linÃ©aire multiple, oÃ¹ la matrice de valeurs prÃ©dites par la rÃ©gression est assujettie Ã  une analyse en composantes principales. Il est ainsi possible de superposer les scores des variables explicatives Ã  ceux des variables rÃ©ponse.</p>
<p>Plus prÃ©cisÃ©ment, une RDA effectue les Ã©tapes suivantes (<a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>) entre une matrice de variables indÃ©pendantes (explicatives) <span class="math inline">\(X\)</span> et une matrice de variables dÃ©pendantes (rÃ©ponse) <span class="math inline">\(Y\)</span>.</p>
<section id="rÃ©gression-entre-y-et-x" class="level5" data-number="10.4.2.2.1"><h5 data-number="10.4.2.2.1" class="anchored" data-anchor-id="rÃ©gression-entre-y-et-x">
<span class="header-section-number">10.4.2.2.1</span> 1. RÃ©gression entre <span class="math inline">\(Y\)</span> et <span class="math inline">\(X\)</span>
</h5>
<p>Pour chacune des variables rÃ©ponse de <span class="math inline">\(Y\)</span> (<span class="math inline">\(y_1\)</span>, <span class="math inline">\(y_2\)</span>, , <span class="math inline">\(y_j\)</span>), effectuer une rÃ©gression linÃ©aire sur les variables explicatives <span class="math inline">\(X\)</span>.</p>
<p><span class="math display">\[\hat{y}_j = b_j + m_{1, j} \times x_1 + m_{2, j} \times x_2 + ... + m_{i, j} \times x_i\]</span></p>
<p><span class="math display">\[\hat{y}_j = y_j + y_{res, j}\]</span></p>
<p>Pour chaque observation (<span class="math inline">\(n\)</span>), nous obtenons une sÃ©rie de valeurs de <span class="math inline">\(\hat{y}_j\)</span> et de <span class="math inline">\(y_{res, j}\)</span>. Donc chaque cellule de la matrice <span class="math inline">\(Y\)</span> a ses pendant <span class="math inline">\(\hat{y}\)</span> et <span class="math inline">\(y_{res}\)</span>. Nous obtenons ainsi une matrice de prÃ©diction <span class="math inline">\(\hat{Y}\)</span> et une matrice des rÃ©sidus <span class="math inline">\(Y_{res} = Y - \hat{Y}\)</span>.</p>
</section><section id="analyse-en-composantes-principales-1" class="level5" data-number="10.4.2.2.2"><h5 data-number="10.4.2.2.2" class="anchored" data-anchor-id="analyse-en-composantes-principales-1">
<span class="header-section-number">10.4.2.2.2</span> 2. Analyse en composantes principales</h5>
<p>Ensuite, on effectue une analyse en composantes principales (ACP) sur la matrice des prÃ©dictions <span class="math inline">\(\hat{Y}\)</span>. On obtient ainsi ses valeurs et vecteurs propres. Nommons <span class="math inline">\(U\)</span> ses vecteurs propres. Les fonctions de RDA mettent souvent ces vecteurs Ã  lâ€™Ã©chelle avant de les retourner Ã  lâ€™utilisateur. En ordination Ã©cologique, ces vecteurs mis Ã  lâ€™Ã©chelle sont souvent appelÃ©s les <em>scores des espÃ¨ces</em>, bien quâ€™il ne sâ€™agisse pas nÃ©cessairement dâ€™espÃ¨ces, mais plus gÃ©nÃ©ralement des variables de la matrice dÃ©pendante <span class="math inline">\(Y\)</span>.</p>
<p>Il est aussi possible dâ€™effectuer une ACP sur <span class="math inline">\(Y_{res}\)</span>.</p>
</section><section id="calculer-les-scores" class="level5" data-number="10.4.2.2.3"><h5 data-number="10.4.2.2.3" class="anchored" data-anchor-id="calculer-les-scores">
<span class="header-section-number">10.4.2.2.3</span> 3. Calculer les scores</h5>
<p>Les vecteurs propres <span class="math inline">\(U\)</span> sont utilisÃ©s pour calculer les <em>scores des sites</em>, <span class="math inline">\(Y \times U\)</span>, ainsi que les <em>contraintes de site</em> <span class="math inline">\(\hat{Y} \times U\)</span>.</p>
</section><section id="application-7" class="level5" data-number="10.4.2.2.4"><h5 data-number="10.4.2.2.4" class="anchored" data-anchor-id="application-7">
<span class="header-section-number">10.4.2.2.4</span> Application</h5>
<p>Nous allons utiliser la fonction <code>rda</code> du module vegan. En ce qui a trait aux donnÃ©es, utilisons les donnÃ©es varespec (matrice Y) et varechem (matrice X). La fonction <code>rda</code> peut fonctionner avec lâ€™interface-formule de R, oÃ¹ Ã  gauche du <code>~</code> on retrouve le Y (la matrice de la communautÃ© Ã©cologique, i.e.&nbsp;les abondances dâ€™espÃ¨ces) contre le X (l), Ã  gauche, ce qui peut Ãªtre pratique pour lâ€™analyse dâ€™interactions. Mais pour comparer deux matrices, nous pouvons dÃ©finir X et Y. Ce qui est mÃ©langeant, câ€™est que vegan, contrairement aux conventions, dÃ©fini X comme Ã©tant la matrice rÃ©ponse et Y comme Ã©tant la matrice explicative.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vare_rda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">rda</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">varespec</span>, Y <span class="op">=</span> <span class="va">vareclr</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="va">vare_rda</span>, scaling <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"text"</span>, main <span class="op">=</span> <span class="st">"Scaling 1: triplot de distance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="va">vare_rda</span>, scaling <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"text"</span>, main <span class="op">=</span> <span class="st">"Scaling 2: triplot de corrÃ©lation"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-vare-rda-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>La fonction <code>ordiplot</code> permet de crÃ©er un triplot de base. La reprÃ©sentation des wascores est rÃ©putÃ©e plus robuste (moins susceptible dâ€™Ãªtre bruitÃ©e), mais leur interprÃ©tation porte Ã  confusion (<a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>).</p>
<p><strong>Triplot de distance (scaling 1)</strong>. Les angles entre les variables explicatives reprÃ©sentent leur corrÃ©lation (non pas les variables rÃ©ponse).</p>
<p><strong>Triplot de corrÃ©lation (scaling 2)</strong>. Les angles entre les variables reprÃ©sentent leurs corrÃ©lation, que les variables soient rÃ©ponse ou explicative, ou entre variables rÃ©ponses et variables explicatives. Les distances entre les objets sur le triplot ne sont pas des approximation de leur distance euclidienne.</p>
<p>Les triplots montrent que les variables ont toutes un rÃ´le important sur la dispersion des sites autour des axes principaux. Le premier axe principal est composÃ© de maniÃ¨re plus marquÃ©e par le clr de lâ€™Al et celui du Fe. Le deuxiÃ¨me axe principal est composÃ© de maniÃ¨re plus marquÃ©e par le clr du S, du P et du K. Le triplot de corrÃ©lation ne prÃ©sente pas de tendance apprÃ©ciable pour la plupart des espÃ¨ces, qui ne possÃ¨dent pas de niche particuliÃ¨re. Toutefois, lâ€™espÃ¨ce Cladstel, prÃ©sente surtout dans les sites 9 et 10, est liÃ©e Ã  de basses teneurs en N et Ã  de faibles valeurs de Baresoil (sol nu). Lâ€™espÃ¨ce Pleuschr est liÃ©e Ã  des sols oÃ¹ lâ€™on retrouve une grande Ã©paisseur dâ€™humus, ainsi que des teneurs Ã©levÃ©es en nutriment K, P, S, Ca, Mg et Zn. Elle semble apprÃ©cier les sols Ã  bas pH, mais Ã  faible teneur en Fe et Al. La teneur en N lui semble plus indiffÃ©rente (son vecteur Ã©tant presque perpendiculaire).</p>
<p>On pourra personnaliser les graphiques en extrayant les scores.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scaling</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">sites</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">vare_rda</span>, display <span class="op">=</span> <span class="st">"wa"</span>, scaling <span class="op">=</span> <span class="va">scaling</span><span class="op">)</span></span>
<span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">vare_rda</span>, display <span class="op">=</span> <span class="st">"species"</span>, scaling <span class="op">=</span> <span class="va">scaling</span><span class="op">)</span></span>
<span><span class="va">env</span> <span class="op">&lt;-</span> <span class="fu">vegan</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/scores.html">scores</a></span><span class="op">(</span><span class="va">vare_rda</span>, display <span class="op">=</span> <span class="st">"reg"</span>, scaling <span class="op">=</span> <span class="va">scaling</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, type <span class="op">=</span> <span class="st">"n"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>, asp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, v <span class="op">=</span> <span class="fl">0</span>, col <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">sites</span><span class="op">/</span><span class="fl">2</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sites</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fl">0.7</span>, col <span class="op">=</span> <span class="st">"grey50"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">species</span><span class="op">/</span><span class="fl">2</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"green"</span>, cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu">segments</span><span class="op">(</span>x0 <span class="op">=</span> <span class="fl">0</span>, y0 <span class="op">=</span> <span class="fl">0</span>, x1 <span class="op">=</span> <span class="va">env</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, y1 <span class="op">=</span> <span class="va">env</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">env</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">env</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"blue"</span>, cex <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-vare-rda-plot-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<p>On pourra effectuer une analyse de Kaiser-Guttmann ou de broken-stick de la mÃªme maniÃ¨re que prÃ©cÃ©demment. Ã‰tant une collection de rÃ©gressions, une RDA est en mesure dâ€™effectuer des tests statistiques sur les coefficients de la rÃ©gression en utilisant des permutations pour tester la signification des coefficients et des axes dâ€™une RDA. On doit nÃ©anmoins obligatoirement effectuer la RDA avec lâ€™interface formule. La variable Ã  gauche du <code>~</code> peut Ãªtre une matrice ou un tableau, et celui de droite est dÃ©fini dans <code>data</code>. Le <code>.</code> dans lâ€™interface formule signifie â€œune combinaison linÃ©aire de toutes les variables, sans interactionâ€.</p>
<div class="cell">
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">vare_rda</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">rda</a></span><span class="op">(</span><span class="va">varespec</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">vareclr</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">perm_test_term</span> <span class="op">&lt;-</span> <span class="fu">anova</span><span class="op">(</span><span class="va">vare_rda</span>, by <span class="op">=</span> <span class="st">"term"</span><span class="op">)</span></span>
<span><span class="co">#perm_test_axis &lt;- anova(vare_rda, by = "axis")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>La signification des axes est difficile Ã  interprÃ©ter. Toutefois, celui des variables prÃ©sente un intÃ©rÃªt.</p>
<div class="cell">
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">perm_test_term</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

Model: rda(formula = varespec ~ N + P + K + Ca + Mg + S + Al + Fe + Mn + Zn + Mo + Fv + Baresoil + Humdepth + pH, data = vareclr, scale = FALSE)
         Df Variance      F Pr(&gt;F)   
N         1   216.13 4.8470  0.006 **
P         1   272.71 6.1159  0.005 **
K         1   194.97 4.3724  0.012 * 
Ca        1    24.92 0.5589  0.661   
Mg        1    52.61 1.1799  0.310   
S         1   100.07 2.2441  0.089 . 
Al        1   177.91 3.9900  0.008 **
Fe        1   118.59 2.6595  0.050 * 
Mn        1    25.96 0.5822  0.638   
Zn        1    35.81 0.8030  0.483   
Mo        1    23.51 0.5273  0.678   
Baresoil  1    98.64 2.2122  0.103   
Humdepth  1    43.59 0.9777  0.393   
pH        1    38.93 0.8730  0.447   
Residual  9   401.31                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>La p-value est la probabilitÃ© que les pentes calculÃ©es pour les variables Ã©mergent de distributions dont la moyenne est nulle. Au seuil 0.05, les variables significatives sont (les clr de) lâ€™azote, le phosphore, le potassium et lâ€™aluminium.</p>
<p>Dans le cas des matrices dâ€™abondance (ce nâ€™est pas le cas de varespec, constituÃ©e de donnÃ©es de recouvrement), il est prÃ©fÃ©rable avec les RDA de les transformer prÃ©alablement avec la transformation compositionnelle, de chord ou de Hellinger (<a href="08-explorer.html">chapitre&nbsp;<span>9</span></a>). Une autre option est dâ€™effectuer une RDA sur des matrices dâ€™association en passant par une analyse en coordonnÃ©es principales (<a href="http://www.jstor.org/stable/2657192">Legendre et Anderson, 1999</a>). Enfin, les donnÃ©es dâ€™abondance Ã  lâ€™Ã©tat brutes devraient plutÃ´t passer utiliser une analyse canonique des corrÃ©lations.</p>
</section></section><section id="analyse-canonique-des-correspondances-acc" class="level4" data-number="10.4.2.3"><h4 data-number="10.4.2.3" class="anchored" data-anchor-id="analyse-canonique-des-correspondances-acc">
<span class="header-section-number">10.4.2.3</span> Analyse canonique des correspondances (ACC)</h4>
<p>Lâ€™analyse canonique des correspondances (<em>Canonical correspondance analysis</em>), ACC, a Ã©tÃ© Ã  lâ€™origine conÃ§ue pour Ã©tudier les liens entre des variables environnementales et lâ€™abondance (dÃ©compte) ou lâ€™occurrence (prÃ©sence-absence) dâ€™espÃ¨ces (<a href="https://www.ohio.edu/plantbio/staff/mccarthy/multivariate/terBraak1986.pdf">ter Braak, 1986</a>). Lâ€™ACC est Ã  la RDA ce que la CA est Ã  lâ€™ACP. Alors que la RDA prÃ©serve les distance euclidiennes entre variables dÃ©pendantes et indÃ©pendantes, lâ€™ACC prÃ©serve les distances du <span class="math inline">\(\chi^2\)</span>. Tout comme lâ€™AC, elle hÃ©rite du coup une propriÃ©tÃ© importante de la distance du <span class="math inline">\(\chi^2\)</span>: il y a davantage davantage dâ€™importance aux espÃ¨ces rares.</p>
<p>Lâ€™analyse des correspondances canoniques est souvent utilisÃ©e dans la littÃ©rature, mais dans bien des cas une RDA sur des donnÃ©es dâ€™abondance transformÃ©es donnera des rÃ©sultats davantage interprÃ©tables (<a href="https://doi.org/10.1007/s004420100716">Legendre et Gallagher, 2001</a>).</p>
<section id="application-8" class="level5" data-number="10.4.2.3.1"><h5 data-number="10.4.2.3.1" class="anchored" data-anchor-id="application-8">
<span class="header-section-number">10.4.2.3.1</span> Application</h5>
<p>Cet exemple dâ€™application concerne des donnÃ©es dâ€™abondance. Nous allons consÃ©quemment utiliser une CCA avec la fonction <code>cca</code>, toujours avec le module vegan.</p>
<p>Les tableaux <a href="https://rdrr.io/cran/ade4/man/doubs.html"><code>doubs_fish</code> et <code>doubs_env</code></a> comprennent respectivement des donnÃ©es dâ€™abondance dâ€™espÃ¨ces de poissons et dans diffÃ©rents environnements de la riviÃ¨re Doubs (Europe) publiÃ©es dans <a href="https://www.worldcat.org/title/cours-deau-de-franche-comte-massif-du-jura-recherches-ecologiques-sur-le-reseau-hydrographique-du-doubs-essai-de-biotypologie/oclc/496763306">Verneaux. (1973)</a> et exportÃ©es du module <a href="https://rdrr.io/rforge/ade4/"><code>ade4</code></a>.</p>
<div class="cell">
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"doubs"</span><span class="op">)</span></span>
<span><span class="va">doubs_fish</span> <span class="op">&lt;-</span> <span class="va">doubs</span><span class="op">$</span><span class="va">fish</span></span>
<span><span class="va">doubs_env</span> <span class="op">&lt;-</span> <span class="va">doubs</span><span class="op">$</span><span class="va">env</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Sur le site no 8, aucun poisson nâ€™a pas Ã©tÃ© observÃ©. Les observations ne comprenant que des zÃ©ro doivent Ãªtre prÃ©alablement retirÃ©es.</p>
<div class="cell">
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tot_spec</span> <span class="op">&lt;-</span> <span class="va">doubs_fish</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/transmute.html">transmute</a></span><span class="op">(</span>tot_spec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">.</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">doubs_fish</span> <span class="op">&lt;-</span> <span class="va">doubs_fish</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">tot_spec</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using one column matrices in `filter()` was deprecated in dplyr 1.1.0.
â„¹ Please use one dimensional logical vectors instead.</code></pre>
</div>
<div class="sourceCode" id="cb180"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">doubs_env</span> <span class="op">&lt;-</span> <span class="va">doubs_env</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">tot_spec</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>De la mÃªme maniÃ¨re quâ€™avec la fonction <code>rda</code> de vegan, nous utilisons <code>cca</code> pour lâ€™ACC.</p>
<div class="cell">
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">doubs_cca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/vegan/man/cca.html">cca</a></span><span class="op">(</span><span class="va">doubs_fish</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">doubs_env</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Comparons les rÃ©sultats</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="va">doubs_cca</span>, scaling <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"text"</span>, main <span class="op">=</span> <span class="st">"CCA - Scaling 1 - Triplot de distance"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vegan/man/ordiplot.html">ordiplot</a></span><span class="op">(</span><span class="va">doubs_cca</span>, scaling <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"text"</span>, main <span class="op">=</span> <span class="st">"CCA - Scaling 2 - Triplot de corrÃ©lation"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="09-ordination_files/figure-html/apo-doubs-cca-ordiplot-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p><strong>Triplot de distance (scaling 1)</strong>.</p>
<blockquote class="blockquote">
<ol type="1">
<li>La projection des variables rÃ©ponse Ã  angle droit sur les variables explicatives est une approximation de la rÃ©ponse sur lâ€™explication. (2) Un objet (site ou rÃ©ponse) situÃ© prÃ¨s dâ€™une variable explicative est plus susceptible dâ€™avoir le dÃ©compte <code>1</code>. (3) Les distances entre les variables (rÃ©ponse et explicatives) approximent la distance du <span class="math inline">\(\chi^2\)</span> (traduction adaptÃ©e de <a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>).</li>
</ol>
</blockquote>
<p><strong>Triplot de corrÃ©lation (scaling 2)</strong>.</p>
<blockquote class="blockquote">
<ol type="1">
<li>La valeur optmiale de lâ€™espÃ¨ce sur une variable environnementale quantitative peut Ãªtre obtenue en projetant lâ€™espÃ¨ce Ã  angle droit sur la variable. (2) Une espÃ¨ce se trouvant prÃ¨s dâ€™une variable environnementale est susceptible de se trouver en plus grande abondance aux sites de statut <code>1</code> pour cette variable. (3) Les distances nâ€™approximent pas la distance du <span class="math inline">\(\chi^2\)</span> (traduction adaptÃ©e de <a href="http://www.springer.com/us/book/9781441979759">Borcard et al.&nbsp;(2011)</a>).</li>
</ol>
</blockquote>


</section></section></section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./08-explorer.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Explorer R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./10-imputation.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">DÃ©tection de valeurs aberrantes et imputation de donnÃ©es manquantes</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">Analyse et modÃ©lisation dâ€™agroÃ©cosystÃ¨mes est construit avec <a href="https://quarto.org/">Quarto</a>.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Lien vers le <a href="https://essicolo.github.io/ecologie-mathematique-R/">manuel original</a>.</div>
  </div>
</footer>


</body></html>